{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions and Imports for Ovarian Cancer Ratio Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports \n",
    "This includes transcriptional regulatory interactions from TRRUST v2 (https://www.grnpedia.org/trrust/downloadnetwork.php)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "import seaborn as sns\n",
    "import cptac\n",
    "tf_data = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/ProteinComplexes/Data_Files/trrust_rawdata.human.tsv', sep='\\t', header=None)\n",
    "tf_data.columns = ['TF', 'Regulated_Gene', 'Regulation_Type', 'ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All files already downloaded and correct.\n",
      "All files already downloaded and correct.\n",
      "All files already downloaded and correct.\n",
      "All files already downloaded and correct.\n",
      "                                    \r"
     ]
    }
   ],
   "source": [
    "# Create a workable dictionary of protein complex information (all cancer types)\n",
    "# Select for human complexes\n",
    "complexData = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/ProteinComplexes/Data_Files/allComplexes.txt', sep='\\t')\n",
    "complexData = complexData.loc[complexData['Organism'] == 'Human']\n",
    "\n",
    "# Split the proteins in each complex into invdividual columns of a new df\n",
    "subunitNames = complexData['subunits(Gene name)'].str.split(';', expand = True)\n",
    "subunitNames.index = complexData.ComplexName\n",
    "\n",
    "# Create a dictionary (key = complex name, value = list of proteins in complex) and remove None values\n",
    "subunitNames = subunitNames.loc[~subunitNames.index.duplicated()]\n",
    "subunitNames = subunitNames.transpose().to_dict('list')\n",
    "for key, val in subunitNames.items():\n",
    "    subunitNames[key] = [value for value in val if value != None]\n",
    "\n",
    "# Ovarian Data\n",
    "\n",
    "cptac.download(dataset='ovarian', version='latest')\n",
    "ov = cptac.Ovarian()\n",
    "ov_clinical = ov.get_clinical()\n",
    "ov_mutations = ov.get_mutations()\n",
    "ov_mutation_data = ov_clinical[['Patient_ID', 'Sample_Tumor_Normal']].join(ov_mutations).dropna(axis = 0)\n",
    "ov_mutation_data = ov_mutation_data.loc[ov_mutation_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "#Only for ovarian: Import non-normalized proteomics data for ratio analysis\n",
    "data = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/ProteinComplexes/Data_Files/proteinGroups_cleaned.txt', sep='\\t')\n",
    "\n",
    "# Remove duplicate rows from the dataset\n",
    "data = data.set_index('Gene_Name')\n",
    "data = data.loc[~data.index.duplicated()]\n",
    "\n",
    "# Endometrial Data\n",
    "cptac.download(dataset='endometrial', version='latest')\n",
    "en = cptac.Endometrial()\n",
    "en_clinical = en.get_clinical()\n",
    "en_mutations = en.get_mutations()\n",
    "en_mutation_data = en_clinical[['Patient_ID', 'Proteomics_Tumor_Normal']].join(en_mutations).dropna(axis = 0)\n",
    "en_mutation_data = en_mutation_data.loc[en_mutation_data['Proteomics_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "# Colon Data\n",
    "cptac.download(dataset='colon', version='latest')\n",
    "colon = cptac.Colon()\n",
    "colon_clinical = colon.get_clinical()\n",
    "colon_mutations = colon.get_mutations()\n",
    "colon_mutation_data = colon_clinical[['Patient_ID', 'Sample_Tumor_Normal']].join(colon_mutations).dropna(axis = 0)\n",
    "colon_mutation_data = colon_mutation_data.loc[colon_mutation_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "# Renal Data\n",
    "cptac.download(dataset='renalccrcc', version='latest')\n",
    "renal = cptac.RenalCcrcc()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define basic analysis and plotting functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get_ratio_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "get_ratio_df\n",
    "-----------\n",
    "All patients, single protein, tumor vs normal\n",
    "Returns two dataframes: tumor and normal containing ratios for the proteins\n",
    "\n",
    "Parameters:\n",
    "    prot1, prot2 = (gene) names of the two proteins to be compared (ratio of prot1 / prot2)\n",
    "    \n",
    "Returns:\n",
    "    Two dataframes: tumor and normal containing ratios for the proteins\n",
    "  \n",
    "\"\"\"\n",
    "def get_ratio_df(prot1, prot2):\n",
    "    # Make sure that both proteins are in our dataset\n",
    "    if not data.index.contains(prot1) or not data.index.contains(prot2): return None, None\n",
    "    \n",
    "    # Create a dataframe of ratios for easy testing\n",
    "    tumor_rows_list = []\n",
    "    normal_rows_list = []\n",
    "    for patient in data.columns:\n",
    "        \n",
    "        # Determine the sample type\n",
    "        sample_type = 'Tumor'\n",
    "        if '_NM' in patient: sample_type = 'Normal'\n",
    "            \n",
    "        # Find the ratio prot1 / prot2 and create a new row for the dataframe\n",
    "        if data.at[prot2, patient] == 0: \n",
    "            ratio = np.nan\n",
    "        else: \n",
    "            ratio = data.at[prot1, patient] / data.at[prot2, patient]\n",
    "        # Create a row entry for the dictionary\n",
    "        row_dict = {'Ratio': ratio, 'Sample_Type': sample_type, 'Patient_ID': patient}\n",
    "        \n",
    "        # Add the new row to the tumor or normal list, depending on the sample type\n",
    "        if sample_type == 'Tumor':\n",
    "            tumor_rows_list.append(row_dict)\n",
    "        else:\n",
    "            normal_rows_list.append(row_dict)\n",
    "    \n",
    "    # Convert the row lists into dataframes\n",
    "    tumor_ratio_df = pd.DataFrame(tumor_rows_list)\n",
    "    normal_ratio_df = pd.DataFrame(normal_rows_list)\n",
    "    \n",
    "    # If there were no valid ratios for either the tumor or normal samples, return None\n",
    "    if (not np.isfinite(tumor_ratio_df['Ratio']).any()) or (not np.isfinite(normal_ratio_df['Ratio']).any()): \n",
    "        return None, None\n",
    "    \n",
    "    # Drop na values\n",
    "    tumor_ratio_df = tumor_ratio_df.dropna(axis = 0)\n",
    "    normal_ratio_df = normal_ratio_df.dropna(axis = 0)\n",
    "    \n",
    "    return tumor_ratio_df, normal_ratio_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test_complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "test_complex\n",
    "------------\n",
    "Perform statistical tests on every combination of proteins in a given complex, printing those with significant p-values\n",
    "\n",
    "Parameters:\n",
    "    complex_name = name of the complex\n",
    "    test_type = type of test to perform, levene or ttest\n",
    "    \n",
    "Returns:\n",
    "    Classification of the complex: whether it has proteins missing in the normal, missing proteins in cancer,\n",
    "    or other (proteins present in both)\n",
    "    Prints tests with statistically significant values\n",
    "\n",
    "\"\"\"\n",
    "def test_complex(complex_name, test_type = 'ttest'):\n",
    "    \n",
    "    prot_list = subunitNames[complex_name]\n",
    "    sig_result_found = False\n",
    "    \n",
    "    # Keep track of how many significant ratios in this complex include only 0 values for tumor/normal \n",
    "    num_missing_tumor = 0\n",
    "    num_missing_normal = 0\n",
    "    \n",
    "    # Calculate the cutoff for significance (Bonferroni correction)\n",
    "    alpha = 0.05 / (len(subunitNames)*(len(subunitNames) - 1))\n",
    "    \n",
    "    for i in range(0, len(prot_list)):\n",
    "        for j in range(0, len(prot_list)):\n",
    "            if j == i: continue\n",
    "                \n",
    "            # Test the two proteins (ratio of prot1 / prot2)\n",
    "            tumor_ratio_df, normal_ratio_df = get_ratio_df(prot_list[i], prot_list[j])\n",
    "            if tumor_ratio_df is None or normal_ratio_df is None: continue\n",
    "\n",
    "            # Perform the selected statistical test on the ratios\n",
    "            \n",
    "            # If both tumor and normal only have 0 values for the ratio, the tests don't work\n",
    "            if (sum(tumor_ratio_df['Ratio']) == 0) and (sum(normal_ratio_df['Ratio']) == 0): continue   \n",
    "            # If there are not enough samples, the tests also don't work\n",
    "            if (len(tumor_ratio_df) < 2) or (len(normal_ratio_df) < 2): continue\n",
    "                \n",
    "            if test_type == 'ttest':\n",
    "                test_result = scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "            else:\n",
    "                # This test throws a warning if the variance in both is 0, so we check for that\n",
    "                if (np.var(tumor_ratio_df['Ratio']) == 0) and (np.var(normal_ratio_df['Ratio']) == 0): continue\n",
    "                test_result = scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "            \n",
    "            # If the p-value is significant, print the result and return the appropriate classification\n",
    "            if test_result < alpha:\n",
    "                sig_result_found = True\n",
    "                # Determine the classification of this ratio\n",
    "                if not (normal_ratio_df['Ratio'] != 0).any():\n",
    "                    num_missing_normal += 1\n",
    "                    print('Normal missing ' + prot_list[i] + ' / ' + prot_list[j])\n",
    "                elif not (tumor_ratio_df['Ratio'] != 0).any():\n",
    "                    num_missing_tumor += 1\n",
    "                    print('Tumor missing ' + prot_list[i] + ' / ' + prot_list[j])\n",
    "    \n",
    "                print(prot_list[i] + ' / ' + prot_list[j] + ': ' + str(test_result))\n",
    "                \n",
    "    if sig_result_found: \n",
    "        \n",
    "        print(complex_name)\n",
    "        print('---------------------------')\n",
    "        print('---------------------------')\n",
    "        \n",
    "        # Classify the complex as a whole depending on the classification of the majority of ratios\n",
    "        # If none of the ratios had all 0 values for tumor or normal, it is classified as \"Other\"\n",
    "        if num_missing_normal == 0 and num_missing_tumor == 0:\n",
    "            return \"Other\"\n",
    "        elif num_missing_normal >= num_missing_tumor:\n",
    "            return \"Normal Missing Proteins\"\n",
    "        elif num_missing_tumor > num_missing_normal:\n",
    "            return \"Tumor Missing Proteins\"\n",
    "        \n",
    "    return \"NS\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test_all_complexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "test_all_complexes\n",
    "------------------\n",
    "Test all complexes in the dataset for significance\n",
    "\n",
    "Parameters:\n",
    "    test_type (optional) = ttest or levene\n",
    "    \n",
    "Returns:\n",
    "    A dictionary: {\"Normal Missing Proteins\": [List of complexes], \n",
    "    \"Tumor Missing Proteins\": [List of complexes], \n",
    "    \"Other\": [List of complexes]}\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "def test_all_complexes(test_type = 'ttest'):\n",
    "    \n",
    "    if test_type != 'ttest' and test_type != 'levene':\n",
    "        print(\"Error: test_type must equal 'ttest' or 'levene'\")\n",
    "        return\n",
    "    \n",
    "    classified_complexes = {\"Normal Missing Proteins\": [], \"Tumor Missing Proteins\": [], \"Other\": [], \"NS\": []}\n",
    "    \n",
    "    for key, val in subunitNames.items():\n",
    "        complex_type = test_complex(key, test_type = test_type)\n",
    "        classified_complexes[complex_type].append(key)\n",
    "        \n",
    "    return classified_complexes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_mutations\n",
    "--------------\n",
    "Search mutational data for mutations in proteins that change in a complex or transcription factors\n",
    "known to regulate said proteins\n",
    "\n",
    "Parameters:\n",
    "    complex_dict = dictionary of protein complexes (output of test_all_complexes)\n",
    "    transcription_factors = if True, will test for mutations in transcription factors known to regulate\n",
    "        proteins in the complex\n",
    "        \n",
    "Returns:\n",
    "    Prints results (how many patients have mutations in the protein or transcription factor)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def find_mutations(complex_dict, transcription_factors = False):\n",
    "    \n",
    "    for key, val in complex_dict.items():\n",
    "        if key == 'NS': continue\n",
    "        print(key)\n",
    "        print('--------------')\n",
    "        print('--------------')\n",
    "        for complex_name in val:\n",
    "            changed_proteins = []\n",
    "            mutation_rate = {}\n",
    "\n",
    "            # Calculate the cutoff for significance (Bonferroni correction)\n",
    "            alpha = 0.05 / (len(subunitNames)*(len(subunitNames) - 1))\n",
    "\n",
    "            # Perform tests to find statistically significant differences in ratios\n",
    "            protein_list = subunitNames[complex_name]\n",
    "            for prot1 in protein_list:\n",
    "                for prot2 in protein_list:\n",
    "                    if prot1 == prot2: continue\n",
    "\n",
    "                    # Test the two proteins (ratio of prot1 / prot2)\n",
    "                    tumor_ratio_df, normal_ratio_df = get_ratio_df(prot1, prot2)\n",
    "                    if tumor_ratio_df is None or normal_ratio_df is None: continue\n",
    "\n",
    "                    # Perform the selected statistical tests on the ratios\n",
    "                    \n",
    "                    # If both tumor and normal only have 0 values for the ratio, the tests don't work\n",
    "                    if (sum(tumor_ratio_df['Ratio']) == 0) and (sum(normal_ratio_df['Ratio']) == 0): continue\n",
    "                    # If there are not enough samples, the tests also don't work\n",
    "                    if (len(tumor_ratio_df) < 2) or (len(normal_ratio_df) < 2): continue\n",
    "                \n",
    "                    ttest_result = scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "                    # The levene test throws a warning if the variance in both is 0, so we check for that\n",
    "                    if (np.var(tumor_ratio_df['Ratio']) != 0) and (np.var(normal_ratio_df['Ratio']) != 0):\n",
    "                        levene_result = scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "                    else:\n",
    "                        levene_result = 1\n",
    "\n",
    "                    # If the p-value is significant, add the numerator protein to changed_proteins\n",
    "                    if ttest_result < alpha or levene_result < alpha:\n",
    "                        changed_proteins.append(prot1)\n",
    "\n",
    "            changed_proteins = list(set(changed_proteins))\n",
    "            for protein in changed_proteins:\n",
    "                if transcription_factors:\n",
    "                    tfs = tf_data.loc[tf_data['Regulated_Gene'] == protein]['TF']\n",
    "                    for tf in tfs:\n",
    "                        num_mutations = len(set(mutation_data.loc[mutation_data['Gene'] == tf]['Patient_ID']))\n",
    "                        tf_title = tf + ' (regulates ' + protein + ')'\n",
    "                        if num_mutations > 0:\n",
    "                            mutation_rate[tf_title] = num_mutations\n",
    "                else:\n",
    "                    num_mutations = len(set(mutation_data.loc[mutation_data['Gene'] == protein]['Patient_ID']))\n",
    "                    if num_mutations > 0:\n",
    "                        mutation_rate[protein] = num_mutations\n",
    "            if len(mutation_rate) > 0:\n",
    "                print(complex_name)\n",
    "                print(mutation_rate)\n",
    "                print('--------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_unique_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_unique_mutations\n",
    "---------------------\n",
    "Find the number of patients with a mutation in any one of the proteins in a list\n",
    "\n",
    "Parameters:\n",
    "    prot_list = list of proteins to look for mutations in\n",
    "    \n",
    "Returns:\n",
    "    prints the number of patients with a mutation in a protein in prot_list as well as their identifiers\n",
    "    returns a list of those patients\n",
    "\"\"\"\n",
    "\n",
    "def find_unique_mutations(prot_list, cancer_type = 'ov'):\n",
    "    # Set the cancer type\n",
    "    if cancer_type == 'ov': mutation_data = ov_mutation_data\n",
    "    elif cancer_type == 'colon': mutation_data = colon_mutation_data\n",
    "    elif cancer_type == 'en': mutation_data = en_mutation_data\n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "        \n",
    "    patients = []\n",
    "    mutated_proteins = {}\n",
    "    for protein in prot_list:\n",
    "        mutations = list(mutation_data.loc[mutation_data['Gene'] == protein]['Patient_ID'])\n",
    "        if len(mutations) > 0:\n",
    "            mutated_proteins[protein] = len(mutations)\n",
    "            patients = patients + mutations\n",
    "    patients = list(set(patients))\n",
    "    print('Patients with a mutation in any of the ' + str(len(prot_list)) + ' given proteins: ' + str(len(patients)) + '\\n')\n",
    "    print(mutated_proteins)\n",
    "    return patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_ratios\n",
    "-----------\n",
    "Create a seaborn plot for the ratios in a dataframe\n",
    "\n",
    "Parameters:\n",
    "    prot1, prot2 = the two proteins to plot (ratio of prot1 / prot2)\n",
    "    by_patient (optional) = T/F whether or not to visualize matched samples\n",
    "    mutation_list (optional) = plot will distinguish patients with a mutation in any of the proteins in this list\n",
    "    \n",
    "Returns:\n",
    "    Displays a plot of the data\n",
    "    \n",
    "\"\"\"\n",
    "def plot_ratios(prot1, prot2, by_patient = False, mutation_list = None):\n",
    "    \n",
    "    # Set up dataframe for plotting\n",
    "    tumor_ratio_df, normal_ratio_df = get_ratio_df(prot1, prot2)\n",
    "    plot_data = tumor_ratio_df.append(normal_ratio_df)\n",
    "    \n",
    "    # Reformat for visualizing matched samples if necessary\n",
    "    if by_patient or mutation_list:\n",
    "        # Create a new column for matched status\n",
    "        tumor_ratio_df['Matched_Status'] = 'Unmatched'\n",
    "        normal_ratio_df['Matched_Status'] = 'Unmatched'\n",
    "        # Classify samples as matched/unmatched\n",
    "        normal_ratio_df.loc[((normal_ratio_df['Patient_ID']).str.replace('_NM', '')).isin(tumor_ratio_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_ratio_df.loc[((tumor_ratio_df['Patient_ID']) + '_NM').isin(normal_ratio_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        # Label samples accordingly in plot_data\n",
    "        plot_data = tumor_ratio_df.append(normal_ratio_df)\n",
    "        if mutation_list:\n",
    "            # Find patients with mutations in the given proteins\n",
    "            patient_list = find_unique_mutations(mutation_list)\n",
    "            plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Matched_Status'] = 'Has_Mutation'\n",
    "        plot_data.loc[plot_data['Matched_Status'] == 'Unmatched', 'Patient_ID'] = 'Unmatched_Sample'\n",
    "        plot_data['Patient_ID'] = plot_data['Patient_ID'].str.replace('_NM','')\n",
    "    \n",
    "    # Reformat for visualizing a certain list of patients if necessary\n",
    "    if mutation_list and not by_patient:\n",
    "        plot_data.loc[~plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Other Mutation'\n",
    "        plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Given Mutation'\n",
    "    elif mutation_list:\n",
    "        plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Given Mutation'\n",
    "        \n",
    "    # Print results of statistical tests\n",
    "    print('T-test p-value: ' + str(scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]))\n",
    "    print('Levene p-value: ' + str(scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]))\n",
    "    \n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "\n",
    "    # Create the plot\n",
    "    if by_patient or patient_list:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y='Ratio', color='w', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y='Ratio', hue='Patient_ID', size=10, dodge=True, jitter=True)\n",
    "        boxplt.get_legend().set_bbox_to_anchor((1, 1, 0, 0))\n",
    "    else:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y='Ratio', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y='Ratio', dodge=True, jitter=True, color='.3')\n",
    "\n",
    "    # Add styling\n",
    "    boxplt.set_title('Ratio of ' + prot1 + ' / ' + prot2, fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Protein Expression Ratio', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='15')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ttest_proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "ttest_proteomics\n",
    "----------------\n",
    "Performs and prints a basic t-test on the proteomics of a given protein (tumor vs normal)\n",
    "\n",
    "Parameters:\n",
    "    protein = protein to test\n",
    "    use_cptac (optional) = T/F whether to use the CPTAC proteomics data (default) or the non-normalized data\n",
    "    \n",
    "Returns:\n",
    "    p-value of the t-test\n",
    "\"\"\"\n",
    "\n",
    "def ttest_proteomics(protein, use_cptac = True, cancer_type = 'ov'):\n",
    "    \n",
    "    # This uses the normalized proteomics data from the CPTAC package\n",
    "    if use_cptac:\n",
    "        # Get data from the appropriate cancer type\n",
    "        if cancer_type == 'ov':\n",
    "            test_data = ov.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'colon': \n",
    "            test_data = colon.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'en': \n",
    "            test_data = en.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "            \n",
    "        test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "        if cancer_type == 'en':\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            test_data.loc[test_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "        else:\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "        tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "        normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "    \n",
    "    # Otherwise use the non-normalized data to make the plot\n",
    "    else:\n",
    "        test_data = pd.DataFrame(data.loc[data.index == protein].transpose())\n",
    "        test_data['Sample_Type'] = 'Tumor'\n",
    "        test_data.loc[test_data.index.str.contains('_NM'), 'Sample_Type'] = 'Normal'\n",
    "\n",
    "        tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Perform the t-test\n",
    "    return scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_differential_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_differential_expression\n",
    "----------------------------\n",
    "Searches through all the proteomic data and finds proteins with differential expression in tumor cells using t-tests.\n",
    "Organizes results into categories: tumor lower and normal lower\n",
    "\n",
    "Parameters:\n",
    "    cancer_type = 'ov', 'en', or 'colon'; the cancer to search for\n",
    "    \n",
    "Returns:\n",
    "    A dictionary of the form {tumor_lower: [list_of_proteins], normal_lower: [list_of_proteins]}\n",
    "    'tumor_lower' means that there are lower expression levels of that protein in tumor samples\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "def find_differential_expression(cancer_type = 'ov'):\n",
    "    results = {'tumor_lower': [], 'normal_lower': []}\n",
    "    \n",
    "    # Get data from the appropriate cancer type\n",
    "    if cancer_type == 'ov':\n",
    "        alpha = 0.05 / len(ov.get_proteomics().columns)\n",
    "        for protein in ov.get_proteomics().columns:\n",
    "            test_data = ov.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                    metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                    omics_genes = protein)\n",
    "            test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "\n",
    "            tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "            normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "            \n",
    "            pval = scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]\n",
    "            if pval < alpha:\n",
    "                mean_tumor = tumor_df[protein].mean()\n",
    "                mean_normal = normal_df[protein].mean()\n",
    "                if mean_tumor < mean_normal: \n",
    "                    results['tumor_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (tumor lower)')\n",
    "                else: \n",
    "                    results['normal_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (normal lower)')\n",
    "            \n",
    "        \n",
    "    elif cancer_type == 'colon': \n",
    "        alpha = 0.05 / len(colon.get_proteomics().columns)\n",
    "        for protein in colon.get_proteomics().columns:\n",
    "            test_data = colon.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                    metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                    omics_genes = protein)\n",
    "            test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "\n",
    "            tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "            normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "            \n",
    "            pval = scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]\n",
    "            if pval < alpha:\n",
    "                mean_tumor = tumor_df[protein].mean()\n",
    "                mean_normal = normal_df[protein].mean()\n",
    "                if mean_tumor < mean_normal: \n",
    "                    results['tumor_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (tumor lower)')\n",
    "                else: \n",
    "                    results['normal_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (normal lower)')\n",
    "        \n",
    "    elif cancer_type == 'en': \n",
    "        alpha = 0.05 / len(en.get_proteomics().columns)\n",
    "        for protein in en.get_proteomics().columns:\n",
    "            test_data = en.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                    metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                    omics_genes = protein)\n",
    "            test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            test_data.loc[test_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "\n",
    "            tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "            normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "            \n",
    "            pval = scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]\n",
    "            if pval < alpha:\n",
    "                mean_tumor = tumor_df[protein].mean()\n",
    "                mean_normal = normal_df[protein].mean()\n",
    "                if mean_tumor < mean_normal: \n",
    "                    results['tumor_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (tumor lower)')\n",
    "                else: \n",
    "                    results['normal_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (normal lower)')\n",
    "        \n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "    \n",
    "    return results\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_phosphoproteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/24685012/pandas-dataframe-renaming-multiple-identically-named-columns\n",
    "# Rename duplicated columns in a dataframe so that each column has a unique name\n",
    "def df_column_uniquify(df):\n",
    "    df_columns = df.columns\n",
    "    new_columns = []\n",
    "    for item in df_columns:\n",
    "        counter = 1\n",
    "        newitem = item\n",
    "        while newitem in new_columns:\n",
    "            counter += 1\n",
    "            newitem = \"{}_{}\".format(item, counter)\n",
    "        new_columns.append(newitem)\n",
    "    df.columns = new_columns\n",
    "    return df\n",
    "\n",
    "def plot_phosphoproteomics(protein, by_patient = False, print_pvals = True, remove_duplicates = False, cancer_type = 'ov'):\n",
    "    \n",
    "    # Get data from the appropriate cancer type\n",
    "    if cancer_type == 'ov':\n",
    "        phos_data = ov.join_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    elif cancer_type == 'colon': \n",
    "        phos_data = colon.join_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    elif cancer_type == 'en': \n",
    "        phos_data = en.join_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "    \n",
    "    if remove_duplicates: phos_data = phos_data.loc[:, ~phos_data.columns.duplicated()]\n",
    "    else: phos_data = df_column_uniquify(phos_data)\n",
    "    plot_data = pd.melt(phos_data, id_vars = ['Patient_ID', 'Sample_Tumor_Normal'], var_name = 'Location', value_name = 'Reading').dropna(axis = 0)\n",
    "    plot_data['Location'] = plot_data['Location'].str.replace('_phosphoproteomics', '')\n",
    "    \n",
    "    # Perform t-tests on tumor vs normal phosphoproteomics for each site\n",
    "    if print_pvals:\n",
    "        for column in phos_data.columns:\n",
    "            if column != 'Patient_ID' and column != 'Sample_Tumor_Normal':\n",
    "                sitedf = phos_data[['Patient_ID', 'Sample_Tumor_Normal', column]]\n",
    "                tumordf = sitedf.loc[sitedf['Sample_Tumor_Normal'] == 'Tumor'].dropna(axis = 0)\n",
    "                normaldf = sitedf.loc[sitedf['Sample_Tumor_Normal'] == 'Normal'].dropna(axis = 0)\n",
    "                if len(tumordf) > 2 and len(normaldf) > 2:\n",
    "                    pval = scipy.stats.ttest_ind(tumordf[column], normaldf[column])[1]\n",
    "                    print(column + ' t-test: ' + str(pval))\n",
    "    \n",
    "    # Plot the data\n",
    "    a4_dims = (20, 20)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "\n",
    "    boxplt = sns.boxplot(data=plot_data, x='Location', y='Reading', hue='Sample_Tumor_Normal', showfliers=False, palette = 'Blues')\n",
    "    boxplt = sns.stripplot(data=plot_data, x='Location', y='Reading', hue='Sample_Tumor_Normal', dodge=True, jitter=True, color='.3')\n",
    "\n",
    "    # Add styling\n",
    "    boxplt.set_title(protein + ' Phosphorylation', fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Phosphorylation Level', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='10')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_proteomics\n",
    "---------------\n",
    "Create a seaborn plot of the proteomics data for a single protein\n",
    "\n",
    "Parameters:\n",
    "    protein = the protein to plot\n",
    "    use_cptac (optional) = T/F whether to use the CPTAC proteomics data (default) or the non-normalized data\n",
    "    by_patient (optional) = T/F whether or not to visualize matched samples\n",
    "    print_pvals (optional) = T/F whether or not to print the p-values of the statistical tests\n",
    "    \n",
    "Returns:\n",
    "    Displays a plot of the data\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def plot_proteomics(protein, use_cptac = True, by_patient = False, print_pvals = True, cancer_type = 'ov'):\n",
    "    \n",
    "    # This uses the normalized proteomics data from the CPTAC package\n",
    "    if use_cptac:\n",
    "        # Get data from the appropriate cancer type\n",
    "        if cancer_type == 'ov':\n",
    "            plot_data = ov.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'colon': \n",
    "            plot_data = colon.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'en': \n",
    "            plot_data = en.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "            \n",
    "        plot_data = plot_data.loc[:, ~plot_data.columns.duplicated()]\n",
    "        if cancer_type == 'en':\n",
    "            plot_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            plot_data.loc[plot_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "        else:\n",
    "            plot_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "        plot_data = plot_data.dropna(axis = 0)\n",
    "        plot_data['Matched_Status'] = 'Unmatched'\n",
    "        tumor_df = plot_data.loc[plot_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = plot_data.loc[plot_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Otherwise use the non-normalized data to make the plot\n",
    "    else:\n",
    "        plot_data = pd.DataFrame(data.loc[data.index == protein].transpose())\n",
    "        plot_data['Sample_Type'] = 'Tumor'\n",
    "        plot_data.loc[plot_data.index.str.contains('_NM'), 'Sample_Type'] = 'Normal'\n",
    "\n",
    "        plot_data['Patient_ID'] = plot_data.index\n",
    "        plot_data['Matched_Status'] = 'Unmatched'\n",
    "        tumor_df = plot_data.loc[plot_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = plot_data.loc[plot_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Format to show matched patients if necessary\n",
    "    if by_patient:\n",
    "        pd.options.mode.chained_assignment = None \n",
    "        \n",
    "        # Classify samples as matched/unmatched\n",
    "        normal_df.loc[((normal_df['Patient_ID']).str.replace('_NM|N', '')).isin(tumor_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[((tumor_df['Patient_ID']) + '_NM').isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[((tumor_df['Patient_ID']) + 'N').isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[('N' + (tumor_df['Patient_ID'])).isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        \n",
    "        # Label samples accordingly in plot_data\n",
    "        plot_data = tumor_df.append(normal_df)\n",
    "        plot_data.loc[plot_data['Matched_Status'] == 'Unmatched', 'Patient_ID'] = 'Unmatched_Sample'\n",
    "        plot_data['Patient_ID'] = plot_data['Patient_ID'].str.replace('_NM|N','')\n",
    "    \n",
    "    if print_pvals:\n",
    "        # Print results of statistical tests\n",
    "        print('T-test p-value: ' + str(scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]))\n",
    "        print('Levene p-value: ' + str(scipy.stats.levene(tumor_df[protein], normal_df[protein])[1]))\n",
    "    \n",
    "    # Create the plot\n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "    \n",
    "    if by_patient:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y=protein, color='w', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y=protein, hue='Patient_ID', size=10, dodge=True, jitter=True)\n",
    "        boxplt.get_legend().set_bbox_to_anchor((1, 1, 0, 0))\n",
    "    else:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y=protein, showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y=protein, dodge=True, jitter=True, color='.3')\n",
    "        \n",
    "    # Add styling\n",
    "    boxplt.set_title(protein + ' Proteomics', fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Protein Expression', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='15')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_complex_clinical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_complex_clinical\n",
    "---------------------\n",
    "Creates a countplot of the number of proteins in a complex that are differentially expressed in tumor/normal matched samples\n",
    "Also displays the given clinical feature \n",
    "\n",
    "Parameters:\n",
    "    cancer_type = can be colon, \n",
    "    clinical_feature = the name of the column from the clinical dataframe that you want to visualize\n",
    "    protein_list = list of proteins to analyze (most likely the proteins in a complex)\n",
    "    expression_change = increased, decreased, or nochange; will determine if the plot shows the number\n",
    "        of proteins with increased expression in tumor samples, decreased in tumor samples, or no change between\n",
    "        tumor/normal\n",
    "    bin_size = size of the bins to put patients in as number of proteins on the x-axis (default is 1)\n",
    "    complex_name (optional) = The title for the plot\n",
    "    show_percentages = whether to display percentages of each group in each category or raw numbers\n",
    "    \n",
    "Returns:\n",
    "    Displays a Seaborn countplot\n",
    "    x-axis: Number of proteins with differential expression\n",
    "    y-axis: Number of patients in this category\n",
    "    hue: clinical feature\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "def plot_complex_clinical(cancer_type, clinical_feature, protein_list, expression_change = 'increased', bin_size = 1, complex_name = '', show_percentages = True):\n",
    "    if cancer_type == 'colon':\n",
    "        # Start out with a df including the patient IDs and the given clinical feature\n",
    "        plot_data = pd.DataFrame()\n",
    "        plot_data = colon.get_clinical()[['Patient_ID', clinical_feature]]\n",
    "        plot_data['Sample_ID'] = plot_data.index\n",
    "        \n",
    "        # Add protein expression values for all proteins in the given list\n",
    "        proteomics_data = colon.get_proteomics()[protein_list]\n",
    "        plot_data = plot_data.join(proteomics_data)\n",
    "\n",
    "        # Extract all the normal matched samples into a new df and remove the 'N' to match them with tumor samples\n",
    "        normal = plot_data.loc[plot_data['Patient_ID'].str.contains('N')]\n",
    "        normal['Patient_ID'] = normal['Patient_ID'].str.replace('N', '')\n",
    "        normal.index = normal['Patient_ID']\n",
    "\n",
    "        # Create new column to count the number of proteins with increased/decreased/nochange levels in tumor samples\n",
    "        # Remove the normal samples from plot_data\n",
    "        plot_data[expression_change] = 0 \n",
    "        plot_data = plot_data.loc[~plot_data['Patient_ID'].str.contains('N')]\n",
    "        plot_data.index = plot_data['Patient_ID']\n",
    "\n",
    "        # Remove unmatched samples from each dataframe\n",
    "        plot_data = plot_data.loc[plot_data['Patient_ID'].isin(normal['Patient_ID'])]\n",
    "        normal = normal.loc[normal['Patient_ID'].isin(plot_data['Patient_ID'])]\n",
    "\n",
    "        # Fill out the Num_* column by comparing the protein expression for matched patients in plot_data/normal dfs\n",
    "        for protein in protein_list:\n",
    "            mean_diff = np.abs(np.mean(plot_data[protein]) - np.mean(normal[protein]))\n",
    "            if expression_change == 'increased':\n",
    "                plot_data.loc[((plot_data[protein] - normal[protein]) >= mean_diff), expression_change] += 1\n",
    "            elif expression_change == 'decreased':\n",
    "                plot_data.loc[((plot_data[protein] - normal[protein]) <= (-1*mean_diff)), expression_change] += 1\n",
    "            else:\n",
    "                plot_data.loc[((plot_data[protein] - normal[protein]) <= mean_diff)\n",
    "                          & ((plot_data[protein] - normal[protein]) >= (-1*mean_diff)), expression_change] += 1\n",
    "\n",
    "        # Call another function to create the plot\n",
    "        create_complex_clinical_plot(plot_data, clinical_feature, protein_list, bin_size, expression_change, complex_name, show_percentages)\n",
    "    \n",
    "    else: print('Error: invalid or unimplemented cancer type')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create_complex_clinical_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "create_complex_clinical_plot\n",
    "----------------------------\n",
    "Function called by plot_complex_clinical; this prepares the dataframe for plotting and creates the actual plot\n",
    "\n",
    "Parameters are the same as those passed to plot_complex_clinical\n",
    "\n",
    "\"\"\"\n",
    "def create_complex_clinical_plot(plot_data, clinical_feature, protein_list, bin_size, expression_change, complex_name, show_percentages):\n",
    "    # Drop the proteomics data (not necessary at this point)\n",
    "    plot_data.drop(protein_list, axis = 1, inplace=True)\n",
    "    category_col_title = expression_change + ' category'\n",
    "    \n",
    "    # If indicated, bin the patients into groups based on number of proteins increased/decreased/nochange\n",
    "    # This makes it easier to view, otherwise there are lots of ticks on the x-axis\n",
    "    if bin_size > 1:\n",
    "        # Create the list of cutoff values for each bin based on the given bin_size\n",
    "        bin_vals = list(range(0, len(protein_list), bin_size))\n",
    "        # Create the list of strings to use as category names\n",
    "        category_names = []\n",
    "        for i in range(0, len(bin_vals) - 1):\n",
    "            category_names.append(str(bin_vals[i]) + '-' + str(bin_vals[i+1] - 1))\n",
    "        category_names.append(str(bin_vals[len(bin_vals)-1]) + '+')\n",
    "        \n",
    "        # Place each sample into the correct category\n",
    "        plot_data[category_col_title] = category_names[0]\n",
    "        for i in range(1, len(bin_vals) - 1):\n",
    "            plot_data.loc[((plot_data[expression_change] > bin_vals[i]) & (plot_data[expression_change] <= bin_vals[i+1])), category_col_title] = category_names[i]\n",
    "            \n",
    "        # Create the appropriate plot based on the desired differential expression type\n",
    "        # increased means increased in tumor samples\n",
    "        a4_dims = (10, 10)\n",
    "        fig, ax = plt.subplots(figsize=a4_dims)\n",
    "        x_order = category_names\n",
    "        if show_percentages:\n",
    "            percent_df = (plot_data[category_col_title]\n",
    "                            .groupby(plot_data[clinical_feature])\n",
    "                            .value_counts(normalize=True)\n",
    "                            .rename('Percentage ' + clinical_feature)\n",
    "                            .reset_index())\n",
    "            countplot = sns.barplot(data=percent_df, x=category_col_title, y=('Percentage ' + clinical_feature), hue=clinical_feature, order = x_order)\n",
    "            countplot.set_ylabel('Percentage of Individuals', fontsize = '15')\n",
    "            tick_spacing = 0.05\n",
    "            \n",
    "        else:\n",
    "            countplot = sns.countplot(x=category_col_title, hue=clinical_feature, data=plot_data, order = x_order)\n",
    "            countplot.set_ylabel('Number of Individuals', fontsize = '15')\n",
    "            tick_spacing = 2\n",
    "            \n",
    "        countplot.set_xlabel('Number of proteins with ' + expression_change + ' expression in tumor samples', fontsize = '15')      \n",
    "        ax.yaxis.set_major_locator(ticker.MultipleLocator(tick_spacing))\n",
    "        countplot.set_title(complex_name, fontsize = '25')\n",
    "    \n",
    "    else:\n",
    "        # Same idea as above, just without binning the patients\n",
    "        plot_data[category_col_title] = plot_data[expression_change].astype(str)\n",
    "        a4_dims = (10, 10)\n",
    "        fig, ax = plt.subplots(figsize=a4_dims)\n",
    "        \n",
    "        x_axis_labels = list(set(plot_data[category_col_title].astype(int).sort_values()))\n",
    "        x_axis_labels = [str(i) for i in x_axis_labels] \n",
    "        countplot = sns.countplot(x=category_col_title, hue=clinical_feature, data=plot_data, order = x_axis_labels)\n",
    "        tick_spacing = 2\n",
    "        countplot.set_ylabel('Number of Individuals', fontsize = '15')\n",
    "        \n",
    "    countplot.set_xlabel('Number of proteins with ' + expression_change + ' expression in tumor samples', fontsize = '15')      \n",
    "    ax.yaxis.set_major_locator(ticker.MultipleLocator(tick_spacing))\n",
    "    countplot.set_title(complex_name, fontsize = '25')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/envs/ptmresearch/lib/python3.7/site-packages/ipykernel_launcher.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAnAAAAJrCAYAAABgAq77AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xm4HGWZ///3TRICBgZZXQghQdlCEgIkgBvCIJsgKKCCigZ03BCdGYFhdIwaF1TUUX8uuIIiAoobCuqAgIj6AwLKLsoSMGxGwioJScj9/aPqnNPpdPepJN3npJL367r66q6qp6rv3nI+earqqchMJEmSVB/rDHcBkiRJWjEGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJGkVRMSmEfFIRMyLiDE92P7eEZERsdygnZ2WqZrV7T2MiKPKer4z3LVo9WaAU+1FxIf6/gFuui2MiLkRcUFEvCYiYphr/FBEjB+uGtQzHwI2Aj6Vmf8crHFEXN3wHX1Lz6trX8f4Nr+bpyNifkT8ISJmRsRmw1jjK8vfzSuHq4Zh8H3gFuANEbHrcBej1ZcBTmuaBxtuCWwJvAI4D7gwIkYPU10fLG/jh+n51QMRsR3wdmAe8KUK7ScB0xtmvblHpa2oxxj43TwKbAzsCXwYuDkipg1TXa+k+N30MsA9CdxW3oZdZi4FPgIE8OlhLkerMQOc1iiZ+ey+GzAGmARcXC4+CPjosBWnNdF/AiOBb2fmkxXa9wW2M4HHgT0jYmKPalsR72n47WwCbAL8D7AY2AL4UUSsN6wV9khmXp2ZO2TmDsNdS4PzKf5TsM8whmet5gxwWmNl5tLMvBk4FLi9nP22iBg5jGVpDRERGwCvKye/W6H9usAbysnTgR+Wj4/rfnWrJjMfzsyPAZ8sZ20FHDaMJa1VMnMJxa5UgLcNZy1afRngtMbLzIXAD8rJDYH+/2lHxJnlcT9nRuEtEXFlRDxUzp/RuK2I2Kg8Lui6iHgsIhZExF8j4isRsU3zc/dtv2HWZU3HG81pVXN57M9PIuK+iFgUEQ9HxBUR8faIGNVmnY3KA6DPjogby+OYFkbE3RHxvYjYs9P7FBF7lOveVa73z3Ld30TEByJibJv1nh0Rp0XEzRHxRLnezRHxqYh4Vpt1Go+/Gh8RW0fE1yPinvK574iIj0bDSQERMSkivhsRfyvb/DUi/qfd+9FU3yci4vqIeLRc986I+MYq9n4dTfF9ujUzr6/Q/jBgM+C2zLwK+HY5/42DvYZhdFbD4/5dv9F04H9E7FJ+d+ZGxOKIuLx5QxFxeET8PCIeLL/TD5bTr2rRdu9y228qZ70plj9Wb+8W6z0vIv6/iLi1/C4+WT7+XESMa/UCm19L07IZjb/TiNgtIr4fEfdHxFPl9+izEbFxuzdwZX9XwPfK+6Oj+M+CtKzM9Oat1jeKg8iz+Dq3bfPOvjbACxvmn1nO+zZFyEvgaWB+eT+joe1OwN8atrOA4tihvumFwBFNz/t54IGGNvPL6b7bNU3tNwB+1tA+KY5JWtow/Xtg407vQ3l7vKypb3op8O4278+bmp5jYfm8jdub0WK9lwIPN7T5J/BE0+t9cYv1xje0ObxhG48CSxqWXQGMAg4ut53AI021ntvhcz+kfB/62i5qqu8p4I0r+b37YbmN0yu2/2XZ/n3ldAB39b0HHdbbu6/eFVlWoZ7Gz2C5z7Zs84yGNl9r9bzAEeX72vf5LQAub2i7LnBuQ/vG31ffvO8BoxrWeSHF72MBA7+1B5puL2yq9d8a6uj7Dj/ZMP0osN8Kvr8zymVzKHpb+7b/SFP9NwEbdOt3Va47quH1H7Qy31Fva/Zt2Avw5m1Vb1QLcJ9q+Adzh4b5ZzIQdhYD7wX+pVy2AfCc8vGGwJ1l27nAy4F1ymU7A39o+Ad65xbP3/fcew/yWn5ctvsrZQ9POX89il3Bd5TLf9xi3bcDnwX2AJ5ZzgtgAvC58g/JEmCXpvWewUAQPQt4XsOyMcBu5fv38qb1tmIgeN0MvKhh2UuAP5fLHgK2bFp3fMN78jBwCTCxXLY+cAIDQe4j5R/Mc4GtGz6bjzZs42Ut3o/dKQJaUuyy3AEYUS4bR3HSQZaf+7SV+N71nShzXIW2W1H8wV8KjGuYP6vcxs87rLs3bb7fnZZVqKnxM5jRps3EhjafaPW8FL+dC1n2d7Vtw+NPM/AfiFkN382NgY+12n6L3+eZg7yWVzIQ0E8Ftqb47gewPcXuyL4QN65p3U7v7wwG/mOyEPg6sFXD7+Z4BkLdrG78rpq28bt27403b8NegDdvq3pjkAAH/AtwLwNhYp2GZX1/IBI4ocNz/FfDH4hJLZZvyEBvynJ/jBueY+8Oz3Fw2eZ+mgJPQ5uxDPQgTV3B9+mL5XrfaJq/ezn/CWDkCmzvKwz0sj27Ta19vQ1fbFo2vuE9uQkY3WL97zS0+T8gWrS5otVrKpdd3eoPa1Obz5dtfrKC7+U2DbXtVqH9zLLtr5vmP7+cv6TDZ753u+93p2UVamr8DGYM8p1J4JWtnhe4ijIYt1h/S4qAnMDH27T5TMNv6zlNy/p+n2d2eB3rUvynqmOYBn5atvncCry/MxpeZ8saGur/azd+V23e/9+szPre1uybx8BpjRURz4yIfYFLgeeWsz+fxWn6zR4Gvtphc68t78/PzJuaF2bm4xT/mwY4KCI2WomS+8YEOysz723VIDPnApeVkwes4PYvLO9f3DT/kfJ+XWDTKhuKiABeU06enpkPNLcpaz29nDyqw+b+NzOfajH/Vw2PP5GZ2aHNlKb6dqY4ZmsxxR/YdvoGS31ZRIzo0K7Zcxsez+vUsHyvjm16PgAy83aKXeIjGDjea1hFxLoRsWNEfAF4Rzn7r8DP26xyWmY+3WbZERRn6S4EPtGmzUcpekpHAUeuRMkHUQTFB4EzOrTre+9X9HfTp90Z7D8t758fEc9omL/Cv6sW/lHeP7djK62VPBtPa5RWByI3+C7FLptWrsnMRW22uS4DAeGSDtvvG65kHWBXBoJWVX3B6q0R8cYO7frC4dbNC6I4keKdwD7A8yh6Bpv/o9Z80PQdFLs7dwCuioivUASjGzv8YZ5AMdQEDP6enAxsGhETMvOuFm2ubrPugw2PrxmkTfNB5H3v5TrAbdF+DOe+0DaG4o/s39s1bLJ5w+P5g7Tdl6K3658MnHna6NsUx3wdGxGntgmqvXZGRLQLP3cBh2ZxZmQrv+uw3b4hMK7JzMdaNcjMhyNiNvCihvYrou+z3hi4v8NnvW55v9zvpoL5Zdhu5b6GxxtTHHcHK/e7Wu55y/vNO7bSWskApzVN4x/9pyj+B/tH4OzM7BSoOv3h3oSBP/Qte8ZKcxseb9GpyGblWYh9I95vxEBI66Txf/uUZ/OdAzQOVvwYAycyrEvxB2aZyz1l5tMRcRTF8XcTKHpKPgE8GRG/B37E8uOcNb6+FXlPWgW4x9us2x8Yyh7OTm2az+Ls67EYAbQ8E7aFZwzepF/jmGiteg8b9Q0T8uPMfKLF8vModuU+H9gL+M0K1NEtj1EcMA/FsXqPAX+hOPHiO9n5ChOdfjt935NO3xEY+J6s0O+m1PdZr0u1z3r9lXiOdt8/aPie0vA9XMnfVbO+z2SNHINPq8YApzVKFgP4royq/yPu1DuSbR5X0bj77qjMPG9FVo6ITSmOFxpNsct4FnB1Zi5oaLMvbXrLMvP6iNiB4qzNAyh6hHYCXlbe/jsiDs7MG1utXrHMoexZ6ns//5yZO/Zg+w81PN6YNiGmHF6ib5iMN0TEG1q1a/BmhifAvSczz1yZFSv2JvXyO9L3Wf8yMw9aifV7ZhV/VzDQy/1Qm+Vai3kMnDS4viEPoDibsJ3GZR2Pi2qWxVh1j5aTk1dk3dLLKU7WeBh4RWb+pjG8lTqG28xclJk/ysy3ZeZkit02b6d4/VsxMG4ZLBtYOr0njbtrV+g9WUV9x+RtEz24wDzLvpZN2raC17NivSdHRMS/rFxJq6W+70mn7wgMfE9W5jvS91mvzO+m51bwd9Ws77s1lL8d1YQBThpEeWzcDeXkvh2avqy8Xwpc17yZ8r7tAToMHEv06ohY0d9m3x/I2zrsknlZm/ktZeZDmflVijNwAXYpe/qg2BXad3xOlffkoTbHv/VK33u5LgM9YN30FwZ2nS03gHODvktnfZ7ieMR2t40o/kg/g2L4mDXF7PJ+WrsTeyLimTQcK9e0uO+Eoyq/my0jovkEndXOIL+rZhPK+1t7X5nqxgAnVXNueX9kFBckX0Y5UvrJ5eRFmfloU5O+A7if2eE5vlbebwec1KmYiBhTnlzRp+/5tosW16yMiKkMXPapednoVvMbNPbkPQ3leAvFsVtQXJ5sud69iHguA5cBOmeQ5+i22RTHPgJ8LCI6HgQeEZ160ZZTHhPWF9J3b7PNXYGp5eQ5mflEh9tjFMdEwepzgftu+CFF0F2PgcDS7H0Uu/4Xs/xJHlV+Nz+jGHoH4PNNZ4IuZ0U/65W1Mr+rFvYo74djt7pWcwY4qZqvUPQ6jQJ+EREH9fWSRcRkirPLJlCMZfU/LdbvG3rk9e3+wGTmTykOeAb4RBSX59qub3k5vMMeEfFJ4G6WPeD7/yh6KzYBzo6ILRvWeU25vN2B2EdFxO8i4m3RcDmwiBgREQcwMPzDHzLzkYb1Pk4xVMImwCUR8cKGdV9EcbzdMyl66toNIdETZcB8O8UJBuMozgI8svG9j4gtI+INEXExA9f8XBGXl/d7tFneF8TuyeLSWYPpu/bl9Fb/Saijcjicz5eTp0TEh8set75hfj7CwH9WPpuZ9zdtou9385LyWLJWz7GQgSut7Ar8LiIOaPwPTkRMKL/fV5dth8LK/q762j2L4rsLBji1MtwD0Xnztqo3KlyJocO6Z1JhpPey7SQGBgxNiv9BN14WZyFwZJt139DQblG5nTnAlU3tnkHRW5UNtydY/tJDyfJXN/hE0/JHGBgl/k6KHrjl3ieWHay073X8o+n57qVhpP2GdV9aPk9jrY2XqnoYeEmL9cY3tBnf5j3be7DPtaH2OW2W71e+lr7nWlJO/5NlX/PXV+K7M7Vc90nKq3c0LFuPgatUfLri9kYwcHWH/63yPlR5jzo8X+NnMGMF1638vBS7sc9reK5BL6XVsG7fCSJ97eZR/G7mAHs2tX190+e6uPysGy8nl8D7q76Wwb5fnb7LrMLvqlz/rWWbP67oZ+tt7bjZAydVlMUAvjtRBMY/UYSB0RTjPZ0O7JSZ57dZ97vAMcCVFH/wn0MxHtXYpnZPZubRFOO4nUURvNahuHTU3ynOMD2Z4lJF9zatewrwRopx1RZQ9BbeTtFTtgvLjlfV6IJyvTOA6ylC6UYUPXZXAx8oX9ufW7yu31CMc/UZiuN01qE4XulWikso7ZiZv23zvD2XmRdTDM/x3xTv/aMUvYJLgVuAb1JcouyEldj2nyjen/Uprufa6AgGdvt9nwqyOJuzbzfqG5p2kddWFgfxv5biPfkFxRmVG5b3v6C4DuzrMnNxi3Ufphha5VyKsLMRxe9ma5pODsnMsyk+649S7EJ/guIzWEjxe/0ixTGZK9PbujJW+ndVen1532mAca3FInM4xoyUpPorB1z+NnBZZv7rcNejNUNEjKf4z9vjwNhsPw6i1mL2wEnSyjuboidvn4hoeTKDtBL+i6In+1TDm9qxB06SVkFEHExxndBfZObLh7se1VtEbEVx6MMDwPZZnKQhLccrMUjSKsjMCyPi34FnRsQG2fpyWVJVWwOnUuyWN7ypLXvgJEmSamaN7oHbbLPNcvz48cNdhiRJ0qCuvfbaf2Rmx4HH+6zRAW78+PHMnj178IaSJEnDLCLurtrWs1AlSZJqxgAnSZJUMwY4SZKkmlmjj4FrZfHixcydO5eFCz07u07WW289xo4dy6hRo4a7FEmSht1aF+Dmzp3LhhtuyPjx44mI4S5HFWQmDz30EHPnzmXChAnDXY4kScNurduFunDhQjbddFPDW41EBJtuuqm9ppIklda6AAcY3mrIz0ySpAFrZYCTJEmqMwMcRe/OMccc0z+9ZMkSNt98cw455JCO6z3yyCN8+ctfHnT7ze3uu+8+jjzyyJUvuIXx48czefJkdt55Z/bff38eeOABADbYYIOuPk+zqu+BJEnqHgMcMGbMGG666SYWLFgAwMUXX8yWW2456HorG+Ce+9zncv755698wW1cdtllXH/99UybNo2Pf/zjXd9+KwY4SZKGngGudNBBB3HhhRcCcM4553D00Uf3L/vQhz7Epz/96f7pSZMmMWfOHE455RTuuOMOpk6dykknncQTTzzBvvvuy6677srkyZP56U9/CrBcuzlz5jBp0iSgOKni2GOPZfLkyeyyyy5cdtllAJx55pkcfvjhHHjggWy77bacfPLJlV/LXnvtxe23394//f73v5+dd96ZPffckwcffBCAefPmccQRRzB9+nSmT5/O7373u/7Xetxxx7H33nuzzTbb8IUvfKF/O5/97GeZNGkSkyZN4nOf+1zL13bMMcf0v26A17/+9VxwwQWceeaZHHbYYRx44IFsv/32fPjDH+5v893vfpfdd9+dqVOn8ra3vY2nn3668muVJGmtlJlr7G233XbLZrfccsty88aMGZPXX399HnHEEblgwYLceeed87LLLsuDDz44MzM/+MEP5mmnndbffqeddsq77ror77rrrtxpp5365y9evDgfffTRzMycN29ePu95z8ulS5cu165x+tOf/nTOmDEjMzNvvfXW3GqrrXLBggV5xhln5IQJE/KRRx7JBQsW5Lhx4/Kee+5ZrvY+W2+9dc6bNy8zM48//vg8+eSTMzMTyAsuuCAzM0866aT8yEc+kpmZRx99dP72t7/NzMy77747d9hhh/7X+oIXvCAXLlyY8+bNy0022SQXLVqUs2fPzkmTJuUTTzyRjz/+eE6cODGvu+665V7b5ZdfnocddlhmZj7yyCM5fvz4XLx4cZ5xxhn57Gc/O//xj3/kk08+mTvttFNec801ecstt+QhhxySixYtyszMd7zjHfntb3+75Wts9dlJkrSmAGZnxYyz1o0D186UKVOYM2cO55xzDi9/+ctXahuZyfve9z6uuOIK1llnHe69997+Hq92rrzySk444QQAdthhB7beemv+8pe/ALDvvvuy0UYbATBx4kTuvvtuttpqq7bb2meffRgxYgRTpkzhox/9KADrrrtu/7F8u+22GxdffDEAl1xyCbfcckv/uo899hiPP/44AAcffDCjR49m9OjRbLHFFjz44INceeWVvOpVr2LMmDEAHH744fz2t7/l0EMPXaaGl770pRx//PH8/e9/50c/+hFHHHEEI0cWX7P99tuPTTfdtH/9K6+8kpEjR3Lttdcyffp0ABYsWMAWW2zR8T2TJGltZ4BrcOihh3LiiSdy+eWX89BDD/XPHzlyJEuXLu2fbjce2dlnn828efO49tprGTVqFOPHjx907LIicLc2evTo/scjRoxgyZIlHbd12WWXsdlmmy0zb9SoUf1DcDRuY+nSpfzhD39g/fXXr/S8nepsdswxx3D22Wdz7rnn8q1vfat/fvNQIBFBZvKmN72JU089tfL2JUla23kMXIPjjjuOmTNnMnny5GXmjx8/nuuuuw6A6667jrvuuguADTfcsL/XCuDRRx9liy22YNSoUVx22WXcfffdLds12muvvTj77LMB+Mtf/sI999zD9ttv3/XX1mz//ffni1/8Yv/0n/70p47t99prL37yk5/w5JNP8s9//pMf//jHvOQlL2n52mbMmNF/jNxOO+3UP//iiy9m/vz5LFiwgJ/85Ce86EUvYt999+X888/n73//OwDz58/vf98kSVJrBrgGY8eO5T3vec9y84844gjmz5/P1KlT+cpXvsJ2220HwKabbsqLXvQiJk2axEknncTrX/96Zs+ezbRp0zj77LPZYYcdWrZr9M53vpOnn36ayZMn89rXvpYzzzxzmR6wXvnCF77A7NmzmTJlChMnTuT000/v2H7XXXdlxowZ7L777uyxxx685S1vYZdddmn52p71rGex4447cuyxxy6zjRe/+MUcc8wxTJ06lSOOOIJp06YxceJEPvrRj7L//vszZcoU9ttvP+6///6evW5JktYEsSK7xupm2rRpOXv27GXm3Xrrrey4447DVNHa4cknn2Ty5Mlcd911/cfwnXnmmcyePXuZXr8V5WcnSVqTRcS1mTmtSlt74NRVl1xyCTvssAMnnHBCf3iTJEnd5UkMNbPHHnvw1FNPLTPvrLPOWu64veHyspe9jHvuuWe5+TNmzGDGjBlDX5AkSWsgA1zNXHXVVcNdgiRJGmbuQpUkSaoZA5wkSVLNGOAkSZJqxgC3hokIjjnmmP7pJUuWsPnmm/dfTuvBBx/kkEMOYeedd2bixIn9lw2bM2cOkyZNarnNGTNmcP755y8zb4MNNujRK5AkSYPxJIYe2u2k73R1e9ee9sZB24wZM4abbrqJBQsWsP7663PxxRez5ZZb9i+fOXMm++23X/+AxTfccENXa5RWxT2zVv1s6nEzb+xCJZK0erMHbg100EEHceGFFwJwzjnncPTRR/cvu//++xk7dmz/9JQpU4a8PkmStGoMcGugo446inPPPZeFCxdyww03sMcee/QvO/7443nzm9/MPvvsw8c+9jHuu+++Sts86aSTmDp1av9NkiQNH3ehroGmTJnCnDlzOOecc/qPcetzwAEHcOedd/LLX/6SX/ziF+yyyy7cdNNNg27ztNNO48gjj+yf9hg4SZKGjz1wa6hDDz2UE088cZndp3022WQTXve613HWWWcxffp0rrjiimWWH3vssUydOnW58CdJklYP9sCtoY477jg22mgjJk+ezOWXX94//9JLL2XPPffkGc94Bo8//jh33HEH48aNW2bdM844Y4irlSRJK8IAt4YaO3Zs/5mmja699lre9a53MXLkSJYuXcpb3vIWpk+fzpw5c4a+SEmStFIiM4e7hp6ZNm1azp49e5l5t956KzvuuOMwVaRV4We35nMYEUlrs4i4NjOnVWnrMXCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwK1hIoJjjjmmf3rJkiVsvvnmHHLIIQA8+OCDHHLIIey8885MnDix/2oLS5cu5d3vfjeTJk1i8uTJTJ8+nbvuumtYXoMkSerMgXx7qBtjWjWqMr7VmDFjuOmmm1iwYAHrr78+F198MVtuuWX/8pkzZ7Lffvv1D/J7ww03AHDeeedx3333ccMNN7DOOuswd+5cxowZ09X6JUlSd9gDtwY66KCDuPDCCwE455xzlrke6v3338/YsWP7p6dMmdI//znPeQ7rrFN8JcaOHcvGG288hFVLkqSqDHBroKOOOopzzz2XhQsXcsMNN7DHHnv0Lzv++ON585vfzD777MPHPvYx7rvvPgBe85rX8LOf/YypU6fy3ve+lz/+8Y/DVb4kSRqEAW4NNGXKFObMmcM555zTf4xbnwMOOIA777yTf/u3f+PPf/4zu+yyC/PmzWPs2LHcdtttnHrqqayzzjrsu+++/PrXvx6mVyBJkjoxwK2hDj30UE488cRldp/22WSTTXjd617HWWedxfTp07niiisAGD16NAcddBCnnXYa73vf+/jJT34y1GVLkqQKDHBrqOOOO46ZM2cyefKyJ1JceumlPPnkkwA8/vjj3HHHHYwbN47rrruuf3fq0qVLueGGG9h6662HvG5JkjQ4A9waauzYsf1nmja69tprmTZtGlOmTOEFL3gBb3nLW5g+fTp///vfecUrXsGkSZOYMmUKI0eO5F3vetcwVC5JkgYTmTncNfTMtGnTcvbs2cvMu/XWW9lxxx2HqSKtCj+7NV83ht6pMtyOJK2OIuLazJxWpa09cJIkSTVjgJMkSaoZA5wkSVLNrJUBbk0+7m9N5WcmSdKAtS7Arbfeejz00EMGghrJTB566CHWW2+94S5FkqTVwlp3MfuxY8cyd+5c5s2bN9ylaAWst956y1zDVZKktdlaF+BGjRrFhAkThrsMSZKklbbW7UKVJEmqOwOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlm1rpx4KRuuGfW5FXexriZN3ahEknS2sgeOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzQx5gIuIAyPitoi4PSJOabF8r4i4LiKWRMSRTcuejog/lbcLhq5qSZKk1cfIoXyyiBgBfAnYD5gLXBMRF2TmLQ3N7gFmACe22MSCzJza80IlSZJWY0Ma4IDdgdsz806AiDgXOAzoD3CZOadctnSIa5MkSaqFod6FuiXwt4bpueW8qtaLiNkR8f9HxCu7W5okSVI9DHUPXLSYlyuw/rjMvC8itgEujYgbM/OOZZ4g4q3AWwHGjRu38pVKkiStpoa6B24usFXD9FjgvqorZ+Z95f2dwOXALi3afC0zp2XmtM0333zVqpUkSVoNDXWAuwbYNiImRMS6wFFApbNJI2LjiBhdPt4MeBENx85JkiStLYY0wGXmEuBdwK+AW4HvZ+bNETErIg4FiIjpETEXeDXw1Yi4uVx9R2B2RFwPXAaWOiflAAAgAElEQVR8ounsVUmSpLXCUB8DR2ZeBFzUNG9mw+NrKHatNq/3e2ByzwuUJElazXklBkmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNXMyOEuQFL97XbSd7qynR9v2JXNrDG68b5ee9obu1DJ4OpUa534vqode+AkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNjKzSKCJGAiMy86mGefsDE4ErMvO6HtUnSZKkJpUCHHAe8ChwHEBEvBv4HPAUMCIiDs/Mn/emREmSJDWqugt1T+CihumTgM9k5vrAN4D3d7swSZIktVY1wG0KPAAQEZOB5wKnl8t+QLErVZIkSUOgaoB7EBhfPj4QuDsz7yin1weWdrkuSZIktVH1GLgfAJ+MiJ2BY4EvNizbBfhrtwuTJElSa1UD3CnAY8B04CvAxxuW7UZxkoMkSZKGQKUAl5lLgFltlh3e1YokSZLUkQP5SpIk1UzbHriImAdk1Q1l5hZdqUiSJEkdddqF+iVWIMBJkiRpaLQNcJn5oSGsQ5IkSRV5DJwkSVLNVB1GhIh4AfBmYDtgveblmbl7F+uSJElSG5V64CJiP+AKYCzwYmAe8ASwM8Vltm7qVYGSJElaVtVdqLOAzwMHl9MfyMx/peiNWwxc3v3SJEmS1ErVADcR+AXFNU8TGAOQmXcDHwLe34viJEmStLyqAW4hsE5mJnA/8LyGZY9R7FqVJEnSEKh6EsP1wPbAxcCvgf+OiHuBRRS7V2/sTXmSJElqVrUH7nMMDOr7PuCfwK+Ay4AtgOO7X5okSZJaqXox+4saHt8bEbsBzwfWB/6cmYt6VJ8kSZKaVB4HrlF5LNxfu1yLJEmSKqgU4CLiU4O1ycyTV70cSZIkDaZqD9yrW8zbGPgX4FHgYcAAJ0mSNASqHgM3odX8iNgD+Brw9m4WJUmSpPZW6WL2mXkVcBrwxe6UI0mSpMGsUoArPUQxRpwkSZKGQNWTGJ7RYva6wI4UA/ne3M2iJEmS1F7VkxieYGAg30YB3Au8smsVSZIkqaOqAe44lg9wC4G5wNWZubirVUmSJKmtqmehntnjOiRJklRRN05ikCRJ0hBq2wMXEUtpfdxbS5k5oisVSZIkqaNOu1DfzUCAGwW8l+Jkhp8CfweeBRwGjAE+08MaJUmS1KBtgMvM/sF5I+KzwFXAq8sL2ffNPwX4AdDySg2SJEnqvqrHwL0R+HpjeAMop78OvKHbhUmSJKm1qgFuBMWgva3stALbkSRJ0iqqOg7c2cDHI2IkcAHFMXBbUBwDNwv4Zm/KkyRJUrOqAe4/gcUUYe2TDfOfAr4KnNzluiRJktRG1YF8FwH/EREfASYDzwYeAG7MzPk9rE+SJElNqvbAAVCGtd/0qBZJkiRV0Gkg35cDV2bmY+XjjjLzoq5WJkmSpJY69cD9HNgTuLp8nEC0aZsUZ6pKkiSpxzoFuAnA/Q2PJUmStBrodCWGu1s9liRJ0vCqNABvRMyJiE9GxC69LkiSJEmdVb2CwvnAa4HZEfGXiJgVEZN6WJckSZLaqBTgMvPEzBwPvBi4CDgOuD4iboqID0TEtj2sUZIkSQ1W6BqmmfmHzPx3YCtgb+By4ATg1q5XJkmSpJZW9iL0Y4BxwNbARhSX1JIkSdIQqBzgImL9iHhNRPyQ4mL236QY/+044Fk9qk+SJElNKl1KKyLOAw4GRgOXAscDP87MR3pYmyRJklqoei3UZwEnAudn5j96WI8kSZIGUSnAZebePa5DkiRJFXW6mP1E4I7MfKp83FFm3tLVyiRJktRSpx64mxi4mP1NFCcstBJ4MXtJkqQh0ynA7QPc0vBYkiRJq4FOF7P/TavHkiRJGl4rO5CvJEmShkmnkxiW0v64t+VkpsfASZIkDYFOx8C9m4EANwp4L/AE8FOKKzE8CziM4rJan+lhjZIkSWrQ6Ri4L/Y9jojPAlcBr87MbJh/CvADYEIvi5QkSdKAqsfAvRH4emN4Ayinvw68oduFSZIkqbWqAW4EsGObZTutwHYkSZK0iqpeC/Vs4OMRMRK4gOIYuC0ojoGbBXyzN+VJkiSpWdUA95/AYoqw9smG+U8BXwVO7nJdkiRJaqPqxewXAf8RER8BJgPPBh4AbszM+T2sT5IkSU2q9sABUIY1r8ogSZI0jCoHuIhYD9gLGAus17Q4M/Mr3SxMkiRJrVUKcBHxYuBHwGZtmiRggJMkSRoCVXvgvgDcAewH3JKZi3tXktZW98yavMrbGDfzxi5UInXWje8q+H2VtPKqBrjtgcMz8/peFiNJkqTBVR2A9waKM08lSZI0zKoGuHdQDCPy0l4WI0mSpMFV3YV6MfAM4NKIWAw81twgM7foZmGSJElqrWqA+xLFmaaSJEkaZlWvxPChHtchSZKkiqoeAydJkqTVRNseuIj4/opsKDNfs+rlSJIkaTCddqFuPmRVSJIkqbK2AS4z9xnKQiRJklTNkB8DFxEHRsRtEXF7RJzSYvleEXFdRCyJiCOblr0pIv5a3t40dFVLkiStPoY0wEXECIohSQ4CJgJHR8TEpmb3ADOA7zWtuwnwQWAPYHfggxGxca9rliRJWt0MdQ/c7sDtmXlnZi4CzgUOa2yQmXMy8wZgadO6BwAXZ+b8zHyYYnDhA4eiaEmSpNXJUAe4LYG/NUzPLef1el1JkqQ1RtsAVx6LtkGXny9azKt6hYdK60bEWyNidkTMnjdv3goVJ0mSVAedeuAuozhOjYi4MyJ27sLzzQW2apgeC9zXzXUz82uZOS0zp22+uSOhSJKkNU+nAPc40HeSwHhg3S483zXAthExISLWBY4CLqi47q+A/SNi4/Lkhf3LeZIkSWuVTgP5/h74RkRcVU6fGhHz27TNzHztYE+WmUsi4l0UwWsE8K3MvDkiZgGzM/OCiJgO/JgiPL4iIj6cmTtl5vyI+AhFCASYlZnt6pEkSVpjdQpwxwHvB3agONZsY4rQtUoy8yLgoqZ5MxseX0Oxe7TVut8CvrWqNUiSJNVZpysxPACcABARS4F3ZObVQ1WYJEmSWuvUA9cvM4f8ig2SJElqrVKAA4iIZwJvA14MbALMB34LfC0zH+lNeZIkSWpWqWctIp4H3AjMAsZQXO5qTDl9Q7lckiRJQ6BqD9z/Ao8Ae2bmvX0zI2JL4BfAZ2m6JJYkSZJ6o+qxbXsDMxvDG0A5/WFgny7XJUmSpDaqBrik/RAi61D9cliSJElaRVUD3GXARyJi68aZ5fQs4NfdLkySJEmtVT0G7t+BS4G/RsR1wIPAFsBuwN+A/+xNeZIkSWpWqQcuM+dQXJHh3cDNwCjgFuBdwI7lckmSJA2ByuPAZeYi4PTyJkmSpGHiFRYkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYqB7iI2CIiPhkRv46Iv0TETuX890TEC3pXoiRJkhpVGkYkInYHLgbmAb+huDbq6HLxc4D3Akf2oD6tontmTV7lbYybeWMXKtFw8TsgSWueqj1w/0txOa3tgLcB0bDsamD3LtclSZKkNqoO5LsrcFhmLo2IaFr2EMVltSRJkjQEqga4R4HN2yzbhuLaqFIt7HbSd1Z5Gz/esAuFDKIbdcLQ1CrVSTd+W9ee9sYuVCKtvKq7UH8KfDgitmmYlxGxGXAi8KOuVyZJkqSWqga4U4DHKC5gf0U573TgNmABMLP7pUmSJKmVSrtQM/PhiNgTOAbYF/gnMB/4BvCdzHyqdyVKkiSpUdVj4MjMRcA3y5skSZKGiVdikCRJqplKAS4ilkbE021uSyJifkRcFhGv6nXBkiRJa7uqu1D/s7w9BvyM4ooMWwCvADak2K36EuD8iHhTZn63B7VKkiSJ6gHuucDvMvPopvmnRMS5wMaZ+bKI+A5wMmCAkyRJ6pGqx8AdC5zRZtkZFGenApwHbLuqRUmSJKm9qgFuJLBDm2U7NmxnEbBwVYuSJElSe1V3oZ4LnBoRIxk4Bm5z4DBgFgO9c7sCf+52kZIkSRpQNcC9h6J37aPAaQ3znwK+DpxUTl8F/Lpr1UmSJGk5Va/EsAh4T0R8GJgMPBt4ALgxM+c3tLu8F0VKkiRpQOUrMQCUYe03PapFkiRJFVQOcBERwIuA7YD1mpdn5pe7WJckSZLaqBTgIuJZFMe2TQQSiHJRNjQzwEmSJA2BqsOIfAZ4FNiKIrztAYwHPgD8laJXTpIkSUOg6i7Ul1KciXp/OR2ZeQ/w8YhYh6L37YAe1CdJkqQmVXvgngnMy8ylFNdD3aJh2e+BF3a7MEmSJLVWNcDdBTynfHwz8PqGZa8A5i+3hiRJknqi6i7UC4H9ge9TDOb704iYCywGxgH/1ZvyJEmS1KzqQL7/3fD4FxHxQuBVwPrAxZn5ix7VJ0mSpCZVhxEZB9yfmYsBMnM2MLtcNioixpUnNUiSJKnHVuQYuF3aLJtSLpckSdIQqBrgosOy9Sguai9JkqQh0HYXakRMAaY2zHp5ROzQ1Gw94DXAX3pQmyRJklrodAzcq4APlo8TmNmm3V3A27pZlCRJktrrtAv148CGwL9Q7EL913K68TY6M5+XmZf0ulBJkiQV2vbAlWecLi4nqx4rJ0mSpB6rOpAvABGxHTCW4ti3ZWTmRd0qSpIkSe1VHQduInAeMJHWZ6QmMKKLdUmSJKmNqj1wXwXWBQ4HbgEW9awiSZIkdVQ1wO0CHJWZP+9lMZIkSRpc1QB3By2Oe5MkqVvumTV5lbcxbuaNXahEWv1VPbv0vcD7ImKbXhYjSZKkwVXtgTsV2BL4c0TMAR5pbpCZu3exLkmSJLVRNcDdVN4kSZI0zCoFuMw8tteFSJIkqZoVusJCFLaKiBdGxJheFSVJkqT2Kge4iHgncC9wN/BbYPty/o8i4t97U54kSZKaVQpwEXES8Fng6xQXtW+8GsPlwGu7XpkkSZJaqnoSw/HAzMz8VEQ0XzLrNmC77pYlSZKkdqruQn02cG2bZUtxkF9JkqQhUzXA3Q68tM2yvSiujypJkqQhUHUX6ueAL0fEIuD8ct4WEfFm4D+Bf+tFcZIkSVpe1XHgvhERGwMzgQ+Xsy8CngQ+lJnf61F9kiRJalK1B47MPC0iTgdeAGwGzAf+kJmP9qo4SZIkLa9ygAPIzMeB/+tRLZIkSaqg6jhwH4uIr7ZZdnpEfKS7ZUmSJKmdqmehHk1x9YVWfgu8rjvlSJIkaTBVA9xzKS6j1cp95XJJkiQNgaoB7gFg1zbLdgXmdaccSZIkDaZqgPs+MDMiDm6cGREvBz4AnNvtwiRJktRa1bNQZwJTgZ9FxEPA/cBzgE0ozkr9QG/KkyRJUrOqA/kuBPaPiAOAfYBNgYeAX2fmxT2sT5IkSU0GDXARMRo4Efh5Zv4K+FXPq5IkSVJbgx4Dl5lPAe8Hntn7ciRJkjSYqicxXAXs1stCJEmSVE3VkxhOBr4XEYsoLmL/IJCNDTLzyS7XJkmSpBaqBriryvsvAJ9v02bEqpcjSZKkwVQNcMfR1OMmSZKk4VF1GJEze1yHJEmSKqraAwdAREykOJlhK+BbmflARDwfeDAzH+9FgZIkSVpWpQAXERsA3wKOBBaX6/2S4hqpHwfuoRgrTpIkST1WdRiRzwIvBPYFNgSiYdlFwIFdrkuSJEltVN2Fejjwnsy8LCKazza9G9i6u2VJkiSpnao9cOtTXPu0lQ2Bp7tTjiRJkgZTNcBdA7yxzbIjgd93pxxJkiQNpuou1P8BLomIS4AfUIwJ9/KI+A+KALdXj+qTJElSk0o9cJl5JcUJDKOBL1KcxPBhYBvgZZl5Tc8qlCRJ0jIqjwOXmb8DXhIR6wMbA4+srdc/vWfW5FXexriZN3ahEkmStDbqGODKsPZyYDzFmG+XZOaDwILelyZJkqRW2ga4iNgGuIQivPV5LCJek5n/1+vCJEmS1FqnY+A+BSwFXgI8A9gJ+CPw1SGoS5IkSW10CnAvAP4nM3+XmQsz81bgbcC4iHjO0JQnSZKkZp0C3HOAO5vm3UFxBuqze1aRJEmSOhpsGJEckiokSZJU2WDDiPwqIpa0mP/r5vmZuUX3ypIkSVI7nQLch4esCkmSJFXWNsBlpgFOkiRpNVT1YvaSJElaTRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMjh7sASVLv3DNrcle2M27mjV3ZjqTusAdOkiSpZgxwkiRJNTPkAS4iDoyI2yLi9og4pcXy0RFxXrn8qogYX84fHxELIuJP5e30oa5dkiRpdTCkx8BFxAjgS8B+wFzgmoi4IDNvaWj2ZuDhzHx+RBwFfBJ4bbnsjsycOpQ1S5IkrW6Gugdud+D2zLwzMxcB5wKHNbU5DPh2+fh8YN+IiCGsUZIkabU21AFuS+BvDdNzy3kt22TmEuBRYNNy2YSI+GNE/CYiXtLrYiVJklZHQz2MSKuetKzY5n5gXGY+FBG7AT+JiJ0y87FlVo54K/BWgHHjxnWhZEmSpNXLUPfAzQW2apgeC9zXrk1EjAQ2AuZn5lOZ+RBAZl4L3AFs1/wEmfm1zJyWmdM233zzHrwESZKk4TXUAe4aYNuImBAR6wJHARc0tbkAeFP5+Ejg0szMiNi8PAmCiNgG2Ba4c4jqliRJWm0M6S7UzFwSEe8CfgWMAL6VmTdHxCxgdmZeAHwTOCsibgfmU4Q8gL2AWRGxBHgaeHtmzh/K+iVJklYHQ34prcy8CLioad7MhscLgVe3WO+HwA97XqAkSdJqzisxSJIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmhnygXyH224nfWeVt/HjDbtQSAXWKnWf31VJawJ74CRJkmrGACdJklQzBjhJkqSaMcBJkiTVjAFOkiSpZgxwkiRJNWOAkyRJqhkDnCRJUs0Y4CRJkmrGACdJklQzBjhJkqSaMcBJkiTVjAFOkiSpZgxwkiRJNWOAkyRJqhkDnCRJUs2MHO4CJEmqm3tmTV7lbYybeWMXKtHayh44SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmjHASZIk1YwBTpIkqWYMcJIkSTVjgJMkSaoZA5wkSVLNGOAkSZJqxgAnSZJUMwY4SZKkmhk53AVIkqTeuWfW5FXexriZN3ahEnWTPXCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmDHCSJEk1Y4CTJEmqGQOcJElSzRjgJEmSasYAJ0mSVDMGOEmSpJoxwEmSJNWMAU6SJKlmhjzARcSBEXFbRNweEae0WD46Is4rl18VEeMblv13Of+2iDhgKOuWJElaXQxpgIuIEcCXgIOAicDRETGxqdmbgYcz8/nA/wKfLNedCBwF7AQcCHy53J4kSdJaZah74HYHbs/MOzNzEXAucFhTm8OAb5ePzwf2jYgo55+bmU9l5l3A7eX2JEmS1ipDHeC2BP7WMD23nNeyTWYuAR4FNq24riRJ0hovMnPonizi1cABmfmWcvoYYPfMPKGhzc1lm7nl9B0UPW2zgD9k5nfL+d8ELsrMHzY9x1uBt5aT2wO39eClbAb8owfb7QVr7Q1r7Y261FqXOsFae8Vae2Ntr3XrzNy8SsORXX7iwcwFtmqYHgvc16bN3IgYCWwEzK+4Lpn5NeBrXax5ORExOzOn9fI5usVae8Nae6MutdalTrDWXrHW3rDW6oZ6F+o1wLYRMSEi1qU4KeGCpjYXAG8qHx8JXJpFN+EFwFHlWaoTgG2Bq4eobkmSpNXGkPbAZeaSiHgX8CtgBPCtzLw5ImYBszPzAuCbwFkRcTtFz9tR5bo3R8T3gVuAJcDxmfn0UNYvSZK0OhjqXahk5kXARU3zZjY8Xgi8us26HwM+1tMCq+npLtous9besNbeqEutdakTrLVXrLU3rLWiIT2JQZIkSavOS2lJkiTVjAGujQqX/JoaEX+IiJsj4oaIeO0w1bleRFwd/6+9Mw+3qyrv/+cLIYSAEGoAaQNGKFEGNUEJQyXkJxpkEBArqFgMaFEGFZRSJiVYsTIJFqsICImIiihoHyKTQAjQJhQ0IwkBMUCAkMQQhoQhkPf3x/se7sq+e59z7r3n3Jtb1/d59nPOXnsN3/Xud71r3kuaGVzOqfD3hqQZcRU3jvQaJA2R9CtJ8yXNk7RniZ+TIy9zJP1c0qBe5PeVSHeupJMq/FwlaYmkOQX3T0S4NZJavjOpLF1JF4QsZ0m6UdKQirB1uUnaVtJLkk5pI9cJkp5K9PCArnCVdGQSdkY8H9kmriMlTYt0HpBU+tFwSdeGnZgT8WwQ7pL0H2E/Zknatac8I95tJN0VZWeupK+Ee1O6V6UvkkYncp0p6WPt4po8P0WSSRpaEb5UtvFsbHCdK+nudvCUH+lYk8lCSTMqwlfJdKCkqyXNDpmO7QnPJL1Smy/fHDhd0iPBfWBF+FuSsJcpTjRq1pa0gOePw22WvC7YpIs8m7IjbeIuSedKWhD68uVWp90lmFm+Che+weJPwHbAQGAmsFPBzwhgh/j/t8AzwJA+4Cpgk/i/ATAd2KPE30t9LdfgMQn4fPwfWJQZ/nHmPwMbxf0vgfG9xG0XYA4wGF8f+vvaOy74GwPsCswpuO+If3twCvD+NvDrlC4wDhgQ/88DzqsIW5cb8GvgeuCUNnKd0Ez8zcgReDfwWBu53gbsH/8PAKZUhD0gyqCAnwPHJe43h/sewPQWcd0a2DX+vwVYgB9L2JTuVelLTeeTNJbU7lvNNe63wTezPQ4M7aJsh+Cb2baN+y3bxTPxcxHwjS7K9ATg6hpH4EFgvRboQKnNx23lJ8P9spq8SsJvmsTz6yRMU7akBTw3Tfx8Fzitizwn0CI71Q3uRwM/qb3HMt0DJgJj28mvduURuHI0PPLLzBaY2SPx/2nc4DX18b1Wwhwvxe0Gca2TCxslbYpXlj8GMLPXzGxFidcBwEby7wAOpuR7f23CjsA0M1tlfgrI3UCnkQgzm4rvkC66zzOzdnw4ujJdM7stuAJMw7+PWBa2kpukQ4HHgLnt5NqFsM3I8VN4pd5jVHA1YNP4vxkVOmhmv4syaPhnjWryPwT4STyaBgyRtHULuD5jZn+I/y8C84C/a1b3qvQl0XmAQbTAhlRxjccXA6fWS6eObD8N3GBmT4S/JW3kiSQBh1Ohb3XK4E7AHQnHFUCPR+br2PwP4sdPgneUD60I/0L8HYB3oq1BPlrKs5Z+yHUjKnSgimdvoI6MjwO+aWZrwl+PdK+nyA24cnTp2C759MpAfNSu1yFp/RjeXwLcbmbTS7wNkk8FTYsKuy+wHbAUuFrSHyVdKWnj1IOZPQVcCDyBj2o+b2a39RK/OcAYSW+VNBgfAdimQZh1Ccfgoz5NI+T/r0Dp1HsbcGJMnVwlafMexHMELWrAVeAk4AJJT+L6eHo9zzG990/ALeHU9qP/JA0HRuGjA93BWvoiaXf5STizgS8mlXmPkXKVdDDwlJnNbDJsUbYjgM0lTZH0oKSj2sEzcd4beLbWYW+AVKYzgUMkDZB/u/R9tMieFG0+XvesSN5Zozrr1gj7Ih2Nvqp8tIxnrW6SdDWwGHgXcGk3eLbKjnSV+/bAEVGX3ixph3ak3SxyA64cKnErbf1Hr/oa4Ohaq7y3YWZvmNlIvMc0WtIuJd62Nf9i9KeBSyRt36skHQPwqaofmtkoYCWw1vrCKIyHAO/Ap6Y3lvSZ3iBnZvPwqYPb8cpiJv7NwXUekhLkAkcAABVbSURBVM7EuV7bxaDnABcnvc124oe4ARyJN84v6k4kknYHVpnZnIaeu4/jgJPNbBvgZGLUuA5+AEw1s3tqNEv8tGwEIdYN/Ro4KRmp6Er4TvpiZtPNbGdgN+B0tWjtaco10jwT+EbdQGujKNsBeGPoQGA/4OuSRrSSZ0GmTY32lsj0Krwh9QBwCfDftMieFG0+PnvQyVud8PvhU8cb4iN3b6IHtqQhz1rdZGZH4/Z9Ht4Z6wrPltiRbnLfEHgl6tIr8HeMpP1qa/KAg4Er4767naumkBtw5Sg7tmtZsmjyYHhzSnAycFZMk/QpYjpyCnBgkWtM82Jmj4WfUX1AcRGwKBkh/BWwV8L1i8CHgD+b2VIzWw3cAOzVWwTN7MdmtquZjcGn1VYU+K1zkPRZ4CDgyJhuQr54eoak39UPze7A+ZIW4hXsGfKPbbccZvZsGMU1uPEb3UWuNXyS9o6+gZ8Gc0P8v54OrrcG1ytrHiWdjS+f+GoSvqmj/7qDGJH6NXCtmd3QwG8n2ZbpS4royKzE14S2muv2eOdsZujcMOAPkt7WRdneYmYrzWwZMBV4b4t51twHAIcB1yVuTcnUzF43s5PNbKSZHYKv3WtmFK9pJDZ/D3yavvZt12HA07VRpLi+WQj7Cn7C0ZvLgxrpRgt4fiRxewOX68e7wrPKjrQLBe6LcD0BuBF4T/i5Nd7zyOD6+bjfvZ3c2r7Irj9eeA/vMdzQ1DYx7FzwMxBf33BSH3PdgtgIgK8nuAc4qOBnc2DD+D8UNyI79SbPhMs9wDvj/wTggsLz3fG1WIPxUYxJwJd6kd+W8bstMB/YvMLfcAqbGJJnU2jDJoaydHGj8hCwRZPhK7nR4sXBJVy3Tv6fDPyiq1zxTuciYLs2y3UesRAZ2Bd4sCLc5/GRlY0K7gey9iaG+1vEU/gi6ku6o3tV+hK2rraA/e14Y7N0c0GruIafhVXp1JHtjrjtra2RnQPs0g6eIa+7G4SvkulgYOP4/2F8FLEVOlBq8/GORrqJ4fiSsJvUymHI7zrgxHr5aDHPjwJ/n8j9QuDCLvLskh1psYy/AxwT7mOB/y0JO5Fe2sTQ9gT664Wvf1qAry04s+T5Z4DVwIzkGtkHPN8D/BGYFYas004pfARrNt4QnQ18rg/lOhKfUpgF/IaSBhI+rTc/8nMN0fjsJX73hBGbCexb4efn+ND9arwx8blw/1jcvwo8C9zaYm6d0gUexdda1XTwsoqwDbnRwgZcBddrQv9m4b3UrbvKNYzmtF6Q6wfwXYMz8fVQ76sI+3rYiJr8vxHuAv4zns2mRQ364GUhw1qaBzSre1X6gq8xmxtufwAObRfXgp+FVDfgSmUbz/4lyukcetiJrscTr4y/2CB8lUyHAw/jnYHfA29vkQ6U2nx8jfH9wed6SuwmsBV+JvmseN+X0tFwb8qW9IQn3gG7L8rEHHyadtMu8mzKjrRJxkPwWbfZwP8A7y0JO5FeasDlkxgyMjIyMjIyMvoZ8hq4jIyMjIyMjIx+htyAy8jIyMjIyMjoZ8gNuIyMjIyMjIyMfobcgMvIyMjIyMjI6GfIDbiMjIyMjIyMjH6G3IDrA0iaIMnimJDis19JmtKLXMYGlx5/tLMdkLSjpHskrQyew/uQy6mSxnYz7BRJZUfW9DlCricm98eq5Lg1SQslXdiN+CdIWtZTnv+XIGmTkPv4vuayLkDS8JDHQb2UXsvsnqTD83vsPRTt1V8zcgOubzFO0m59TWIdxwX4t3cOBvbEv9fVVzgV/w5Zd3A8Dc7T7EPsiX83qoZjqTgIu5u4Ej/yKCOjCs/genhvL6X3h0ivFedXHw6Mb0E8GRldwoDGXjLahOX4hzfPpLWV5ToFSYPMj0LpLt4F/JeZ3dEqTilawK8pmNlD7U6ju7A2HwNnZotwXW87eut9/rWgF8vHq0CvHUdoftZpnx9/2BfIZeT/DvIIXN/BgG8DB0t6d5WnqumnkmmvhZIulHSapGckPS/pIjkOkDRX0ouSfiM/ML6Iv5V0U0xVPlF27qekD0i6W9IqSX+RdIWktyTPxwev0TFl+DL+xfSqvI2UdEfE95ykayVtFc+GSzL87MSTI94pdeIySV+V9D1JyyWtkHSppIHN8JM0VNKkyNeqeP7+VL7AW4GzIw6rTadKWi/k/qikVyUtiDMFU35rTaHW3qukUZKmRZp/lLR3IdzBkh6M9/KcpOmS9qkjh8clnZHcfyG4fjlx+5qkpwqyO7HGEz8o/LNJPscX0jhZ0qLg8wtJQ6r4pHlN7mvTV2MlXS/pJUmPSTq+JOwYSXeFn+dDjqPiWb33OUjS+ZKejHcyU9IBhbiPknRv6Mtzkc77C352lnRL+FkpaZ6kEwp+DpH0gKRXJC2OdDco+Pl46MXLkqbiHZOGaJQPSXtIel3SMYnbZuH/p4lbT8tHM/Ksq6uSPie3Qy+H7t8taed41mkKVX4+5gS5PXo1wn66kObEkP2HJc2KtO+txVtHrp2mUOP+K5K+LWmppCWS/lPShnXimQh8HNgnKS8T4lmnJQeJjDcp8NhX0m+D/yOSxkX+LwhZPSXpqyXpHy5pdsjnSUnnquM81C7ZZElDJF0p6enQ5SckXZE8f5e8vD8pt1dzJZ0kab3ET7fzk7zLQyXNDw73StqpSv5J2LplUNIwSb+Md/qypD9J+rdG8a7z6I3jHvLV6aiNCcAyvAE9n+QsN/yA9ylFvyVxGHE2XNwvxEc5bsDPtDsz/FyMHwl0GHAk8BzJESn4lKDhR6h8G5/q+lG4HZT4+wf8mJ7r8KN7/gl4CvhV4md8hPsTcArw/4BRFTLYAliBH0dyKH402SL86JKBwIb4GZLP4Met7EGd81sj3afwg4b3j/RfJTlrtR4/fOpmMXA0fl7fVOBFOs7tGxV8rwwuexBHwOBHJr2ET7F+CDgPeKMgvykFWU0AVkV+jw7O00IvBoef7YHX8GnkD4bcvw58rI4cfgrcnNxfC7wMXJ+4/Za1de5NXQJ2wo/+mZzkc4tEx54Abgoux0a+f9CMvpfo3CPAWfg5kVeF2+iCv9XAbXgl+RHg32pybfA+bwKWAMcB4+K9vU5y3B1+tM+x+Fmn++NH9KwiOWc14p4c+d0Xnwo/LXl+eLzrH0Q6x4WeXJj42TXSvp4O3XwsuI9vILtm8vHvwPPAtnE/CS8Lmyd+elo+6vKgga4CY+Jdnh7v9eDgvVc8H05nm3NuhDkLt0uXh59PJX4mBq8ZwBER7wL8+CXVkevYiGuXgoyeiDj3wxs6rwOn1olne+BOfEq2Vl6GJeXlwoL/mow3KfB4NNIbhx+79QLwQ/xM03HA98LfHklc48JtEl42To13elkz77QkL1fh9dERwD64Tb48eb4vfszhR4P3SbjenV4i1+7kZyKwFC8bR+J11my8bhpUp+5rpgzeSUddMxY4Bji/XtnrD1efE/hrvEgqtChgbwAj4r4nDbhHgfUTt/txA/SOxO184NnkvlbgLi/EfzvJmZP4GaF3Ffx8kMQIJsbiK03I4DtRyDZN3EbT2UAvpOSw4wp5zAfWS9zOxCvkv6nHDzd+BuyTuG0cxuRHidsyYEIh7N8Da4DPFtx/QnLQMeUNOAM+mLiNDLePxP0/An/pom59ATeq68X9E8D3gcVxr8jHCXV06QFgYkncC/GKYEDidkkt7mb0vaBz30zcNgh5fydx+5/gUloR13mf+xbfZ7hPJWnIFp6thy8pmU/HuYdDI553V4QR8DhwdcH9GLzR/Na4/yV+bqcSP7UO1vg6cmsqH3iHZxZeUR4SYYpnjvakfDTk0UhX8cbDg3WeDydpwAF/A6wEzi74+x3wcHI/EbdxOyRuh0Zc76qTXk0Hiw24qQV/v6HB2bsUbHahvDTbgDs78bNTuN1Z0M/FwHmJ2zQ62+RT8fpkWCG9ZmzyHOBLjfwluj8AOAN4rESu3cnPxPC3V+L29ni/X0zc3rRXNF8GXwI+2kze+tOVp1D7Hj/FK9lWLHCfYmZvJPePAgvN7M8Fty3SqZPAjYX7G4D3xbD3YHzB7y8lDahd+KjVanzKLcXkJriOBm4zX4sCgJndjxu9DzQRvgy/NbM1hTxsBBR3mhX5jQaWmtndCZeV+KhDIy774g24GwuyuQMYKWn9OmFX4w27Gmrr5IbF72xgM/nU7jhJGzfgAt7Q3hR4r3zH7jC80T5U0g7AzvhU8D1NxFWGu8zs9QLnLUv0qRncVvtjZqvxEblhAJHX3YFJFha4Dorv80N45XBfyTtJp8V3lHSjpGfxSm818E5gRHhZjvf+L5N0hKQtC+mMALalc7m4ExhEh96Nxtdxpvm4oUGems6Hmb0GHIWPcl0HXGlmvyuJr7vloxkejXR1BjBK0sXyafFG+rILMJi1N9cQ+RtReBcLzeyR5L5YjrqC2wr3D3Uznq4iXeP7aPzeWXOI9/YY8Hfg08v4yG6ZfNbD7XWKZmzyDOBfJB0vaUTxoXwa/RxJj+IjfavxUdJ3pNO23clPgiVm9t+Jv8fxGaTRFZybLYMzgH+PKeVtK+Lqd8gNuD5GVIbnA5+R9PYeRreicP9ahZvwXnuKJSX3A/BRiM2B9fEh6tXJ9So+crJNIeyzTXDdusLfs3jvuzsoy0MtrWIareIyFJfN86wtm4m4/Ippp3ghrVCjIgY3PJjZw/iIynb4yMMyST+TtEVVhOabJZYBe8c1x8yewA1YzW0F3tvuDprVp+7GNSj+bx7xNrPruPjuhgJvY+33sRofCdwGQL5287a4/youl92AmXTIfw0+JbMYn15aLP+kzagkHfB3k6ZT6zDVysXbqNbNemiYjwQz8cbGhng5LUN3y0dDHo101cx+jy8VGIN3WpZJ+kGdTkmNU5FL7T5dx1umR9ChS11BPZ1sJ95MN7ED9bgMxW1vlXyKdqsZm3wiPuL4DeDhWLv2yeT5efhI6uX4FPluwLfiWVFGXc1PDWXlYgnVdrTZMngEPpp/MfC4pBmS9q2Is98g70JdN3AVvs7jX0uevUKhclT5JoSeoji6sCU+dL0ML2SGG+yynv3ThftGIybgFXMxTYCt8B5Xd1CWh1paKYr86nFZ3iDN5bic/gEfiSuimYq6EmY2GZgsaTPgQHzK8lLgk3WC3UtHQ21quN0TboOA+wojMesinsPlWa8BXEPxfS7H13vV2929Jz6y8mEzm19zDDl3ROzPPh4LovfGK7HJkobRoRvHAn8sSaNWiSymWjfroZl81HASvjFiPnCppDEl77i75aMpHo101cwmAZOiUXcYXpm+AJxWEl2N05bAXxL3rRJO6zo62W663zktYhneUCm+0yr5NLTJZrYC+DLwZUnvwadjr5U0KzqGnwAuNbPza2EkHdhN/lUoKxdb4msay9BUGTSzp4DxseFiNF6X/Zekbc3sLyXh+gXyCNw6APMt9Bfi8/bFCmsR8BZJ6VDzuDbQ+FjJ/YNm9kZMJ04D3mlmD5RcxQZcM5gO7Ke1d7Huhq+F6e63oA5Jd0ThlcTLNB5tmo5PA45JuAzGK6GUS1mP8U58BG6zCtm8RgtgZs+b2c/wqe5Gu7JqjbUxdDTgptIxAtdo+rS3Rh0qETo3HThKkroY/A58xOilsncSfjaK31drgSTthetfGZ/VZnYn8F28jA4BHsYbNsMr3n2tYvhffLd5mo/DWpQPJL0Tn8o6C69kdwNOLomvu+WjKR41NNJVM1tqZj/C9bBKl+fg6/M+UXA/HFhgZksbcO5NVJWXRcCOBbcPtyLBWCrzIOXyWYOvH+1J/LPwTQjr0bFjeiPWLi/rU78j2R1sGeWwlsa2+FTx/RX+my2DgI+qm3826Rx8ir6ns159ijwCt+7gR/iC0L2AuxP3W3Aje5Wki4B3AJ0+8dEC7C/p3Ej7MNzQHJI8PxW4Q9IafNHui/jagwOBM81sQRfT+y6+W+hWSecBm+AbG2bjO+W6g7cA18u3vu+MTwV838zq9tbN7FZJ9wHXSToN7/GfghusCxKv84EDJd2CL4p92MwelnQZ8AtJ5+PD9IMi/RFm9vlu5gVJX8BHim7BRzl3wA32TxoEnQpchPfGaw24e/Edc9C4ATcfb1zvh8viz33USz0NX5h/s6TL8UXtewIPmNlNdcLdDtwK3B66NRdfFzgS3812Ot4heQm4It7bMLxXnn5e5T14x+o6fL3O5vgo+cyaTkn6GnCNpE2Bm/HKfDt8tOofzWwVPmo3HV+n82N8Xc7nmsh/w3xEJToJH334rpmtkXQ28C1Jk9PRRbpZPprkUVdXJZ2Djz5NwUePRuE7HctG3zCz5ZIuAc6S9Dperg7Dp+4+1Vh0vYr5eOP4ULzR9nR0am/ER0PPwBvxh+FybxXOxu3n1cAvgHfju7SvMP/2Ypcg6V6c8xx8xO6f8TJXazzdDpwQa+CWAyfgU/atxDK8PH0dr/e+ic9iTCzzHPpetwziU8234rq4IDh/DR8Zn9di/r2Lejsc8tWei+qdpWfgBWdKwX1/3GiuwivfHSnfhVrc8TQRr+xSt/GU74LaD1f+VbgROr6E3+64gX4BL9gP4Q2xzcribkIOo/ARrFX4dN/PgK0KfjrlqyIuw9cyfR+ffnse/7zHhlV5L4TfAi/gz+GG425gt4Kf9+EV/8qIZ2y4C5/Cmov3UJdG+KOSsFPovAu17u5ivEKcjFeIr+DTAeeleaqQxfp4A3tBwX1e5G1gVZpxvx3ecHqeZKdkhY41fOfFvFKyA7BMRuG2D94IrenIXXR8uqLe+9wQ72U/ihv0xbjuHpj4+QheWb2M7+I8IOWAT91cgzfeXok4fk58rqNQPu8JvXgBX2/4LdberfuJ4PIK3pjeLZVtHdnVzQe++Wkla+/CXB8fgZlO7Eqn5+WjEY+6ugochI/kLY3nD+ONN8Xz4XT+jMj6keaTkeZDwJFN2LhOcZXkZyzlu1BPLPibQEk5LfgZijd8lkccE8J9A9w+Lg6Zfw+f6iuzv8WyUMZlCp3LxxF4p/c13G6fy9p6V/lOS/JxQcT1Ih1lbe/k+VaRzxfwNXXn4428luSn9i7xhu4C3Jbe12RclWUQ190rQudW4Y3Em6jYXd6frlrhycjo15B/9PdLZvb9vuaSkbGuIZePjHUd8o8i72Jm72/kN8OR18BlZGRkZGRkZPQz5AZcRkZGRkZGRkY/Q55CzcjIyMjIyMjoZ8gjcBkZGRkZGRkZ/Qy5AZeRkZGRkZGR0c+QG3AZGRkZGRkZGf0MuQGXkZGRkZGRkdHPkBtwGRkZGRkZGRn9DLkBl5GRkZGRkZHRz/D/AQtllsHzGf7UAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "proteasome_proteins = subunitNames['PA700 complex'] + subunitNames['26S proteasome'] + subunitNames['20S proteasome'] + subunitNames['PA28-20S proteasome'] + subunitNames['PA700-20S-PA28 complex'] + subunitNames['PA28gamma-20S proteasome']\n",
    "proteasome_proteins = list(set(proteasome_proteins))\n",
    "plot_complex_clinical('colon', 'Mutation_Phenotype', proteasome_proteins, expression_change='increased', complex_name = 'Proteasome (All Proteins)', bin_size = 3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tidy_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Thanks StackOverflow! \n",
    "# https://stackoverflow.com/questions/12680754/split-explode-pandas-dataframe-string-entry-to-separate-rows\n",
    "\n",
    "def tidy_split(df, column, sep='|', keep=False):\n",
    "    \"\"\"\n",
    "    Split the values of a column and expand so the new DataFrame has one split\n",
    "    value per row. Filters rows where the column is missing.\n",
    "\n",
    "    Params\n",
    "    ------\n",
    "    df : pandas.DataFrame\n",
    "        dataframe with the column to split and expand\n",
    "    column : str\n",
    "        the column to split and expand\n",
    "    sep : str\n",
    "        the string used to split the column's values\n",
    "    keep : bool\n",
    "        whether to retain the presplit value as it's own row\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        Returns a dataframe with the same columns as `df`.\n",
    "    \"\"\"\n",
    "    indexes = list()\n",
    "    new_values = list()\n",
    "    #df = df.dropna(subset=[column])\n",
    "    for i, presplit in enumerate(df[column].astype(str)):\n",
    "        values = presplit.split(sep)\n",
    "        if keep and len(values) > 1:\n",
    "            indexes.append(i)\n",
    "            new_values.append(presplit)\n",
    "        for value in values:\n",
    "            indexes.append(i)\n",
    "            new_values.append(value)\n",
    "    new_df = df.iloc[indexes, :].copy()\n",
    "    new_df[column] = new_values\n",
    "    return new_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
