{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting functions\n",
    "Note: This is meant to be run AFTER Analysis_Functions.ipynb, as it requires datasets loaded in that notebook. This setup is to prevent having to re-load the data with every small change to a plotting function."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_ratios\n",
    "-----------\n",
    "Create a seaborn plot for the ratios in a dataframe\n",
    "\n",
    "Parameters:\n",
    "    prot1, prot2 = the two proteins to plot (ratio of prot1 / prot2)\n",
    "    by_patient (optional) = T/F whether or not to visualize matched samples\n",
    "    mutation_list (optional) = plot will distinguish patients with a mutation in any of the proteins in this list\n",
    "    \n",
    "Returns:\n",
    "    Displays a plot of the data\n",
    "    \n",
    "\"\"\"\n",
    "def plot_ratios(prot1, prot2, by_patient = False, mutation_list = None):\n",
    "    \n",
    "    # Set up dataframe for plotting\n",
    "    tumor_ratio_df, normal_ratio_df = get_ratio_df(prot1, prot2)\n",
    "    plot_data = tumor_ratio_df.append(normal_ratio_df)\n",
    "    \n",
    "    # Reformat for visualizing matched samples if necessary\n",
    "    if by_patient or mutation_list:\n",
    "        # Create a new column for matched status\n",
    "        tumor_ratio_df['Matched_Status'] = 'Unmatched'\n",
    "        normal_ratio_df['Matched_Status'] = 'Unmatched'\n",
    "        # Classify samples as matched/unmatched\n",
    "        normal_ratio_df.loc[((normal_ratio_df['Patient_ID']).str.replace('_NM', '')).isin(tumor_ratio_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_ratio_df.loc[((tumor_ratio_df['Patient_ID']) + '_NM').isin(normal_ratio_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        # Label samples accordingly in plot_data\n",
    "        plot_data = tumor_ratio_df.append(normal_ratio_df)\n",
    "        if mutation_list:\n",
    "            # Find patients with mutations in the given proteins\n",
    "            patient_list = find_unique_mutations(mutation_list)\n",
    "            plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Matched_Status'] = 'Has_Mutation'\n",
    "        plot_data.loc[plot_data['Matched_Status'] == 'Unmatched', 'Patient_ID'] = 'Unmatched_Sample'\n",
    "        plot_data['Patient_ID'] = plot_data['Patient_ID'].str.replace('_NM','')\n",
    "    \n",
    "    # Reformat for visualizing a certain list of patients if necessary\n",
    "    if mutation_list and not by_patient:\n",
    "        plot_data.loc[~plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Other Mutation'\n",
    "        plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Given Mutation'\n",
    "    elif mutation_list:\n",
    "        plot_data.loc[plot_data['Patient_ID'].isin(patient_list), 'Patient_ID'] = 'Given Mutation'\n",
    "        \n",
    "    # Print results of statistical tests\n",
    "    print('T-test p-value: ' + str(scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]))\n",
    "    print('Levene p-value: ' + str(scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]))\n",
    "    \n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "\n",
    "    # Create the plot\n",
    "    if by_patient or patient_list:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y='Ratio', color='w', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y='Ratio', hue='Patient_ID', size=10, dodge=True, jitter=True)\n",
    "        boxplt.get_legend().set_bbox_to_anchor((1, 1, 0, 0))\n",
    "    else:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y='Ratio', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y='Ratio', dodge=True, jitter=True, color='.3')\n",
    "\n",
    "    # Add styling\n",
    "    boxplt.set_title('Ratio of ' + prot1 + ' / ' + prot2, fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Protein Expression Ratio', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='15')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_phosphoproteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/24685012/pandas-dataframe-renaming-multiple-identically-named-columns\n",
    "# Rename duplicated columns in a dataframe so that each column has a unique name\n",
    "def df_column_uniquify(df):\n",
    "    df_columns = df.columns\n",
    "    new_columns = []\n",
    "    for item in df_columns:\n",
    "        counter = 1\n",
    "        newitem = item\n",
    "        while newitem in new_columns:\n",
    "            counter += 1\n",
    "            newitem = \"{}_{}\".format(item, counter)\n",
    "        new_columns.append(newitem)\n",
    "    df.columns = new_columns\n",
    "    return df\n",
    "\n",
    "def plot_phosphoproteomics(protein, by_patient = False, print_pvals = True, remove_duplicates = False, cancer_type = 'ov'):\n",
    "    \n",
    "    # Get data from the appropriate cancer type\n",
    "    if cancer_type == 'ov':\n",
    "        phos_data = ov.join_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    elif cancer_type == 'colon': \n",
    "        phos_data = colon.join_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    elif cancer_type == 'en': \n",
    "        phos_data = en.join_metadata_to_omics('clinical', 'phosphoproteomics', \n",
    "                            metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                            omics_genes = protein)\n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "    \n",
    "    if remove_duplicates: phos_data = phos_data.loc[:, ~phos_data.columns.duplicated()]\n",
    "    else: phos_data = df_column_uniquify(phos_data)\n",
    "    plot_data = pd.melt(phos_data, id_vars = ['Patient_ID', 'Sample_Tumor_Normal'], var_name = 'Location', value_name = 'Reading').dropna(axis = 0)\n",
    "    plot_data['Location'] = plot_data['Location'].str.replace('_phosphoproteomics', '')\n",
    "    \n",
    "    # Perform t-tests on tumor vs normal phosphoproteomics for each site\n",
    "    if print_pvals:\n",
    "        for column in phos_data.columns:\n",
    "            if column != 'Patient_ID' and column != 'Sample_Tumor_Normal':\n",
    "                sitedf = phos_data[['Patient_ID', 'Sample_Tumor_Normal', column]]\n",
    "                tumordf = sitedf.loc[sitedf['Sample_Tumor_Normal'] == 'Tumor'].dropna(axis = 0)\n",
    "                normaldf = sitedf.loc[sitedf['Sample_Tumor_Normal'] == 'Normal'].dropna(axis = 0)\n",
    "                if len(tumordf) > 2 and len(normaldf) > 2:\n",
    "                    pval = scipy.stats.ttest_ind(tumordf[column], normaldf[column])[1]\n",
    "                    print(column + ' t-test: ' + str(pval))\n",
    "    \n",
    "    # Plot the data\n",
    "    a4_dims = (20, 20)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "\n",
    "    boxplt = sns.boxplot(data=plot_data, x='Location', y='Reading', hue='Sample_Tumor_Normal', showfliers=False, palette = 'Blues')\n",
    "    boxplt = sns.stripplot(data=plot_data, x='Location', y='Reading', hue='Sample_Tumor_Normal', dodge=True, jitter=True, color='.3')\n",
    "\n",
    "    # Add styling\n",
    "    boxplt.set_title(protein + ' Phosphorylation', fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Phosphorylation Level', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='10')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_proteomics\n",
    "---------------\n",
    "Create a seaborn plot of the proteomics data for a single protein\n",
    "\n",
    "Parameters:\n",
    "    protein = the protein to plot\n",
    "    use_cptac (optional) = T/F whether to use the CPTAC proteomics data (default) or the non-normalized data\n",
    "    by_patient (optional) = T/F whether or not to visualize matched samples\n",
    "    print_pvals (optional) = T/F whether or not to print the p-values of the statistical tests\n",
    "    \n",
    "Returns:\n",
    "    Displays a plot of the data\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def plot_proteomics(protein, use_cptac = True, by_patient = False, print_pvals = True, cancer_type = 'ov', show_subtype = False):\n",
    "    \n",
    "    # This uses the normalized proteomics data from the CPTAC package\n",
    "    if use_cptac:\n",
    "        # Get data from the appropriate cancer type\n",
    "        if cancer_type == 'ov':\n",
    "            plot_data = ov.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'colon': \n",
    "            plot_data = colon.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'en': \n",
    "            plot_data = en.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'renal':\n",
    "            plot_data = renal.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        else: \n",
    "            print('Error: cancer_type must be \"ov\", \"colon\", \"en\", or \"renal\"')\n",
    "            return\n",
    "            \n",
    "        plot_data = plot_data.loc[:, ~plot_data.columns.duplicated()]\n",
    "        if cancer_type == 'en':\n",
    "            plot_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            plot_data.loc[plot_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "        else:\n",
    "            plot_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "        plot_data = plot_data.dropna(axis = 0)\n",
    "        plot_data['Matched_Status'] = 'Unmatched'\n",
    "        tumor_df = plot_data.loc[plot_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = plot_data.loc[plot_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Otherwise use the non-normalized data to make the plot\n",
    "    else:\n",
    "        plot_data = pd.DataFrame(data.loc[data.index == protein].transpose())\n",
    "        plot_data['Sample_Type'] = 'Tumor'\n",
    "        plot_data.loc[plot_data.index.str.contains('_NM'), 'Sample_Type'] = 'Normal'\n",
    "\n",
    "        plot_data['Patient_ID'] = plot_data.index\n",
    "        plot_data['Matched_Status'] = 'Unmatched'\n",
    "        tumor_df = plot_data.loc[plot_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = plot_data.loc[plot_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Format to show matched patients if necessary\n",
    "    if by_patient:\n",
    "        pd.options.mode.chained_assignment = None \n",
    "        \n",
    "        # Classify samples as matched/unmatched\n",
    "        normal_df.loc[((normal_df['Patient_ID']).str.replace('_NM|N', '')).isin(tumor_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[((tumor_df['Patient_ID']) + '_NM').isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[((tumor_df['Patient_ID']) + 'N').isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        tumor_df.loc[('N' + (tumor_df['Patient_ID'])).isin(normal_df['Patient_ID']), 'Matched_Status'] = 'Matched'\n",
    "        \n",
    "        # Label samples accordingly in plot_data\n",
    "        plot_data = tumor_df.append(normal_df)\n",
    "        plot_data.loc[plot_data['Matched_Status'] == 'Unmatched', 'Patient_ID'] = 'Unmatched_Sample'\n",
    "        plot_data['Patient_ID'] = plot_data['Patient_ID'].str.replace('_NM|N','')\n",
    "        \n",
    "        if show_subtype:\n",
    "            plot_data = pd.merge(plot_data, colon.get_clinical()[['Mutation_Phenotype', 'Patient_ID']], on = 'Patient_ID')\n",
    "             \n",
    "    if print_pvals:\n",
    "        # Print results of statistical tests\n",
    "        print('T-test p-value: ' + str(scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]))\n",
    "        print('Levene p-value: ' + str(scipy.stats.levene(tumor_df[protein], normal_df[protein])[1]))\n",
    "    \n",
    "    # Create the plot\n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "    \n",
    "    if by_patient and show_subtype:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y=protein, color='w', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y=protein, hue = 'Mutation_Phenotype', size=10, dodge=True, jitter=True)\n",
    "        boxplt.get_legend().set_bbox_to_anchor((1, 1, 0, 0))\n",
    "    elif by_patient:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y=protein, color='w', showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y=protein, hue='Patient_ID', size=10, dodge=True, jitter=True)\n",
    "        boxplt.get_legend().set_bbox_to_anchor((1, 1, 0, 0))\n",
    "    else:\n",
    "        boxplt = sns.boxplot(data=plot_data, x='Sample_Type', y=protein, showfliers=False)\n",
    "        boxplt = sns.stripplot(data=plot_data, x='Sample_Type', y=protein, dodge=True, jitter=True, color='.3')\n",
    "        \n",
    "    # Add styling\n",
    "    boxplt.set_title(protein + ' Proteomics', fontsize='25')\n",
    "    boxplt.set_xlabel('')\n",
    "    boxplt.set_ylabel('Protein Expression', fontsize='20')\n",
    "    boxplt.tick_params(labelsize='15')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_complex_clinical"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "plot_complex_clinical\n",
    "---------------------\n",
    "Creates a countplot of the number of proteins in a complex that are differentially expressed in tumor/normal matched samples\n",
    "Also displays the given clinical feature \n",
    "\n",
    "Parameters:\n",
    "    cancer_type = can be colon, en (endometrial)\n",
    "    clinical_feature = the name of the column from the clinical dataframe that you want to visualize\n",
    "    protein_list = list of proteins to analyze (most likely the proteins in a complex)\n",
    "    expression_change = increased, decreased, or nochange; will determine if the plot shows the number\n",
    "        of proteins with increased expression in tumor samples, decreased in tumor samples, or no change between\n",
    "        tumor/normal\n",
    "    bin_size = size of the bins to put patients in as number of proteins on the x-axis (default is 1)\n",
    "    complex_name (optional) = The title for the plot\n",
    "    show_proportions = whether to display proportions of each group in each category or raw numbers\n",
    "    \n",
    "Returns:\n",
    "    Displays a Seaborn countplot\n",
    "    x-axis: Number of proteins with differential expression\n",
    "    y-axis: Number of patients in this category\n",
    "    hue: clinical feature\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "def plot_complex_clinical(cancer_type, clinical_feature, raw_protein_list, expression_change = 'increased', bin_size = 1, complex_name = '', show_proportions = True, boxplot = False):\n",
    "    # Remove duplicates from protein list\n",
    "    raw_protein_list = list(set(raw_protein_list))\n",
    "    protein_list = []\n",
    "    \n",
    "    if cancer_type == 'colon':\n",
    "        # Make sure all the proteins in protein_list are in the proteomics df\n",
    "        for protein in raw_protein_list:\n",
    "            if protein in colon.get_proteomics().columns:\n",
    "                protein_list.append(protein)\n",
    "        # Start out with a df including the patient IDs and the given clinical feature\n",
    "        plot_data = pd.DataFrame()\n",
    "        plot_data = colon.get_clinical()[['Patient_ID', clinical_feature]]\n",
    "        plot_data['Sample_ID'] = plot_data.index\n",
    "        \n",
    "        # Add protein expression values for all proteins in the given list\n",
    "        proteomics_data = colon.get_proteomics()[protein_list]\n",
    "        plot_data = plot_data.join(proteomics_data)\n",
    "\n",
    "        # Extract all the normal matched samples into a new df and remove the 'N' to match them with tumor samples\n",
    "        normal = plot_data.copy()\n",
    "        normal = normal.loc[normal['Patient_ID'].str.contains('N')]\n",
    "        normal['Patient_ID'] = normal['Patient_ID'].str.replace('N', '')\n",
    "        normal.index = normal['Patient_ID']\n",
    "\n",
    "        # Create new column to count the number of proteins with increased/decreased/nochange levels in tumor samples\n",
    "        # Remove the normal samples from plot_data\n",
    "        plot_data[expression_change] = 0 \n",
    "        plot_data = plot_data.loc[~plot_data['Patient_ID'].str.contains('N')]\n",
    "        plot_data.index = plot_data['Patient_ID']\n",
    "\n",
    "        # Remove unmatched samples from each dataframe\n",
    "        plot_data = plot_data.loc[plot_data['Patient_ID'].isin(normal['Patient_ID'])]\n",
    "        normal = normal.loc[normal['Patient_ID'].isin(plot_data['Patient_ID'])] \n",
    "    \n",
    "    elif cancer_type == 'en':\n",
    "        # Make sure all the proteins in protein_list are in the proteomics df\n",
    "        for protein in raw_protein_list:\n",
    "            if protein in en.get_proteomics().columns:\n",
    "                protein_list.append(protein)\n",
    "        \n",
    "        # Start out with a df including the patient IDs and the given clinical feature\n",
    "        plot_data = pd.DataFrame()\n",
    "        plot_data = en.get_clinical()[['Patient_ID', 'Proteomics_Tumor_Normal']]\n",
    "        plot_data['Sample_ID'] = plot_data.index\n",
    "        \n",
    "        # Add protein expression values for all proteins in the given list\n",
    "        #proteomics_data = en.get_proteomics()[protein_list]\n",
    "        proteomics_data = en.join_metadata_to_omics('derived_molecular', 'proteomics', clinical_feature, protein_list)\n",
    "        plot_data = plot_data.join(proteomics_data)\n",
    "        plot_data.columns = plot_data.columns.str.replace('_proteomics', '')\n",
    "\n",
    "        # Extract all the normal matched samples into a new df and remove the 'N' to match them with tumor samples\n",
    "        normal = plot_data.copy()\n",
    "        normal = normal.loc[normal['Proteomics_Tumor_Normal'].str.contains('Adjacent_normal')]\n",
    "        normal.index = normal['Patient_ID']\n",
    "\n",
    "        # Create new column to count the number of proteins with increased/decreased/nochange levels in tumor samples\n",
    "        # Remove the normal samples from plot_data\n",
    "        plot_data[expression_change] = 0 \n",
    "        plot_data = plot_data.loc[plot_data['Proteomics_Tumor_Normal'] == 'Tumor']\n",
    "        plot_data.index = plot_data['Patient_ID']\n",
    "\n",
    "        # Remove unmatched samples from each dataframe\n",
    "        plot_data = plot_data.loc[plot_data['Patient_ID'].isin(normal['Patient_ID'])]\n",
    "        normal = normal.loc[normal['Patient_ID'].isin(plot_data['Patient_ID'])] \n",
    "    \n",
    "    elif cancer_type == 'renal':\n",
    "        # Make sure all the proteins in protein_list are in the proteomics df\n",
    "        for protein in raw_protein_list:\n",
    "            if protein in renal.get_proteomics().columns:\n",
    "                protein_list.append(protein)\n",
    "        \n",
    "        # Start out with a df including the patient IDs and the given clinical feature\n",
    "        plot_data = pd.DataFrame()\n",
    "        plot_data = renal.get_clinical()[['Patient_ID', clinical_feature, 'Sample_Tumor_Normal']]\n",
    "        plot_data['Sample_ID'] = plot_data.index\n",
    "        \n",
    "        # Add protein expression values for all proteins in the given list\n",
    "        proteomics_data = renal.get_proteomics()[protein_list]\n",
    "        # The renal dataset has duplicate columns for the same protein in some cases\n",
    "        # We'll keep only the first column with that protein name\n",
    "        proteomics_data = proteomics_data.loc[:, ~proteomics_data.columns.duplicated()]\n",
    "        plot_data = plot_data.join(proteomics_data)\n",
    "\n",
    "        # Extract all the normal matched samples into a new df and remove the 'N' to match them with tumor samples\n",
    "        normal = plot_data.copy()\n",
    "        normal = normal.loc[normal['Sample_Tumor_Normal'] == 'Normal']\n",
    "        normal['Patient_ID'] = normal['Patient_ID'].str.replace('NC', 'C')\n",
    "        normal.index = normal['Patient_ID']\n",
    "\n",
    "        # Create new column to count the number of proteins with increased/decreased/nochange levels in tumor samples\n",
    "        # Remove the normal samples from plot_data\n",
    "        plot_data[expression_change] = 0 \n",
    "        plot_data = plot_data.loc[plot_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "        plot_data.index = plot_data['Patient_ID']\n",
    "        \n",
    "        # Remove unmatched samples from each dataframe\n",
    "        plot_data = plot_data.loc[plot_data['Patient_ID'].isin(normal['Patient_ID'])]\n",
    "        normal = normal.loc[normal['Patient_ID'].isin(plot_data['Patient_ID'])] \n",
    "    \n",
    "    else: \n",
    "        print('Error: invalid or unimplemented cancer type')\n",
    "        return\n",
    "    \n",
    "    # Fill out the Num_* column by comparing the protein expression for matched patients in plot_data/normal dfs\n",
    "    for protein in protein_list:\n",
    "        mean_diff = np.abs(np.mean(plot_data[protein]) - np.mean(normal[protein]))\n",
    "        if expression_change == 'increased':\n",
    "            plot_data.loc[((plot_data[protein] - normal[protein]) >= mean_diff), expression_change] += 1\n",
    "        elif expression_change == 'decreased':\n",
    "            plot_data.loc[((plot_data[protein] - normal[protein]) <= (-1*mean_diff)), expression_change] += 1\n",
    "        else:\n",
    "            plot_data.loc[((plot_data[protein] - normal[protein]) <= mean_diff)\n",
    "                      & ((plot_data[protein] - normal[protein]) >= (-1*mean_diff)), expression_change] += 1\n",
    "\n",
    "    # Call another function to create the plot\n",
    "    if boxplot:\n",
    "        create_complex_clinical_boxplot(plot_data[[clinical_feature, expression_change]], clinical_feature, expression_change, complex_name)\n",
    "    else:\n",
    "        create_complex_clinical_plot(plot_data, clinical_feature, protein_list, bin_size, expression_change, complex_name, show_proportions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_complex_clinical_boxplot(plot_data, clinical_feature, expression_change, complex_name):\n",
    "    clinical_vals = list(set(plot_data[clinical_feature].dropna()))\n",
    "    for i in range(0, len(clinical_vals)):\n",
    "        for j in range(i+1, len(clinical_vals)):\n",
    "            val_1 = plot_data.loc[plot_data[clinical_feature] == clinical_vals[i]]\n",
    "            val_2 = plot_data.loc[plot_data[clinical_feature] == clinical_vals[j]]\n",
    "            pval = scipy.stats.ttest_ind(val_1[expression_change], val_2[expression_change])[1]\n",
    "            print('t-test result ' + str(clinical_vals[i]) + ' vs ' + str(clinical_vals[j]) + ': ' + str(pval))\n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "    sns.set(font_scale = 1.5, style = 'white')\n",
    "    boxplt = sns.boxplot(data = plot_data,\n",
    "                           x = clinical_feature,\n",
    "                           y = expression_change,\n",
    "                           showfliers = False)\n",
    "    boxplt = sns.stripplot(data=plot_data, \n",
    "                           x = clinical_feature, \n",
    "                           y = expression_change, \n",
    "                           dodge=True, \n",
    "                           jitter=True, \n",
    "                           color='.3')\n",
    "    \n",
    "    boxplt.set_title(complex_name, fontsize = '20')\n",
    "    boxplt.set_ylabel('Number of proteins ' + expression_change + ' in ' + complex_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create_complex_clinical_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "create_complex_clinical_plot\n",
    "----------------------------\n",
    "Function called by plot_complex_clinical; this prepares the dataframe for plotting and creates the actual plot\n",
    "\n",
    "Parameters are the same as those passed to plot_complex_clinical\n",
    "\n",
    "\"\"\"\n",
    "def create_complex_clinical_plot(plot_data, clinical_feature, protein_list, bin_size, expression_change, complex_name, show_proportions):\n",
    "    # Drop the proteomics data (not necessary at this point)\n",
    "    plot_data.drop(protein_list, axis = 1, inplace=True)\n",
    "    category_col_title = expression_change + ' category'\n",
    "    sns.set(font_scale = 1, style='white')\n",
    "    # If indicated, bin the patients into groups based on number of proteins increased/decreased/nochange\n",
    "    # This makes it easier to view, otherwise there are lots of ticks on the x-axis\n",
    "    if bin_size > 1:\n",
    "        # Create the list of cutoff values for each bin based on the given bin_size\n",
    "        bin_vals = list(range(0, len(protein_list), bin_size))\n",
    "        # Create the list of strings to use as category names\n",
    "        category_names = []\n",
    "        for i in range(0, len(bin_vals) - 1):\n",
    "            category_names.append(str(bin_vals[i]) + '-' + str(bin_vals[i+1] - 1))\n",
    "        category_names.append(str(bin_vals[len(bin_vals)-1]) + '+')\n",
    "        \n",
    "        # Place each sample into the correct category\n",
    "        plot_data[category_col_title] = category_names[0]\n",
    "        for i in range(1, len(bin_vals) - 1):\n",
    "            plot_data.loc[((plot_data[expression_change] > bin_vals[i]) & (plot_data[expression_change] <= bin_vals[i+1])), category_col_title] = category_names[i]\n",
    "            \n",
    "        # Create the appropriate plot based on the desired differential expression type\n",
    "        # increased means increased in tumor samples\n",
    "        a4_dims = (10, 10)\n",
    "        fig, ax = plt.subplots(figsize=a4_dims)\n",
    "        x_order = category_names\n",
    "        if show_proportions:\n",
    "            percent_df = (plot_data[category_col_title]\n",
    "                            .groupby(plot_data[clinical_feature])\n",
    "                            .value_counts(normalize=True)\n",
    "                            .rename('Proportion ' + clinical_feature)\n",
    "                            .reset_index())\n",
    "            countplot = sns.barplot(data=percent_df, x=category_col_title, y=('Proportion ' + clinical_feature), hue=clinical_feature, order = x_order)\n",
    "            countplot.set_ylabel('Proportion of Individuals', fontsize = '15')\n",
    "            tick_spacing = 0.05\n",
    "            \n",
    "        else:\n",
    "            countplot = sns.countplot(x=category_col_title, hue=clinical_feature, data=plot_data, order = x_order)\n",
    "            countplot.set_ylabel('Number of Individuals', fontsize = '15')\n",
    "            tick_spacing = 2\n",
    "    \n",
    "    else:\n",
    "        # Same idea as above, just without binning the patients\n",
    "        plot_data[category_col_title] = plot_data[expression_change].astype(str)\n",
    "        a4_dims = (10, 10)\n",
    "        fig, ax = plt.subplots(figsize=a4_dims)\n",
    "        x_axis_labels = list(set(plot_data[category_col_title].astype(int).sort_values()))\n",
    "        x_axis_labels = [str(i) for i in x_axis_labels] \n",
    "        \n",
    "        if show_proportions:\n",
    "            percent_df = (plot_data[category_col_title]\n",
    "                            .groupby(plot_data[clinical_feature])\n",
    "                            .value_counts(normalize=True)\n",
    "                            .rename('Proportion ' + clinical_feature)\n",
    "                            .reset_index())\n",
    "            countplot = sns.barplot(data=percent_df, x = category_col_title, y=('Proportion ' + clinical_feature), hue=clinical_feature, order = x_axis_labels)\n",
    "            countplot.set_ylabel('Proportion of Individuals', fontsize = '15')\n",
    "            tick_spacing = 0.05\n",
    "        else:\n",
    "            countplot = sns.countplot(x=category_col_title, hue=clinical_feature, data=plot_data, order = x_axis_labels)\n",
    "            tick_spacing = 2\n",
    "            countplot.set_ylabel('Number of Individuals', fontsize = '15')\n",
    "        \n",
    "    countplot.set_xlabel('Number of proteins with ' + expression_change + ' expression in tumor samples', fontsize = '15')      \n",
    "    ax.yaxis.set_major_locator(ticker.MultipleLocator(tick_spacing))\n",
    "    countplot.set_title(complex_name, fontsize = '25')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### plot_subtype_complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_subtype_complex(raw_protein_list, complex_name = '', cancer_type = 'en', simplify_subtype = True):\n",
    "    protein_list = []\n",
    "    # Make sure all the proteins in protein_list are in the proteomics df\n",
    "    for protein in raw_protein_list:\n",
    "        if protein in en.get_proteomics().columns:\n",
    "            protein_list.append(protein)\n",
    "    if cancer_type == 'en':\n",
    "        # Start out with a df including the patient IDs and molecular subtype\n",
    "        source_data = en.get_derived_molecular()[['Genomics_subtype']]\n",
    "        source_data['Sample_ID'] = source_data.index\n",
    "        \n",
    "        # Add protein expression values for all proteins in the given list\n",
    "        proteomics_data = en.get_proteomics()[protein_list]\n",
    "        patient_ids = en.get_clinical()[['Patient_ID', 'Proteomics_Tumor_Normal']]\n",
    "        source_data = source_data.join(patient_ids)\n",
    "        source_data = source_data.join(proteomics_data)\n",
    "\n",
    "        # Extract all the normal matched samples into a new df and remove the 'N' to match them with tumor samples\n",
    "        normal = source_data.copy()\n",
    "        normal = normal.loc[normal['Proteomics_Tumor_Normal'].str.contains('Adjacent_normal')]\n",
    "        normal.index = normal['Patient_ID']\n",
    "\n",
    "        # Remove the normal samples from the tumor df\n",
    "        tumor = source_data.loc[source_data['Proteomics_Tumor_Normal'] == 'Tumor']\n",
    "        tumor.index = tumor['Patient_ID']\n",
    "\n",
    "        # Remove unmatched samples from each dataframe\n",
    "        tumor = tumor.loc[tumor['Patient_ID'].isin(normal['Patient_ID'])]\n",
    "        normal = normal.loc[normal['Patient_ID'].isin(tumor['Patient_ID'])] \n",
    "        \n",
    "        # Create the plot_data with ratios tumor/normal for each protein\n",
    "        plot_data = tumor.copy()[['Genomics_subtype', 'Sample_ID']]\n",
    "        for protein in protein_list:\n",
    "            plot_data[protein + ' (T/N)'] = tumor[protein] - normal[protein]\n",
    "            \n",
    "        # Create only 3 categories of subtype: MSI-H, POLE, and Other\n",
    "        if simplify_subtype:\n",
    "            plot_data.loc[((plot_data['Genomics_subtype'] != 'MSI-H') \n",
    "                                       & (plot_data['Genomics_subtype'] != 'POLE')), \n",
    "                                      'Genomics_subtype'] = 'Other'\n",
    "        plot_data = pd.melt(plot_data, id_vars = ['Genomics_subtype', 'Sample_ID'], var_name = 'Protein', value_name = 'Ratio')\n",
    "\n",
    "        pole = plot_data.loc[plot_data['Genomics_subtype'] == 'POLE'].groupby('Sample_ID').mean()\n",
    "        msi = plot_data.loc[plot_data['Genomics_subtype'] == 'MSI-H'].groupby('Sample_ID').mean()\n",
    "        other = plot_data.loc[plot_data['Genomics_subtype'] == 'Other'].groupby('Sample_ID').mean()\n",
    "\n",
    "        mwu_msi_pole = scipy.stats.mannwhitneyu(msi['Ratio'], pole['Ratio'])[1]\n",
    "        mwu_msi_other = scipy.stats.mannwhitneyu(msi['Ratio'], other['Ratio'])[1]\n",
    "        mwu_pole_other = scipy.stats.mannwhitneyu(pole['Ratio'], other['Ratio'])[1]\n",
    "        print('p-value for Mann-Whitney rank test test MSI-H vs POLE ' + str(mwu_msi_pole))\n",
    "        print('p-value for Mann-Whitney rank test test MSI-H vs Other ' + str(mwu_msi_other))\n",
    "        print('p-value for Mann-Whitney rank test test POLE vs Other ' + str(mwu_pole_other))\n",
    "        \n",
    "    elif cancer_type == 'colon':\n",
    "        source_data = colon.get_clinical()[['Mutation_Phenotype', 'Patient_ID']]\n",
    "        source_data['Sample_ID'] = source_data.index\n",
    "        \n",
    "        proteomics_data = colon.get_proteomics()[protein_list]\n",
    "        source_data = source_data.join(proteomics_data)\n",
    "\n",
    "        # Extract all the normal matched samples into a new df and remove the 'N' to match them with tumor samples\n",
    "        normal = source_data.copy()\n",
    "        normal = normal.loc[normal['Patient_ID'].str.contains('N')]\n",
    "        normal['Patient_ID'] = normal['Patient_ID'].str.replace('N', '')\n",
    "        normal.index = normal['Patient_ID']\n",
    "\n",
    "        # Remove the normal samples from the tumor df\n",
    "        tumor = source_data.loc[~source_data['Patient_ID'].str.contains('N')]\n",
    "        tumor.index = tumor['Patient_ID']\n",
    "\n",
    "        # Remove unmatched samples from each dataframe\n",
    "        tumor = tumor.loc[tumor['Patient_ID'].isin(normal['Patient_ID'])]\n",
    "        normal = normal.loc[normal['Patient_ID'].isin(tumor['Patient_ID'])] \n",
    "        \n",
    "        plot_data = tumor.copy()[['Sample_ID']]\n",
    "        plot_data['Genomics_subtype'] = tumor['Mutation_Phenotype']\n",
    "        print(tumor)\n",
    "        print(normal)\n",
    "        for protein in protein_list:\n",
    "            plot_data[protein + ' (T/N)'] = (tumor[protein] + 10) - (normal[protein] + 10)\n",
    "        print(plot_data)\n",
    "        plot_data = pd.melt(plot_data, id_vars = ['Genomics_subtype', 'Sample_ID'], var_name = 'Protein', value_name = 'Ratio')\n",
    "        \n",
    "        mss = plot_data.loc[plot_data['Genomics_subtype'] == 'MSS'].groupby('Sample_ID').mean()\n",
    "        msi = plot_data.loc[plot_data['Genomics_subtype'] == 'MSI-H'].groupby('Sample_ID').mean()\n",
    "\n",
    "        whitneyutest = scipy.stats.mannwhitneyu(mss['Ratio'], msi['Ratio'])[1]\n",
    "        print('p-value for Mann-Whitney rank test test MSS vs MSI-H: ' + str(whitneyutest))\n",
    "        \n",
    "    else:\n",
    "        print('Error: invalid or unimplemented cancer type. Options are: en')\n",
    "        return\n",
    "    \n",
    "    create_complex_boxplot(plot_data, complex_name, protein_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create_complex_boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_complex_boxplot(plot_data, complex_name, protein_list):\n",
    "    a4_dims = (100, 50)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "    sns.set(font_scale = 5, style = 'white')\n",
    "    boxplt = sns.boxplot(data = plot_data,\n",
    "                           x = 'Genomics_subtype',\n",
    "                           y = 'Ratio',\n",
    "                           hue = 'Sample_ID',\n",
    "                           showfliers = False)\n",
    "    boxplt = sns.stripplot(data=plot_data, \n",
    "                           x='Genomics_subtype', \n",
    "                           y='Ratio', \n",
    "                           hue = 'Sample_ID',\n",
    "                           dodge=True, \n",
    "                           jitter=True, \n",
    "                           color='.3')\n",
    "    \n",
    "    boxplt.set_title(complex_name, fontsize = '300')\n",
    "    boxplt.set_xlabel('Genomic Subtype', fontsize = '100')\n",
    "    boxplt.set_ylabel('Ratio (Tumor/Normal)', fontsize = '100')\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "    l = plt.legend(handles[0:0], labels[0:0])\n",
    "    boxplt.set_title(complex_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ttest_all_complexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ttest_all_complexes(proteomics_data, change = 'increased', alpha = 0.05):\n",
    "    results = pd.DataFrame(columns = ['complex_name', 'total_protein_num', 'num_increased_tumor', \n",
    "                                      'num_decreased_tumor', 'no_proteomics_data'])\n",
    "    for complex_name, proteins in subunitNames.items():\n",
    "        num_sig_increased = 0\n",
    "        num_sig_decreased = 0\n",
    "        num_no_data = 0\n",
    "        tumor = proteomics_data.loc[proteomics_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "        normal = proteomics_data.loc[proteomics_data['Sample_Tumor_Normal'] == 'Normal']\n",
    "        \n",
    "        for protein in proteins:\n",
    "            if protein in proteomics_data.columns:\n",
    "                tumor_vals = tumor[protein].dropna()\n",
    "                normal_vals = normal[protein].dropna()\n",
    "                if len(tumor_vals) > 2 and len(normal_vals) > 2:\n",
    "                    p_val = scipy.stats.ttest_ind(tumor_vals, normal_vals)[1]\n",
    "                    if p_val < alpha:\n",
    "                        if (np.mean(tumor[protein]) > np.mean(normal[protein])):\n",
    "                            num_sig_increased += 1\n",
    "                        elif (np.mean(tumor[protein]) < np.mean(normal[protein])):\n",
    "                            num_sig_decreased += 1\n",
    "            else:\n",
    "                num_no_data += 1\n",
    "        results = results.append({'complex_name': complex_name,\n",
    "                      'total_protein_num': len(proteins),\n",
    "                      'num_increased_tumor': num_sig_increased,\n",
    "                      'num_decreased_tumor': num_sig_decreased,\n",
    "                      'no_proteomics_data': num_no_data}, ignore_index = True)\n",
    "        \n",
    "    results['percent_increased'] = (results['num_increased_tumor'] / results['total_protein_num'])*100\n",
    "    results['percent_decreased'] = (results['num_decreased_tumor'] / results['total_protein_num'])*100\n",
    "    return results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### make_histogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_histogram(results, show_change = 'percent_increased', size_cutoffs = [10, 20], norm = False, show_complex = None):\n",
    "    a4_dims = (10, 10)\n",
    "    fig, ax = plt.subplots(figsize=a4_dims)\n",
    "    color_vals = ['skyblue', 'olive', 'gold', 'teal', 'red']\n",
    "    if norm:\n",
    "        if size_cutoffs is not None:\n",
    "            plot_data = list(results.loc[results['total_protein_num'] < size_cutoffs[0]][show_change])\n",
    "            distplot = sns.distplot(plot_data, color = color_vals[0], label = 'n < ' + str(size_cutoffs[0]), kde = False, norm_hist = True, bins = 20)\n",
    "            for i in range(1, len(size_cutoffs)):\n",
    "                plot_data = results.loc[(results['total_protein_num'] >= size_cutoffs[i-1]) & (results['total_protein_num'] < size_cutoffs[i])]\n",
    "                plot_data = list(plot_data[show_change])\n",
    "                distplot = sns.distplot(plot_data, color = color_vals[i], kde = False, norm_hist = True, label = (str(size_cutoffs[i-1]) + ' <= n < ' + str(size_cutoffs[i])), bins = 20)\n",
    "            plot_data = list(results.loc[results['total_protein_num'] >= size_cutoffs[len(size_cutoffs) - 1]][show_change])\n",
    "            distplot = sns.distplot(plot_data, color = color_vals[len(color_vals) - 1], kde = False, norm_hist = True, label = 'n >= ' + str(size_cutoffs[len(size_cutoffs)-1]), bins = 20)\n",
    "            plt.legend()\n",
    "        else:\n",
    "            plot_data = list(results[show_change])\n",
    "            distplot = sns.distplot(plot_data, color = color_vals[0], kde = False, norm_hist = True, label = 'No size cutoffs')\n",
    "            plt.legend()\n",
    "    else:\n",
    "        if size_cutoffs is not None:\n",
    "            plot_data = list(results.loc[results['total_protein_num'] < size_cutoffs[0]][show_change])\n",
    "            distplot = sns.distplot(plot_data, color = color_vals[0], norm_hist = False, kde = False, label = 'n < ' + str(size_cutoffs[0]), bins = 20)\n",
    "            for i in range(1, len(size_cutoffs)):\n",
    "                plot_data = results.loc[(results['total_protein_num'] >= size_cutoffs[i-1]) & (results['total_protein_num'] < size_cutoffs[i])]\n",
    "                plot_data = list(plot_data[show_change])\n",
    "                distplot = sns.distplot(plot_data, color = color_vals[i], norm_hist = False, kde = False, label = (str(size_cutoffs[i-1]) + ' <= n < ' + str(size_cutoffs[i])), bins = 20)\n",
    "            plot_data = list(results.loc[results['total_protein_num'] >= size_cutoffs[len(size_cutoffs) - 1]][show_change])\n",
    "            distplot = sns.distplot(plot_data, color = color_vals[len(color_vals) - 1], norm_hist = False, kde = False, label = 'n >= ' + str(size_cutoffs[len(size_cutoffs)-1]))\n",
    "            plt.legend()\n",
    "        else:\n",
    "            plot_data = list(results[show_change])\n",
    "            distplot = sns.distplot(plot_data, color = color_vals[0], norm_hist = False, kde = False, label = 'No size cutoffs', bins = 20)\n",
    "            plt.legend()\n",
    "    \n",
    "    if show_complex is not None:\n",
    "        num_changed = int(results.loc[results['complex_name'] == show_complex][show_change])\n",
    "        plt.axvline(num_changed, 0,0.75, color = 'black')\n",
    "    distplot.set_title('Percent of proteins in complex significantly increased in tumor')\n",
    "    plt.legend().set_title('Total number of proteins in complex')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pancancer_boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "pancancer_boxplot\n",
    "-----------------\n",
    "Create boxplots of tumor vs normal for all proteins in a given complex for all cancer types\n",
    "\n",
    "'''\n",
    "def pancancer_boxplot(complex_name = None, protein_list = None, show_ttest = True):\n",
    "    \n",
    "    if complex_name:\n",
    "        protein_list = subunitNames[complex_name]\n",
    "    else:\n",
    "        complex_name = 'Unique Protein List'\n",
    "        \n",
    "    proteomics_dict = get_pancancer_proteomics(protein_list)\n",
    "    if show_ttest:\n",
    "        pancancer_ttest(complex_name, protein_list)\n",
    "        \n",
    "    for key, val in proteomics_dict.items():\n",
    "        a4_dims = (11.7, 8.27)\n",
    "        fig, ax = plt.subplots(figsize = a4_dims)\n",
    "        plot_data = pd.melt(val, id_vars = ['Sample_Tumor_Normal'], var_name = 'Protein', value_name = 'Proteomics')\n",
    "        boxplt = sns.boxplot(data = plot_data, x = 'Protein', y = 'Proteomics', hue = 'Sample_Tumor_Normal', showfliers = False)\n",
    "        boxplt = sns.stripplot(data = plot_data, x = 'Protein', y = 'Proteomics', hue = 'Sample_Tumor_Normal', dodge = True, jitter = True, color = '.3')\n",
    "        boxplt.set_title(key, fontsize = '20')\n",
    "        boxplt.set_ylabel('Proteomics', fontsize = '15')\n",
    "        boxplt.set_xlabel(complex_name, fontsize = '15')\n",
    "        handles, labels = ax.get_legend_handles_labels()\n",
    "        l = plt.legend(handles[0:2], labels[0:2])\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pancancer_ttest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pancancer_ttest(complex_name, protein_list):\n",
    "    proteomics_dict = get_pancancer_proteomics(protein_list)\n",
    "    print(complex_name + '\\n')\n",
    "    for key, val in proteomics_dict.items():\n",
    "        print(key + ' cancer:')\n",
    "        tumor = val.loc[val['Sample_Tumor_Normal'] == 'Tumor']\n",
    "        normal = val.loc[val['Sample_Tumor_Normal'] == 'Normal']\n",
    "        for protein in protein_list:\n",
    "            pval = scipy.stats.ttest_ind(tumor[protein].dropna(), normal[protein].dropna())[1]\n",
    "            print(protein + ': ' + str(pval))\n",
    "        print('\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get_pancancer_proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pancancer_proteomics(protein_list):\n",
    "    proteomics_dict = {}\n",
    "    \n",
    "    colon_proteomics = colon.join_metadata_to_omics('clinical', 'proteomics', ['Sample_Tumor_Normal'], protein_list)\n",
    "    colon_proteomics.columns = colon_proteomics.columns.str.replace('_proteomics', '')\n",
    "    colon_proteomics = colon_proteomics.loc[:,~colon_proteomics.columns.duplicated()]\n",
    "    proteomics_dict['Colon'] = colon_proteomics\n",
    "\n",
    "    en_proteomics = en.join_metadata_to_omics('clinical', 'proteomics', ['Proteomics_Tumor_Normal'], protein_list)\n",
    "    en_proteomics.columns = en_proteomics.columns.str.replace('Proteomics_Tumor_Normal', 'Sample_Tumor_Normal')\n",
    "    en_proteomics.columns = en_proteomics.columns.str.replace('_proteomics', '')\n",
    "    en_proteomics = en_proteomics.loc[:,~en_proteomics.columns.duplicated()]\n",
    "    en_proteomics.loc[en_proteomics['Sample_Tumor_Normal'] == 'Adjacent_normal', 'Sample_Tumor_Normal'] = 'Normal'\n",
    "    en_proteomics = en_proteomics.loc[(en_proteomics['Sample_Tumor_Normal'] == 'Tumor') | (en_proteomics['Sample_Tumor_Normal'] == 'Normal')]\n",
    "    proteomics_dict['Endometrial'] = en_proteomics\n",
    "    \n",
    "    renal_proteomics = renal.join_metadata_to_omics('clinical', 'proteomics', ['Sample_Tumor_Normal'], protein_list)\n",
    "    renal_proteomics.columns = renal_proteomics.columns.str.replace('_proteomics', '')\n",
    "    renal_proteomics = renal_proteomics.loc[:,~renal_proteomics.columns.duplicated()]\n",
    "    proteomics_dict['Renal'] = renal_proteomics\n",
    "    \n",
    "    ov_proteomics = ov.join_metadata_to_omics('clinical', 'proteomics', ['Sample_Tumor_Normal'], protein_list)\n",
    "    ov_proteomics.columns = ov_proteomics.columns.str.replace('_proteomics', '')\n",
    "    ov_proteomics = ov_proteomics.loc[:,~ov_proteomics.columns.duplicated()]\n",
    "    proteomics_dict['Ovarian'] = ov_proteomics\n",
    "    \n",
    "    return proteomics_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pancancer_analyze_complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pancancer_analyze_complex(complex_name, include_borderline = False):\n",
    "    protein_list = subunitNames[complex_name]\n",
    "    alpha = 0.05 / len(protein_list)\n",
    "    if include_borderline:\n",
    "        print('Borderline changes included; alpha set to 0.05\\n')\n",
    "        alpha = 0.05\n",
    "    \n",
    "    increased_all = []\n",
    "    decreased_all = []\n",
    "    no_change_all = []\n",
    "    mixed_results = []\n",
    "    missing_data = []\n",
    "    \n",
    "    proteomics_dict = get_pancancer_proteomics(protein_list)\n",
    "    \n",
    "    for protein in protein_list:\n",
    "        num_increased = 0\n",
    "        num_decreased = 0\n",
    "        num_no_change = 0\n",
    "        no_data = False\n",
    "        for key, val in proteomics_dict.items():\n",
    "            tumor = val.loc[val['Sample_Tumor_Normal'] == 'Tumor'][protein].dropna()\n",
    "            normal = val.loc[val['Sample_Tumor_Normal'] == 'Normal'][protein].dropna()\n",
    "            if len(tumor) > 2 and len(normal) > 2:\n",
    "                pval = scipy.stats.ttest_ind(tumor, normal)[1]\n",
    "                if pval <= alpha:\n",
    "                    if np.mean(tumor) > np.mean(normal):\n",
    "                        num_increased += 1\n",
    "                    else:\n",
    "                        num_decreased += 1\n",
    "                else:\n",
    "                    num_no_change += 1\n",
    "            else:\n",
    "                no_data = True\n",
    "                missing_data.append(protein)\n",
    "                break\n",
    "                \n",
    "        if not no_data:     \n",
    "            if num_decreased + num_no_change == 0:\n",
    "                increased_all.append(protein)\n",
    "\n",
    "            elif num_increased + num_no_change == 0:\n",
    "                decreased_all.append(protein)\n",
    "\n",
    "            elif num_decreased + num_increased == 0:\n",
    "                no_change_all.append(protein)\n",
    "\n",
    "            else:\n",
    "                mixed_results.append(protein)\n",
    "            \n",
    "    print('Increased in all cancers: ')\n",
    "    print(increased_all)\n",
    "    print('Decreased in all cancers: ')\n",
    "    print(decreased_all)\n",
    "    print('No change in all cancers: ')\n",
    "    print(no_change_all)\n",
    "    print('Mixed results: ')\n",
    "    print(mixed_results)\n",
    "    print('Missing proteomics data in at least one cancer: ')\n",
    "    print(missing_data)\n",
    "    \n",
    "    return {'increased': increased_all, 'decreased': decreased_all, 'no change': no_change_all, 'mixed': mixed_results, 'missing': missing_data}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pancancer_plot_proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pancancer_plot_proteins(complex_name = None, protein_list = None):\n",
    "    if complex_name:\n",
    "        protein_list = subunitNames[complex_name]\n",
    "    else:\n",
    "        complex_name = 'Unique Protein List'\n",
    "        \n",
    "    proteomics_dict = get_pancancer_proteomics(protein_list)\n",
    "    \n",
    "    for protein in protein_list:\n",
    "        plot_data = pd.DataFrame(columns = [protein, 'Sample_Tumor_Normal', 'Cancer_Type'])\n",
    "        for key, val in proteomics_dict.items():\n",
    "            cancer_data = val[[protein, 'Sample_Tumor_Normal']].copy()\n",
    "            cancer_data['Cancer_Type'] = key\n",
    "            plot_data = plot_data.append(cancer_data)\n",
    "            \n",
    "        a4_dims = (11.7, 8.27)\n",
    "        fig, ax = plt.subplots(figsize = a4_dims)\n",
    "        boxplt = sns.boxplot(data = plot_data, x = 'Cancer_Type', y = protein, hue = 'Sample_Tumor_Normal', showfliers = False)\n",
    "        boxplt = sns.stripplot(data = plot_data, x = 'Cancer_Type', y = protein, hue = 'Sample_Tumor_Normal', dodge = True, jitter = True, color = '.3')\n",
    "        boxplt.set_title(protein, fontsize = '20')\n",
    "        boxplt.set_ylabel('Proteomics', fontsize = '15')\n",
    "        boxplt.set_xlabel('Cancer Type', fontsize = '15')\n",
    "        handles, labels = ax.get_legend_handles_labels()\n",
    "        l = plt.legend(handles[0:2], labels[0:2])\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tidy_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Thanks StackOverflow! \n",
    "# https://stackoverflow.com/questions/12680754/split-explode-pandas-dataframe-string-entry-to-separate-rows\n",
    "\n",
    "def tidy_split(df, column, sep='|', keep=False):\n",
    "    \"\"\"\n",
    "    Split the values of a column and expand so the new DataFrame has one split\n",
    "    value per row. Filters rows where the column is missing.\n",
    "\n",
    "    Params\n",
    "    ------\n",
    "    df : pandas.DataFrame\n",
    "        dataframe with the column to split and expand\n",
    "    column : str\n",
    "        the column to split and expand\n",
    "    sep : str\n",
    "        the string used to split the column's values\n",
    "    keep : bool\n",
    "        whether to retain the presplit value as it's own row\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        Returns a dataframe with the same columns as `df`.\n",
    "    \"\"\"\n",
    "    indexes = list()\n",
    "    new_values = list()\n",
    "    #df = df.dropna(subset=[column])\n",
    "    for i, presplit in enumerate(df[column].astype(str)):\n",
    "        values = presplit.split(sep)\n",
    "        if keep and len(values) > 1:\n",
    "            indexes.append(i)\n",
    "            new_values.append(presplit)\n",
    "        for value in values:\n",
    "            indexes.append(i)\n",
    "            new_values.append(value)\n",
    "    new_df = df.iloc[indexes, :].copy()\n",
    "    new_df[column] = new_values\n",
    "    return new_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
