{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions and Imports for Ovarian Cancer Ratio Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports \n",
    "This includes transcriptional regulatory interactions from TRRUST v2 (https://www.grnpedia.org/trrust/downloadnetwork.php)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "import seaborn as sns\n",
    "import cptac\n",
    "import cptac.algorithms as al\n",
    "tf_data = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/ProteinComplexes/Data_Files/trrust_rawdata.human.tsv', sep='\\t', header=None)\n",
    "tf_data.columns = ['TF', 'Regulated_Gene', 'Regulation_Type', 'ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                        0        1    \\\n",
      "ComplexName                                                            \n",
      "BCL6-HDAC4 complex                                     BCL6    HDAC4   \n",
      "BCL6-HDAC5 complex                                     BCL6    HDAC5   \n",
      "BCL6-HDAC7 complex                                     BCL6    HDAC7   \n",
      "Multisubunit ACTR coactivator complex                 EP300   CREBBP   \n",
      "Condensin I complex                                    SMC2    NCAPH   \n",
      "BLOC-3 (biogenesis of lysosome-related organell...     HPS1     HPS4   \n",
      "BLOC-2 (biogenesis of lysosome-related organell...     HPS6     HPS3   \n",
      "MUS81-CDS1 complex                                     CDS1    MUS81   \n",
      "NCOR complex                                          HDAC3    TBL1X   \n",
      "BLOC-1 (biogenesis of lysosome-related organell...   SNAPIN  BLOC1S1   \n",
      "Arp2/3 protein complex                               ARPC1B    ARPC2   \n",
      "PA28gamma complex                                     PSME3     None   \n",
      "PA28 complex                                          PSME1    PSME2   \n",
      "PA700 complex                                        PSMD11   PSMD12   \n",
      "Prefoldin complex                                     PFDN6    PFDN1   \n",
      "AP1 adaptor complex                                   AP1G1    AP1G2   \n",
      "Mi-2/NuRD-MTA3 complex                                 MTA2     MBD3   \n",
      "Gamma-secretase complex (APH1A, PSEN1, PSENEN, ...    PSEN1    NCSTN   \n",
      "DNMT3B complex                                      SMARCA5    KIF4A   \n",
      "RAB5-RABAPTIN5\\u2013RABEX5 complex                    RAB5A   RABEP1   \n",
      "SIN3 complex                                          SAP18    SAP30   \n",
      "HDAC4-ERK1 complex                                    MAPK3    HDAC4   \n",
      "HDAC4-ERK2 complex                                    MAPK1    HDAC4   \n",
      "SMRT complex                                          HDAC3    TBL1X   \n",
      "AP3 adaptor complex                                   AP3B1    AP3D1   \n",
      "Interferon-stimulated gene factor 3 transcripti...    STAT1    STAT2   \n",
      "Mi2/NuRD complex                                       MTA2     MBD3   \n",
      "MeCP1 complex                                          MTA2     MBD3   \n",
      "Cohesin-SA1 complex                                   RAD21    SMC1A   \n",
      "Cohesin-SA2 complex                                   RAD21    SMC1A   \n",
      "...                                                     ...      ...   \n",
      "BORC complex                                         SNAPIN  BLOC1S1   \n",
      "MAPK6-MAPKAPK5-SEPT7 complex                          SEPT7    MAPK6   \n",
      "KEOPS complex                                         LAGE3   TP53RK   \n",
      "PRAME complex                                          RBX1    PRAME   \n",
      "LGR4-RSPO supercomplex                                 WASL     DVL2   \n",
      "INSC-LGN-PAR3 complex                                 GPSM2     INSC   \n",
      "NCKAP1-WASF3 complex                                  WASF3   NCKAP1   \n",
      "NCKAP1-WASF3-RAC1 complex                              RAC1    WASF3   \n",
      "LARP1-PABPC1-RYDEN complex                           PABPC1    LARP1   \n",
      "LGR5-RNF43-RSPO1 complex                               LGR5    RSPO1   \n",
      "7SK RNP complex                                       CCNT1   HEXIM1   \n",
      "SUMO1-SYT7 complex                                     SYT7    SUMO1   \n",
      "LATS2-SIAH2-ZYX complex                               SIAH2      ZYX   \n",
      "APP-PSEN1 complex                                       APP    PSEN1   \n",
      "APP-PSEN2 complex                                       APP    PSEN2   \n",
      "VHL-ElonginB-ElonginC complex                           VHL    TCEB1   \n",
      "APP-APBB1-KAT5 complex                                APBB1      APP   \n",
      "Shieldin complex                                      SHLD3    SHLD2   \n",
      "APP(AICD)-MED12 complex                                 APP    MED12   \n",
      "MAP2K4-MAP3K3-NBR1 complex                           MAP2K4     NBR1   \n",
      "ESR1-TRAP/Mediator coactivator-complex                 MED7    MED14   \n",
      "ESR2-TRAP/Mediator coactivator complex                 MED7    MED14   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12, MED13)             CCNC     CDK8   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12)                    CCNC     CDK8   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED13)                    CCNC     CDK8   \n",
      "FOXO1-ESR1 complex                                     ESR1    FOXO1   \n",
      "APP(AICD)-FOXO3 complex                               FOXO3      APP   \n",
      "APP(AICD)-FOXO1 complex                                 APP    FOXO1   \n",
      "APP(AICD)-FOXO4 complex                                 APP    FOXO4   \n",
      "AKAP13-MAP2K3-MAP3K20-MAPK14-PKN1 complex            MAP2K3   AKAP13   \n",
      "\n",
      "                                                         2        3    \\\n",
      "ComplexName                                                             \n",
      "BCL6-HDAC4 complex                                      None     None   \n",
      "BCL6-HDAC5 complex                                      None     None   \n",
      "BCL6-HDAC7 complex                                      None     None   \n",
      "Multisubunit ACTR coactivator complex                  KAT2B    NCOA3   \n",
      "Condensin I complex                                   NCAPD2    NCAPG   \n",
      "BLOC-3 (biogenesis of lysosome-related organell...      None     None   \n",
      "BLOC-2 (biogenesis of lysosome-related organell...      HPS5     None   \n",
      "MUS81-CDS1 complex                                      None     None   \n",
      "NCOR complex                                           NCOR1     GPS2   \n",
      "BLOC-1 (biogenesis of lysosome-related organell...   BLOC1S3  BLOC1S2   \n",
      "Arp2/3 protein complex                                 ARPC3    ARPC5   \n",
      "PA28gamma complex                                       None     None   \n",
      "PA28 complex                                            None     None   \n",
      "PA700 complex                                          PSMD9   PSMD14   \n",
      "Prefoldin complex                                       VBP1    PFDN5   \n",
      "AP1 adaptor complex                                    AP1S2    AP1S1   \n",
      "Mi-2/NuRD-MTA3 complex                                  BCL6    HDAC1   \n",
      "Gamma-secretase complex (APH1A, PSEN1, PSENEN, ...     APH1A   PSENEN   \n",
      "DNMT3B complex                                          SMC2    HDAC1   \n",
      "RAB5-RABAPTIN5\\u2013RABEX5 complex                   RABGEF1     None   \n",
      "SIN3 complex                                           RBBP4    HDAC1   \n",
      "HDAC4-ERK1 complex                                      None     None   \n",
      "HDAC4-ERK2 complex                                      None     None   \n",
      "SMRT complex                                            GPS2  TBL1XR1   \n",
      "AP3 adaptor complex                                    AP3M2    AP3S2   \n",
      "Interferon-stimulated gene factor 3 transcripti...      IRF9     None   \n",
      "Mi2/NuRD complex                                       RBBP4    HDAC1   \n",
      "MeCP1 complex                                          RBBP4    HDAC1   \n",
      "Cohesin-SA1 complex                                    STAG1     SMC3   \n",
      "Cohesin-SA2 complex                                    STAG2     SMC3   \n",
      "...                                                      ...      ...   \n",
      "BORC complex                                         BLOC1S2   BORCS5   \n",
      "MAPK6-MAPKAPK5-SEPT7 complex                        MAPKAPK5     None   \n",
      "KEOPS complex                                          OSGEP    TPRKB   \n",
      "PRAME complex                                           CUL2    LAGE3   \n",
      "LGR4-RSPO supercomplex                                DIAPH1   MAP2K2   \n",
      "INSC-LGN-PAR3 complex                                  PARD3     None   \n",
      "NCKAP1-WASF3 complex                                    None     None   \n",
      "NCKAP1-WASF3-RAC1 complex                             NCKAP1     None   \n",
      "LARP1-PABPC1-RYDEN complex                             RYDEN     None   \n",
      "LGR5-RNF43-RSPO1 complex                               RNF43     None   \n",
      "7SK RNP complex                                         CDK9    LARP7   \n",
      "SUMO1-SYT7 complex                                      None     None   \n",
      "LATS2-SIAH2-ZYX complex                                LATS2     None   \n",
      "APP-PSEN1 complex                                       None     None   \n",
      "APP-PSEN2 complex                                       None     None   \n",
      "VHL-ElonginB-ElonginC complex                          TCEB2     None   \n",
      "APP-APBB1-KAT5 complex                                  KAT5     None   \n",
      "Shieldin complex                                       SHLD1   MAD2L2   \n",
      "APP(AICD)-MED12 complex                                 None     None   \n",
      "MAP2K4-MAP3K3-NBR1 complex                            MAP3K3     None   \n",
      "ESR1-TRAP/Mediator coactivator-complex                 MED24     MED6   \n",
      "ESR2-TRAP/Mediator coactivator complex                 MED24     MED6   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12, MED13)             MED12    MED13   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12)                    MED12     None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED13)                    MED13     None   \n",
      "FOXO1-ESR1 complex                                      None     None   \n",
      "APP(AICD)-FOXO3 complex                                 None     None   \n",
      "APP(AICD)-FOXO1 complex                                 None     None   \n",
      "APP(AICD)-FOXO4 complex                                 None     None   \n",
      "AKAP13-MAP2K3-MAP3K20-MAPK14-PKN1 complex               PKN1   MAPK14   \n",
      "\n",
      "                                                        4        5        6    \\\n",
      "ComplexName                                                                     \n",
      "BCL6-HDAC4 complex                                     None     None     None   \n",
      "BCL6-HDAC5 complex                                     None     None     None   \n",
      "BCL6-HDAC7 complex                                     None     None     None   \n",
      "Multisubunit ACTR coactivator complex                  None     None     None   \n",
      "Condensin I complex                                    SMC4     None     None   \n",
      "BLOC-3 (biogenesis of lysosome-related organell...     None     None     None   \n",
      "BLOC-2 (biogenesis of lysosome-related organell...     None     None     None   \n",
      "MUS81-CDS1 complex                                     None     None     None   \n",
      "NCOR complex                                         CORO2A  TBL1XR1     None   \n",
      "BLOC-1 (biogenesis of lysosome-related organell...  BLOC1S5   DTNBP1  BLOC1S4   \n",
      "Arp2/3 protein complex                                ARPC4    ACTR3    ACTR2   \n",
      "PA28gamma complex                                      None     None     None   \n",
      "PA28 complex                                           None     None     None   \n",
      "PA700 complex                                         PSMD3   PSMD10    PSMC3   \n",
      "Prefoldin complex                                     PFDN4    PFDN2     None   \n",
      "AP1 adaptor complex                                   AP1B1    AP1S3    AP1M1   \n",
      "Mi-2/NuRD-MTA3 complex                                 MTA3     None     None   \n",
      "Gamma-secretase complex (APH1A, PSEN1, PSENEN, ...     None     None     None   \n",
      "DNMT3B complex                                        SIN3A     SMC4   DNMT3B   \n",
      "RAB5-RABAPTIN5\\u2013RABEX5 complex                     None     None     None   \n",
      "SIN3 complex                                          RBBP7    HDAC2    SIN3A   \n",
      "HDAC4-ERK1 complex                                     None     None     None   \n",
      "HDAC4-ERK2 complex                                     None     None     None   \n",
      "SMRT complex                                          NCOR2     None     None   \n",
      "AP3 adaptor complex                                   AP3B2    AP3S1    AP3M1   \n",
      "Interferon-stimulated gene factor 3 transcripti...     None     None     None   \n",
      "Mi2/NuRD complex                                       CHD4    RBBP7    HDAC2   \n",
      "MeCP1 complex                                          CHD4    RBBP7    HDAC2   \n",
      "Cohesin-SA1 complex                                    None     None     None   \n",
      "Cohesin-SA2 complex                                    None     None     None   \n",
      "...                                                     ...      ...      ...   \n",
      "BORC complex                                         BORCS7   BORCS8   BORCS6   \n",
      "MAPK6-MAPKAPK5-SEPT7 complex                           None     None     None   \n",
      "KEOPS complex                                          None     None     None   \n",
      "PRAME complex                                         TCEB1    TCEB2    OSGEP   \n",
      "LGR4-RSPO supercomplex                               IQGAP1      PXN   MAP2K1   \n",
      "INSC-LGN-PAR3 complex                                  None     None     None   \n",
      "NCKAP1-WASF3 complex                                   None     None     None   \n",
      "NCKAP1-WASF3-RAC1 complex                              None     None     None   \n",
      "LARP1-PABPC1-RYDEN complex                             None     None     None   \n",
      "LGR5-RNF43-RSPO1 complex                               None     None     None   \n",
      "7SK RNP complex                                        None     None     None   \n",
      "SUMO1-SYT7 complex                                     None     None     None   \n",
      "LATS2-SIAH2-ZYX complex                                None     None     None   \n",
      "APP-PSEN1 complex                                      None     None     None   \n",
      "APP-PSEN2 complex                                      None     None     None   \n",
      "VHL-ElonginB-ElonginC complex                          None     None     None   \n",
      "APP-APBB1-KAT5 complex                                 None     None     None   \n",
      "Shieldin complex                                       None     None     None   \n",
      "APP(AICD)-MED12 complex                                None     None     None   \n",
      "MAP2K4-MAP3K3-NBR1 complex                             None     None     None   \n",
      "ESR1-TRAP/Mediator coactivator-complex                 ESR1     CCNC     CDK8   \n",
      "ESR2-TRAP/Mediator coactivator complex                 CCNC     CDK8    EP300   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12, MED13)             None     None     None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12)                    None     None     None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED13)                    None     None     None   \n",
      "FOXO1-ESR1 complex                                     None     None     None   \n",
      "APP(AICD)-FOXO3 complex                                None     None     None   \n",
      "APP(AICD)-FOXO1 complex                                None     None     None   \n",
      "APP(AICD)-FOXO4 complex                                None     None     None   \n",
      "AKAP13-MAP2K3-MAP3K20-MAPK14-PKN1 complex           MAP3K20     None     None   \n",
      "\n",
      "                                                        7      8      9    \\\n",
      "ComplexName                                                                 \n",
      "BCL6-HDAC4 complex                                     None   None   None   \n",
      "BCL6-HDAC5 complex                                     None   None   None   \n",
      "BCL6-HDAC7 complex                                     None   None   None   \n",
      "Multisubunit ACTR coactivator complex                  None   None   None   \n",
      "Condensin I complex                                    None   None   None   \n",
      "BLOC-3 (biogenesis of lysosome-related organell...     None   None   None   \n",
      "BLOC-2 (biogenesis of lysosome-related organell...     None   None   None   \n",
      "MUS81-CDS1 complex                                     None   None   None   \n",
      "NCOR complex                                           None   None   None   \n",
      "BLOC-1 (biogenesis of lysosome-related organell...  BLOC1S6   None   None   \n",
      "Arp2/3 protein complex                                 None   None   None   \n",
      "PA28gamma complex                                      None   None   None   \n",
      "PA28 complex                                           None   None   None   \n",
      "PA700 complex                                         PSMC2  PSMC4  PSMD8   \n",
      "Prefoldin complex                                      None   None   None   \n",
      "AP1 adaptor complex                                   AP1M2   None   None   \n",
      "Mi-2/NuRD-MTA3 complex                                 None   None   None   \n",
      "Gamma-secretase complex (APH1A, PSEN1, PSENEN, ...     None   None   None   \n",
      "DNMT3B complex                                         None   None   None   \n",
      "RAB5-RABAPTIN5\\u2013RABEX5 complex                     None   None   None   \n",
      "SIN3 complex                                           None   None   None   \n",
      "HDAC4-ERK1 complex                                     None   None   None   \n",
      "HDAC4-ERK2 complex                                     None   None   None   \n",
      "SMRT complex                                           None   None   None   \n",
      "AP3 adaptor complex                                    None   None   None   \n",
      "Interferon-stimulated gene factor 3 transcripti...     None   None   None   \n",
      "Mi2/NuRD complex                                       None   None   None   \n",
      "MeCP1 complex                                          MBD2   None   None   \n",
      "Cohesin-SA1 complex                                    None   None   None   \n",
      "Cohesin-SA2 complex                                    None   None   None   \n",
      "...                                                     ...    ...    ...   \n",
      "BORC complex                                           KXD1   None   None   \n",
      "MAPK6-MAPKAPK5-SEPT7 complex                           None   None   None   \n",
      "KEOPS complex                                          None   None   None   \n",
      "PRAME complex                                          None   None   None   \n",
      "LGR4-RSPO supercomplex                                 PTK2   LGR4  RSPO3   \n",
      "INSC-LGN-PAR3 complex                                  None   None   None   \n",
      "NCKAP1-WASF3 complex                                   None   None   None   \n",
      "NCKAP1-WASF3-RAC1 complex                              None   None   None   \n",
      "LARP1-PABPC1-RYDEN complex                             None   None   None   \n",
      "LGR5-RNF43-RSPO1 complex                               None   None   None   \n",
      "7SK RNP complex                                        None   None   None   \n",
      "SUMO1-SYT7 complex                                     None   None   None   \n",
      "LATS2-SIAH2-ZYX complex                                None   None   None   \n",
      "APP-PSEN1 complex                                      None   None   None   \n",
      "APP-PSEN2 complex                                      None   None   None   \n",
      "VHL-ElonginB-ElonginC complex                          None   None   None   \n",
      "APP-APBB1-KAT5 complex                                 None   None   None   \n",
      "Shieldin complex                                       None   None   None   \n",
      "APP(AICD)-MED12 complex                                None   None   None   \n",
      "MAP2K4-MAP3K3-NBR1 complex                             None   None   None   \n",
      "ESR1-TRAP/Mediator coactivator-complex                EP300  MED21   MED1   \n",
      "ESR2-TRAP/Mediator coactivator complex                MED21   MED1  NCOA1   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12, MED13)             None   None   None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12)                    None   None   None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED13)                    None   None   None   \n",
      "FOXO1-ESR1 complex                                     None   None   None   \n",
      "APP(AICD)-FOXO3 complex                                None   None   None   \n",
      "APP(AICD)-FOXO1 complex                                None   None   None   \n",
      "APP(AICD)-FOXO4 complex                                None   None   None   \n",
      "AKAP13-MAP2K3-MAP3K20-MAPK14-PKN1 complex              None   None   None   \n",
      "\n",
      "                                                    ...   133   134   135  \\\n",
      "ComplexName                                         ...                     \n",
      "BCL6-HDAC4 complex                                  ...  None  None  None   \n",
      "BCL6-HDAC5 complex                                  ...  None  None  None   \n",
      "BCL6-HDAC7 complex                                  ...  None  None  None   \n",
      "Multisubunit ACTR coactivator complex               ...  None  None  None   \n",
      "Condensin I complex                                 ...  None  None  None   \n",
      "BLOC-3 (biogenesis of lysosome-related organell...  ...  None  None  None   \n",
      "BLOC-2 (biogenesis of lysosome-related organell...  ...  None  None  None   \n",
      "MUS81-CDS1 complex                                  ...  None  None  None   \n",
      "NCOR complex                                        ...  None  None  None   \n",
      "BLOC-1 (biogenesis of lysosome-related organell...  ...  None  None  None   \n",
      "Arp2/3 protein complex                              ...  None  None  None   \n",
      "PA28gamma complex                                   ...  None  None  None   \n",
      "PA28 complex                                        ...  None  None  None   \n",
      "PA700 complex                                       ...  None  None  None   \n",
      "Prefoldin complex                                   ...  None  None  None   \n",
      "AP1 adaptor complex                                 ...  None  None  None   \n",
      "Mi-2/NuRD-MTA3 complex                              ...  None  None  None   \n",
      "Gamma-secretase complex (APH1A, PSEN1, PSENEN, ...  ...  None  None  None   \n",
      "DNMT3B complex                                      ...  None  None  None   \n",
      "RAB5-RABAPTIN5\\u2013RABEX5 complex                  ...  None  None  None   \n",
      "SIN3 complex                                        ...  None  None  None   \n",
      "HDAC4-ERK1 complex                                  ...  None  None  None   \n",
      "HDAC4-ERK2 complex                                  ...  None  None  None   \n",
      "SMRT complex                                        ...  None  None  None   \n",
      "AP3 adaptor complex                                 ...  None  None  None   \n",
      "Interferon-stimulated gene factor 3 transcripti...  ...  None  None  None   \n",
      "Mi2/NuRD complex                                    ...  None  None  None   \n",
      "MeCP1 complex                                       ...  None  None  None   \n",
      "Cohesin-SA1 complex                                 ...  None  None  None   \n",
      "Cohesin-SA2 complex                                 ...  None  None  None   \n",
      "...                                                 ...   ...   ...   ...   \n",
      "BORC complex                                        ...  None  None  None   \n",
      "MAPK6-MAPKAPK5-SEPT7 complex                        ...  None  None  None   \n",
      "KEOPS complex                                       ...  None  None  None   \n",
      "PRAME complex                                       ...  None  None  None   \n",
      "LGR4-RSPO supercomplex                              ...  None  None  None   \n",
      "INSC-LGN-PAR3 complex                               ...  None  None  None   \n",
      "NCKAP1-WASF3 complex                                ...  None  None  None   \n",
      "NCKAP1-WASF3-RAC1 complex                           ...  None  None  None   \n",
      "LARP1-PABPC1-RYDEN complex                          ...  None  None  None   \n",
      "LGR5-RNF43-RSPO1 complex                            ...  None  None  None   \n",
      "7SK RNP complex                                     ...  None  None  None   \n",
      "SUMO1-SYT7 complex                                  ...  None  None  None   \n",
      "LATS2-SIAH2-ZYX complex                             ...  None  None  None   \n",
      "APP-PSEN1 complex                                   ...  None  None  None   \n",
      "APP-PSEN2 complex                                   ...  None  None  None   \n",
      "VHL-ElonginB-ElonginC complex                       ...  None  None  None   \n",
      "APP-APBB1-KAT5 complex                              ...  None  None  None   \n",
      "Shieldin complex                                    ...  None  None  None   \n",
      "APP(AICD)-MED12 complex                             ...  None  None  None   \n",
      "MAP2K4-MAP3K3-NBR1 complex                          ...  None  None  None   \n",
      "ESR1-TRAP/Mediator coactivator-complex              ...  None  None  None   \n",
      "ESR2-TRAP/Mediator coactivator complex              ...  None  None  None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12, MED13)          ...  None  None  None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12)                 ...  None  None  None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED13)                 ...  None  None  None   \n",
      "FOXO1-ESR1 complex                                  ...  None  None  None   \n",
      "APP(AICD)-FOXO3 complex                             ...  None  None  None   \n",
      "APP(AICD)-FOXO1 complex                             ...  None  None  None   \n",
      "APP(AICD)-FOXO4 complex                             ...  None  None  None   \n",
      "AKAP13-MAP2K3-MAP3K20-MAPK14-PKN1 complex           ...  None  None  None   \n",
      "\n",
      "                                                     136   137   138   139  \\\n",
      "ComplexName                                                                  \n",
      "BCL6-HDAC4 complex                                  None  None  None  None   \n",
      "BCL6-HDAC5 complex                                  None  None  None  None   \n",
      "BCL6-HDAC7 complex                                  None  None  None  None   \n",
      "Multisubunit ACTR coactivator complex               None  None  None  None   \n",
      "Condensin I complex                                 None  None  None  None   \n",
      "BLOC-3 (biogenesis of lysosome-related organell...  None  None  None  None   \n",
      "BLOC-2 (biogenesis of lysosome-related organell...  None  None  None  None   \n",
      "MUS81-CDS1 complex                                  None  None  None  None   \n",
      "NCOR complex                                        None  None  None  None   \n",
      "BLOC-1 (biogenesis of lysosome-related organell...  None  None  None  None   \n",
      "Arp2/3 protein complex                              None  None  None  None   \n",
      "PA28gamma complex                                   None  None  None  None   \n",
      "PA28 complex                                        None  None  None  None   \n",
      "PA700 complex                                       None  None  None  None   \n",
      "Prefoldin complex                                   None  None  None  None   \n",
      "AP1 adaptor complex                                 None  None  None  None   \n",
      "Mi-2/NuRD-MTA3 complex                              None  None  None  None   \n",
      "Gamma-secretase complex (APH1A, PSEN1, PSENEN, ...  None  None  None  None   \n",
      "DNMT3B complex                                      None  None  None  None   \n",
      "RAB5-RABAPTIN5\\u2013RABEX5 complex                  None  None  None  None   \n",
      "SIN3 complex                                        None  None  None  None   \n",
      "HDAC4-ERK1 complex                                  None  None  None  None   \n",
      "HDAC4-ERK2 complex                                  None  None  None  None   \n",
      "SMRT complex                                        None  None  None  None   \n",
      "AP3 adaptor complex                                 None  None  None  None   \n",
      "Interferon-stimulated gene factor 3 transcripti...  None  None  None  None   \n",
      "Mi2/NuRD complex                                    None  None  None  None   \n",
      "MeCP1 complex                                       None  None  None  None   \n",
      "Cohesin-SA1 complex                                 None  None  None  None   \n",
      "Cohesin-SA2 complex                                 None  None  None  None   \n",
      "...                                                  ...   ...   ...   ...   \n",
      "BORC complex                                        None  None  None  None   \n",
      "MAPK6-MAPKAPK5-SEPT7 complex                        None  None  None  None   \n",
      "KEOPS complex                                       None  None  None  None   \n",
      "PRAME complex                                       None  None  None  None   \n",
      "LGR4-RSPO supercomplex                              None  None  None  None   \n",
      "INSC-LGN-PAR3 complex                               None  None  None  None   \n",
      "NCKAP1-WASF3 complex                                None  None  None  None   \n",
      "NCKAP1-WASF3-RAC1 complex                           None  None  None  None   \n",
      "LARP1-PABPC1-RYDEN complex                          None  None  None  None   \n",
      "LGR5-RNF43-RSPO1 complex                            None  None  None  None   \n",
      "7SK RNP complex                                     None  None  None  None   \n",
      "SUMO1-SYT7 complex                                  None  None  None  None   \n",
      "LATS2-SIAH2-ZYX complex                             None  None  None  None   \n",
      "APP-PSEN1 complex                                   None  None  None  None   \n",
      "APP-PSEN2 complex                                   None  None  None  None   \n",
      "VHL-ElonginB-ElonginC complex                       None  None  None  None   \n",
      "APP-APBB1-KAT5 complex                              None  None  None  None   \n",
      "Shieldin complex                                    None  None  None  None   \n",
      "APP(AICD)-MED12 complex                             None  None  None  None   \n",
      "MAP2K4-MAP3K3-NBR1 complex                          None  None  None  None   \n",
      "ESR1-TRAP/Mediator coactivator-complex              None  None  None  None   \n",
      "ESR2-TRAP/Mediator coactivator complex              None  None  None  None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12, MED13)          None  None  None  None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12)                 None  None  None  None   \n",
      "CDK8 subcomplex (CCNC, CDK8, MED13)                 None  None  None  None   \n",
      "FOXO1-ESR1 complex                                  None  None  None  None   \n",
      "APP(AICD)-FOXO3 complex                             None  None  None  None   \n",
      "APP(AICD)-FOXO1 complex                             None  None  None  None   \n",
      "APP(AICD)-FOXO4 complex                             None  None  None  None   \n",
      "AKAP13-MAP2K3-MAP3K20-MAPK14-PKN1 complex           None  None  None  None   \n",
      "\n",
      "                                                     140   141   142  \n",
      "ComplexName                                                           \n",
      "BCL6-HDAC4 complex                                  None  None  None  \n",
      "BCL6-HDAC5 complex                                  None  None  None  \n",
      "BCL6-HDAC7 complex                                  None  None  None  \n",
      "Multisubunit ACTR coactivator complex               None  None  None  \n",
      "Condensin I complex                                 None  None  None  \n",
      "BLOC-3 (biogenesis of lysosome-related organell...  None  None  None  \n",
      "BLOC-2 (biogenesis of lysosome-related organell...  None  None  None  \n",
      "MUS81-CDS1 complex                                  None  None  None  \n",
      "NCOR complex                                        None  None  None  \n",
      "BLOC-1 (biogenesis of lysosome-related organell...  None  None  None  \n",
      "Arp2/3 protein complex                              None  None  None  \n",
      "PA28gamma complex                                   None  None  None  \n",
      "PA28 complex                                        None  None  None  \n",
      "PA700 complex                                       None  None  None  \n",
      "Prefoldin complex                                   None  None  None  \n",
      "AP1 adaptor complex                                 None  None  None  \n",
      "Mi-2/NuRD-MTA3 complex                              None  None  None  \n",
      "Gamma-secretase complex (APH1A, PSEN1, PSENEN, ...  None  None  None  \n",
      "DNMT3B complex                                      None  None  None  \n",
      "RAB5-RABAPTIN5\\u2013RABEX5 complex                  None  None  None  \n",
      "SIN3 complex                                        None  None  None  \n",
      "HDAC4-ERK1 complex                                  None  None  None  \n",
      "HDAC4-ERK2 complex                                  None  None  None  \n",
      "SMRT complex                                        None  None  None  \n",
      "AP3 adaptor complex                                 None  None  None  \n",
      "Interferon-stimulated gene factor 3 transcripti...  None  None  None  \n",
      "Mi2/NuRD complex                                    None  None  None  \n",
      "MeCP1 complex                                       None  None  None  \n",
      "Cohesin-SA1 complex                                 None  None  None  \n",
      "Cohesin-SA2 complex                                 None  None  None  \n",
      "...                                                  ...   ...   ...  \n",
      "BORC complex                                        None  None  None  \n",
      "MAPK6-MAPKAPK5-SEPT7 complex                        None  None  None  \n",
      "KEOPS complex                                       None  None  None  \n",
      "PRAME complex                                       None  None  None  \n",
      "LGR4-RSPO supercomplex                              None  None  None  \n",
      "INSC-LGN-PAR3 complex                               None  None  None  \n",
      "NCKAP1-WASF3 complex                                None  None  None  \n",
      "NCKAP1-WASF3-RAC1 complex                           None  None  None  \n",
      "LARP1-PABPC1-RYDEN complex                          None  None  None  \n",
      "LGR5-RNF43-RSPO1 complex                            None  None  None  \n",
      "7SK RNP complex                                     None  None  None  \n",
      "SUMO1-SYT7 complex                                  None  None  None  \n",
      "LATS2-SIAH2-ZYX complex                             None  None  None  \n",
      "APP-PSEN1 complex                                   None  None  None  \n",
      "APP-PSEN2 complex                                   None  None  None  \n",
      "VHL-ElonginB-ElonginC complex                       None  None  None  \n",
      "APP-APBB1-KAT5 complex                              None  None  None  \n",
      "Shieldin complex                                    None  None  None  \n",
      "APP(AICD)-MED12 complex                             None  None  None  \n",
      "MAP2K4-MAP3K3-NBR1 complex                          None  None  None  \n",
      "ESR1-TRAP/Mediator coactivator-complex              None  None  None  \n",
      "ESR2-TRAP/Mediator coactivator complex              None  None  None  \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12, MED13)          None  None  None  \n",
      "CDK8 subcomplex (CCNC, CDK8, MED12)                 None  None  None  \n",
      "CDK8 subcomplex (CCNC, CDK8, MED13)                 None  None  None  \n",
      "FOXO1-ESR1 complex                                  None  None  None  \n",
      "APP(AICD)-FOXO3 complex                             None  None  None  \n",
      "APP(AICD)-FOXO1 complex                             None  None  None  \n",
      "APP(AICD)-FOXO4 complex                             None  None  None  \n",
      "AKAP13-MAP2K3-MAP3K20-MAPK14-PKN1 complex           None  None  None  \n",
      "\n",
      "[2700 rows x 143 columns]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All files already downloaded and correct.\n",
      "All files already downloaded and correct.\n",
      "All files already downloaded and correct.\n",
      "All files already downloaded and correct.\n",
      "Loading dataframes..                \r"
     ]
    }
   ],
   "source": [
    "# Create a workable dictionary of protein complex information (all cancer types)\n",
    "# Select for human complexes\n",
    "complexData = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/ProteinComplexes/Data_Files/allComplexes.txt', sep='\\t')\n",
    "complexData = complexData.loc[complexData['Organism'] == 'Human']\n",
    "\n",
    "# Split the proteins in each complex into invdividual columns of a new df\n",
    "subunitNames = complexData['subunits(Gene name)'].str.split(';', expand = True)\n",
    "subunitNames.index = complexData.ComplexName\n",
    "\n",
    "# Create a dictionary (key = complex name, value = list of proteins in complex) and remove None values\n",
    "subunitNames = subunitNames.loc[~subunitNames.index.duplicated()]\n",
    "subunitNames = subunitNames.transpose().to_dict('list')\n",
    "for key, val in subunitNames.items():\n",
    "    subunitNames[key] = [value for value in val if value != None]\n",
    "\n",
    "# Ovarian Data\n",
    "\n",
    "cptac.download(dataset='ovarian', version='latest')\n",
    "ov = cptac.Ovarian()\n",
    "ov_clinical = ov.get_clinical()\n",
    "ov_mutations = ov.get_mutations()\n",
    "ov_mutation_data = ov_clinical[['Patient_ID', 'Sample_Tumor_Normal']].join(ov_mutations).dropna(axis = 0)\n",
    "ov_mutation_data = ov_mutation_data.loc[ov_mutation_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "#Only for ovarian: Import non-normalized proteomics data for ratio analysis\n",
    "data = pd.read_csv('/Users/amandaoliphant/Documents/GitHub/ProteinComplexes/Data_Files/proteinGroups_cleaned.txt', sep='\\t')\n",
    "\n",
    "# Remove duplicate rows from the dataset\n",
    "data = data.set_index('Gene_Name')\n",
    "data = data.loc[~data.index.duplicated()]\n",
    "\n",
    "# Endometrial Data\n",
    "cptac.download(dataset='endometrial', version='latest')\n",
    "en = cptac.Endometrial()\n",
    "en_clinical = en.get_clinical()\n",
    "en_mutations = en.get_mutations()\n",
    "en_mutation_data = en_clinical[['Patient_ID', 'Proteomics_Tumor_Normal']].join(en_mutations).dropna(axis = 0)\n",
    "en_mutation_data = en_mutation_data.loc[en_mutation_data['Proteomics_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "# Colon Data\n",
    "cptac.download(dataset='colon', version='latest')\n",
    "colon = cptac.Colon()\n",
    "colon_clinical = colon.get_clinical()\n",
    "colon_mutations = colon.get_mutations()\n",
    "colon_mutation_data = colon_clinical[['Patient_ID', 'Sample_Tumor_Normal']].join(colon_mutations).dropna(axis = 0)\n",
    "colon_mutation_data = colon_mutation_data.loc[colon_mutation_data['Sample_Tumor_Normal'] == 'Tumor']\n",
    "\n",
    "# Renal Data\n",
    "cptac.download(dataset='renalccrcc', version='latest')\n",
    "renal = cptac.RenalCcrcc()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define basic analysis and plotting functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get_ratio_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "get_ratio_df\n",
    "-----------\n",
    "All patients, single protein, tumor vs normal\n",
    "Returns two dataframes: tumor and normal containing ratios for the proteins\n",
    "\n",
    "Parameters:\n",
    "    prot1, prot2 = (gene) names of the two proteins to be compared (ratio of prot1 / prot2)\n",
    "    \n",
    "Returns:\n",
    "    Two dataframes: tumor and normal containing ratios for the proteins\n",
    "  \n",
    "\"\"\"\n",
    "def get_ratio_df(prot1, prot2):\n",
    "    # Make sure that both proteins are in our dataset\n",
    "    if not data.index.contains(prot1) or not data.index.contains(prot2): return None, None\n",
    "    \n",
    "    # Create a dataframe of ratios for easy testing\n",
    "    tumor_rows_list = []\n",
    "    normal_rows_list = []\n",
    "    for patient in data.columns:\n",
    "        \n",
    "        # Determine the sample type\n",
    "        sample_type = 'Tumor'\n",
    "        if '_NM' in patient: sample_type = 'Normal'\n",
    "            \n",
    "        # Find the ratio prot1 / prot2 and create a new row for the dataframe\n",
    "        if data.at[prot2, patient] == 0: \n",
    "            ratio = np.nan\n",
    "        else: \n",
    "            ratio = data.at[prot1, patient] / data.at[prot2, patient]\n",
    "        # Create a row entry for the dictionary\n",
    "        row_dict = {'Ratio': ratio, 'Sample_Type': sample_type, 'Patient_ID': patient}\n",
    "        \n",
    "        # Add the new row to the tumor or normal list, depending on the sample type\n",
    "        if sample_type == 'Tumor':\n",
    "            tumor_rows_list.append(row_dict)\n",
    "        else:\n",
    "            normal_rows_list.append(row_dict)\n",
    "    \n",
    "    # Convert the row lists into dataframes\n",
    "    tumor_ratio_df = pd.DataFrame(tumor_rows_list)\n",
    "    normal_ratio_df = pd.DataFrame(normal_rows_list)\n",
    "    \n",
    "    # If there were no valid ratios for either the tumor or normal samples, return None\n",
    "    if (not np.isfinite(tumor_ratio_df['Ratio']).any()) or (not np.isfinite(normal_ratio_df['Ratio']).any()): \n",
    "        return None, None\n",
    "    \n",
    "    # Drop na values\n",
    "    tumor_ratio_df = tumor_ratio_df.dropna(axis = 0)\n",
    "    normal_ratio_df = normal_ratio_df.dropna(axis = 0)\n",
    "    \n",
    "    return tumor_ratio_df, normal_ratio_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test_complex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "test_complex\n",
    "------------\n",
    "Perform statistical tests on every combination of proteins in a given complex, printing those with significant p-values\n",
    "\n",
    "Parameters:\n",
    "    complex_name = name of the complex\n",
    "    test_type = type of test to perform, levene or ttest\n",
    "    \n",
    "Returns:\n",
    "    Classification of the complex: whether it has proteins missing in the normal, missing proteins in cancer,\n",
    "    or other (proteins present in both)\n",
    "    Prints tests with statistically significant values\n",
    "\n",
    "\"\"\"\n",
    "def test_complex(complex_name, test_type = 'ttest'):\n",
    "    \n",
    "    prot_list = subunitNames[complex_name]\n",
    "    sig_result_found = False\n",
    "    \n",
    "    # Keep track of how many significant ratios in this complex include only 0 values for tumor/normal \n",
    "    num_missing_tumor = 0\n",
    "    num_missing_normal = 0\n",
    "    \n",
    "    # Calculate the cutoff for significance (Bonferroni correction)\n",
    "    alpha = 0.05 / (len(subunitNames)*(len(subunitNames) - 1))\n",
    "    \n",
    "    for i in range(0, len(prot_list)):\n",
    "        for j in range(0, len(prot_list)):\n",
    "            if j == i: continue\n",
    "                \n",
    "            # Test the two proteins (ratio of prot1 / prot2)\n",
    "            tumor_ratio_df, normal_ratio_df = get_ratio_df(prot_list[i], prot_list[j])\n",
    "            if tumor_ratio_df is None or normal_ratio_df is None: continue\n",
    "\n",
    "            # Perform the selected statistical test on the ratios\n",
    "            \n",
    "            # If both tumor and normal only have 0 values for the ratio, the tests don't work\n",
    "            if (sum(tumor_ratio_df['Ratio']) == 0) and (sum(normal_ratio_df['Ratio']) == 0): continue   \n",
    "            # If there are not enough samples, the tests also don't work\n",
    "            if (len(tumor_ratio_df) < 2) or (len(normal_ratio_df) < 2): continue\n",
    "                \n",
    "            if test_type == 'ttest':\n",
    "                test_result = scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "            else:\n",
    "                # This test throws a warning if the variance in both is 0, so we check for that\n",
    "                if (np.var(tumor_ratio_df['Ratio']) == 0) and (np.var(normal_ratio_df['Ratio']) == 0): continue\n",
    "                test_result = scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "            \n",
    "            # If the p-value is significant, print the result and return the appropriate classification\n",
    "            if test_result < alpha:\n",
    "                sig_result_found = True\n",
    "                # Determine the classification of this ratio\n",
    "                if not (normal_ratio_df['Ratio'] != 0).any():\n",
    "                    num_missing_normal += 1\n",
    "                    print('Normal missing ' + prot_list[i] + ' / ' + prot_list[j])\n",
    "                elif not (tumor_ratio_df['Ratio'] != 0).any():\n",
    "                    num_missing_tumor += 1\n",
    "                    print('Tumor missing ' + prot_list[i] + ' / ' + prot_list[j])\n",
    "    \n",
    "                print(prot_list[i] + ' / ' + prot_list[j] + ': ' + str(test_result))\n",
    "                \n",
    "    if sig_result_found: \n",
    "        \n",
    "        print(complex_name)\n",
    "        print('---------------------------')\n",
    "        print('---------------------------')\n",
    "        \n",
    "        # Classify the complex as a whole depending on the classification of the majority of ratios\n",
    "        # If none of the ratios had all 0 values for tumor or normal, it is classified as \"Other\"\n",
    "        if num_missing_normal == 0 and num_missing_tumor == 0:\n",
    "            return \"Other\"\n",
    "        elif num_missing_normal >= num_missing_tumor:\n",
    "            return \"Normal Missing Proteins\"\n",
    "        elif num_missing_tumor > num_missing_normal:\n",
    "            return \"Tumor Missing Proteins\"\n",
    "        \n",
    "    return \"NS\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### test_all_complexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "test_all_complexes\n",
    "------------------\n",
    "Test all complexes in the dataset for significance\n",
    "\n",
    "Parameters:\n",
    "    test_type (optional) = ttest or levene\n",
    "    \n",
    "Returns:\n",
    "    A dictionary: {\"Normal Missing Proteins\": [List of complexes], \n",
    "    \"Tumor Missing Proteins\": [List of complexes], \n",
    "    \"Other\": [List of complexes]}\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "def test_all_complexes(test_type = 'ttest'):\n",
    "    \n",
    "    if test_type != 'ttest' and test_type != 'levene':\n",
    "        print(\"Error: test_type must equal 'ttest' or 'levene'\")\n",
    "        return\n",
    "    \n",
    "    classified_complexes = {\"Normal Missing Proteins\": [], \"Tumor Missing Proteins\": [], \"Other\": [], \"NS\": []}\n",
    "    \n",
    "    for key, val in subunitNames.items():\n",
    "        complex_type = test_complex(key, test_type = test_type)\n",
    "        classified_complexes[complex_type].append(key)\n",
    "        \n",
    "    return classified_complexes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_mutations\n",
    "--------------\n",
    "Search mutational data for mutations in proteins that change in a complex or transcription factors\n",
    "known to regulate said proteins\n",
    "\n",
    "Parameters:\n",
    "    complex_dict = dictionary of protein complexes (output of test_all_complexes)\n",
    "    transcription_factors = if True, will test for mutations in transcription factors known to regulate\n",
    "        proteins in the complex\n",
    "        \n",
    "Returns:\n",
    "    Prints results (how many patients have mutations in the protein or transcription factor)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def find_mutations(complex_dict, transcription_factors = False):\n",
    "    \n",
    "    for key, val in complex_dict.items():\n",
    "        if key == 'NS': continue\n",
    "        print(key)\n",
    "        print('--------------')\n",
    "        print('--------------')\n",
    "        for complex_name in val:\n",
    "            changed_proteins = []\n",
    "            mutation_rate = {}\n",
    "\n",
    "            # Calculate the cutoff for significance (Bonferroni correction)\n",
    "            alpha = 0.05 / (len(subunitNames)*(len(subunitNames) - 1))\n",
    "\n",
    "            # Perform tests to find statistically significant differences in ratios\n",
    "            protein_list = subunitNames[complex_name]\n",
    "            for prot1 in protein_list:\n",
    "                for prot2 in protein_list:\n",
    "                    if prot1 == prot2: continue\n",
    "\n",
    "                    # Test the two proteins (ratio of prot1 / prot2)\n",
    "                    tumor_ratio_df, normal_ratio_df = get_ratio_df(prot1, prot2)\n",
    "                    if tumor_ratio_df is None or normal_ratio_df is None: continue\n",
    "\n",
    "                    # Perform the selected statistical tests on the ratios\n",
    "                    \n",
    "                    # If both tumor and normal only have 0 values for the ratio, the tests don't work\n",
    "                    if (sum(tumor_ratio_df['Ratio']) == 0) and (sum(normal_ratio_df['Ratio']) == 0): continue\n",
    "                    # If there are not enough samples, the tests also don't work\n",
    "                    if (len(tumor_ratio_df) < 2) or (len(normal_ratio_df) < 2): continue\n",
    "                \n",
    "                    ttest_result = scipy.stats.ttest_ind(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "                    # The levene test throws a warning if the variance in both is 0, so we check for that\n",
    "                    if (np.var(tumor_ratio_df['Ratio']) != 0) and (np.var(normal_ratio_df['Ratio']) != 0):\n",
    "                        levene_result = scipy.stats.levene(tumor_ratio_df['Ratio'], normal_ratio_df['Ratio'])[1]\n",
    "                    else:\n",
    "                        levene_result = 1\n",
    "\n",
    "                    # If the p-value is significant, add the numerator protein to changed_proteins\n",
    "                    if ttest_result < alpha or levene_result < alpha:\n",
    "                        changed_proteins.append(prot1)\n",
    "\n",
    "            changed_proteins = list(set(changed_proteins))\n",
    "            for protein in changed_proteins:\n",
    "                if transcription_factors:\n",
    "                    tfs = tf_data.loc[tf_data['Regulated_Gene'] == protein]['TF']\n",
    "                    for tf in tfs:\n",
    "                        num_mutations = len(set(mutation_data.loc[mutation_data['Gene'] == tf]['Patient_ID']))\n",
    "                        tf_title = tf + ' (regulates ' + protein + ')'\n",
    "                        if num_mutations > 0:\n",
    "                            mutation_rate[tf_title] = num_mutations\n",
    "                else:\n",
    "                    num_mutations = len(set(mutation_data.loc[mutation_data['Gene'] == protein]['Patient_ID']))\n",
    "                    if num_mutations > 0:\n",
    "                        mutation_rate[protein] = num_mutations\n",
    "            if len(mutation_rate) > 0:\n",
    "                print(complex_name)\n",
    "                print(mutation_rate)\n",
    "                print('--------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_unique_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_unique_mutations\n",
    "---------------------\n",
    "Find the number of patients with a mutation in any one of the proteins in a list\n",
    "\n",
    "Parameters:\n",
    "    prot_list = list of proteins to look for mutations in\n",
    "    \n",
    "Returns:\n",
    "    prints the number of patients with a mutation in a protein in prot_list as well as their identifiers\n",
    "    returns a list of those patients\n",
    "\"\"\"\n",
    "\n",
    "def find_unique_mutations(prot_list, cancer_type = 'ov'):\n",
    "    # Set the cancer type\n",
    "    if cancer_type == 'ov': mutation_data = ov_mutation_data\n",
    "    elif cancer_type == 'colon': mutation_data = colon_mutation_data\n",
    "    elif cancer_type == 'en': mutation_data = en_mutation_data\n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "        \n",
    "    patients = []\n",
    "    mutated_proteins = {}\n",
    "    for protein in prot_list:\n",
    "        mutations = list(mutation_data.loc[mutation_data['Gene'] == protein]['Patient_ID'])\n",
    "        if len(mutations) > 0:\n",
    "            mutated_proteins[protein] = len(mutations)\n",
    "            patients = patients + mutations\n",
    "    patients = list(set(patients))\n",
    "    print('Patients with a mutation in any of the ' + str(len(prot_list)) + ' given proteins: ' + str(len(patients)) + '\\n')\n",
    "    print(mutated_proteins)\n",
    "    return patients"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ttest_proteomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "ttest_proteomics\n",
    "----------------\n",
    "Performs and prints a basic t-test on the proteomics of a given protein (tumor vs normal)\n",
    "\n",
    "Parameters:\n",
    "    protein = protein to test\n",
    "    use_cptac (optional) = T/F whether to use the CPTAC proteomics data (default) or the non-normalized data\n",
    "    \n",
    "Returns:\n",
    "    p-value of the t-test\n",
    "\"\"\"\n",
    "\n",
    "def ttest_proteomics(protein, use_cptac = True, cancer_type = 'ov'):\n",
    "    \n",
    "    # This uses the normalized proteomics data from the CPTAC package\n",
    "    if use_cptac:\n",
    "        # Get data from the appropriate cancer type\n",
    "        if cancer_type == 'ov':\n",
    "            test_data = ov.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'colon': \n",
    "            test_data = colon.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'en': \n",
    "            test_data = en.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        elif cancer_type == 'renal':\n",
    "            test_data = renal.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                omics_genes = protein)\n",
    "        else: print('Error: cancer_type must be \"ov\", \"colon\", \"en\", or \"renal\"')\n",
    "            \n",
    "        test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "        if cancer_type == 'en':\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            test_data.loc[test_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "        else:\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "        tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "        normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "    \n",
    "    # Otherwise use the non-normalized data to make the plot\n",
    "    else:\n",
    "        test_data = pd.DataFrame(data.loc[data.index == protein].transpose())\n",
    "        test_data['Sample_Type'] = 'Tumor'\n",
    "        test_data.loc[test_data.index.str.contains('_NM'), 'Sample_Type'] = 'Normal'\n",
    "\n",
    "        tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor']\n",
    "        normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal']\n",
    "    \n",
    "    # Perform the t-test\n",
    "    return scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### find_differential_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "find_differential_expression\n",
    "----------------------------\n",
    "Searches through all the proteomic data and finds proteins with differential expression in tumor cells using t-tests.\n",
    "Organizes results into categories: tumor lower and normal lower\n",
    "\n",
    "Parameters:\n",
    "    cancer_type = 'ov', 'en', or 'colon'; the cancer to search for\n",
    "    \n",
    "Returns:\n",
    "    A dictionary of the form {tumor_lower: [list_of_proteins], normal_lower: [list_of_proteins]}\n",
    "    'tumor_lower' means that there are lower expression levels of that protein in tumor samples\n",
    "    \n",
    "\"\"\"\n",
    "\n",
    "def find_differential_expression(cancer_type = 'ov'):\n",
    "    results = {'tumor_lower': [], 'normal_lower': []}\n",
    "    \n",
    "    # Get data from the appropriate cancer type\n",
    "    if cancer_type == 'ov':\n",
    "        alpha = 0.05 / len(ov.get_proteomics().columns)\n",
    "        for protein in ov.get_proteomics().columns:\n",
    "            test_data = ov.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                    metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                    omics_genes = protein)\n",
    "            test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "\n",
    "            tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "            normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "            \n",
    "            pval = scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]\n",
    "            if pval < alpha:\n",
    "                mean_tumor = tumor_df[protein].mean()\n",
    "                mean_normal = normal_df[protein].mean()\n",
    "                if mean_tumor < mean_normal: \n",
    "                    results['tumor_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (tumor lower)')\n",
    "                else: \n",
    "                    results['normal_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (normal lower)')\n",
    "            \n",
    "        \n",
    "    elif cancer_type == 'colon': \n",
    "        alpha = 0.05 / len(colon.get_proteomics().columns)\n",
    "        for protein in colon.get_proteomics().columns:\n",
    "            test_data = colon.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                    metadata_cols = ['Patient_ID', 'Sample_Tumor_Normal'], \n",
    "                                                    omics_genes = protein)\n",
    "            test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Sample_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "\n",
    "            tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "            normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "            \n",
    "            pval = scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]\n",
    "            if pval < alpha:\n",
    "                mean_tumor = tumor_df[protein].mean()\n",
    "                mean_normal = normal_df[protein].mean()\n",
    "                if mean_tumor < mean_normal: \n",
    "                    results['tumor_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (tumor lower)')\n",
    "                else: \n",
    "                    results['normal_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (normal lower)')\n",
    "        \n",
    "    elif cancer_type == 'en': \n",
    "        alpha = 0.05 / len(en.get_proteomics().columns)\n",
    "        for protein in en.get_proteomics().columns:\n",
    "            test_data = en.join_metadata_to_omics('clinical', 'proteomics', \n",
    "                                                    metadata_cols = ['Patient_ID', 'Proteomics_Tumor_Normal'], \n",
    "                                                    omics_genes = protein)\n",
    "            test_data = test_data.loc[:, ~test_data.columns.duplicated()]\n",
    "            test_data.rename(columns={protein+'_proteomics': protein, 'Proteomics_Tumor_Normal': 'Sample_Type'}, inplace = True)\n",
    "            test_data.loc[test_data['Sample_Type'] != 'Tumor', 'Sample_Type'] = 'Normal'\n",
    "\n",
    "            tumor_df = test_data.loc[test_data['Sample_Type'] == 'Tumor'].dropna(axis = 0)\n",
    "            normal_df = test_data.loc[test_data['Sample_Type'] == 'Normal'].dropna(axis = 0)\n",
    "            \n",
    "            pval = scipy.stats.ttest_ind(tumor_df[protein], normal_df[protein])[1]\n",
    "            if pval < alpha:\n",
    "                mean_tumor = tumor_df[protein].mean()\n",
    "                mean_normal = normal_df[protein].mean()\n",
    "                if mean_tumor < mean_normal: \n",
    "                    results['tumor_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (tumor lower)')\n",
    "                else: \n",
    "                    results['normal_lower'].append(protein)\n",
    "                    print(protein + ': ' + str(pval) + ' (normal lower)')\n",
    "        \n",
    "    else: print('Error: cancer_type must be \"ov\", \"colon\", or \"en\"')\n",
    "    \n",
    "    return results\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tidy_split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Thanks StackOverflow! \n",
    "# https://stackoverflow.com/questions/12680754/split-explode-pandas-dataframe-string-entry-to-separate-rows\n",
    "\n",
    "def tidy_split(df, column, sep='|', keep=False):\n",
    "    \"\"\"\n",
    "    Split the values of a column and expand so the new DataFrame has one split\n",
    "    value per row. Filters rows where the column is missing.\n",
    "\n",
    "    Params\n",
    "    ------\n",
    "    df : pandas.DataFrame\n",
    "        dataframe with the column to split and expand\n",
    "    column : str\n",
    "        the column to split and expand\n",
    "    sep : str\n",
    "        the string used to split the column's values\n",
    "    keep : bool\n",
    "        whether to retain the presplit value as it's own row\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pandas.DataFrame\n",
    "        Returns a dataframe with the same columns as `df`.\n",
    "    \"\"\"\n",
    "    indexes = list()\n",
    "    new_values = list()\n",
    "    #df = df.dropna(subset=[column])\n",
    "    for i, presplit in enumerate(df[column].astype(str)):\n",
    "        values = presplit.split(sep)\n",
    "        if keep and len(values) > 1:\n",
    "            indexes.append(i)\n",
    "            new_values.append(presplit)\n",
    "        for value in values:\n",
    "            indexes.append(i)\n",
    "            new_values.append(value)\n",
    "    new_df = df.iloc[indexes, :].copy()\n",
    "    new_df[column] = new_values\n",
    "    return new_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
